// oae-authz/lib/api.js
/*!
 * Copyright 2012 Sakai Foundation (SF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 * 
 *     http://www.osedu.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');

var Cassandra = require('oae-util/lib/cassandra');
var Telemetry = require('oae-telemetry').telemetry('authz');
var Validator = require('oae-authz/lib/validator').Validator;
var AuthzUtil = require('oae-authz/lib/util');
var AuthzConstants = require('oae-authz/lib/constants').AuthzConstants;


/////////////////////////
// ROLES & PERMISSIONS //
/////////////////////////

/**
 * Determine the direct role assigned to a principal on a specified resource.
 * 
 * @param  {String}     principalId     The principal id. This can be a user or a group
 * @param  {String}     resourceId      The resource id. This can be a group as well.
 * @param  {Function}   callback        Standard callback function 
 * @param  {Object}     callback.err    An error that occurred, if any.
 * @param  {String}     callback.role   The role of the principal on the resource. If the principal has no role or there is an error performing the check, role will be null.
 * @api private
 */
var _getDirectRole = function(principalId, resourceId, callback) {
    getDirectRoles([principalId], resourceId, function(err, roles) {
        if (err) {
            return callback(err);
        }

        return callback(null, roles[principalId]);
    });
};

/**
 * Determine the role assigned to the given principals directly on the specified resource instance.
 * 
 * @param   {String[]}     principalIds    Array of principal ids. These can be user or group ids
 * @param   {String}       resourceId      The resource id. This can be a group as well.
 * @param   {Function}     callback        Standard callback function
 * @param   {Object}       callback.err    An error that occurred, if any.
 * @param   {Object}       callback.roles  A hash keyed by principal id, with value set to the role they have directly on the resource. If the principal has no role or there is an error performing the check, role will be null.
 **/
var getDirectRoles = module.exports.getDirectRoles = function(principalIds, resourceId, callback) {
    Cassandra.runQuery('SELECT ? FROM AuthzMembers USING CONSISTENCY QUORUM WHERE resourceId = ?', [principalIds, resourceId], function(err, rows) {
        if (err) {
            return callback(err);
        }

        // Aggregate the roles for the principals
        var roles = Cassandra.rowToHash(rows[0]);
        return callback(null, roles);
    });
};

/**
 * Given a principal and a resource, determine all the effective roles that the principal has on the resource, by virtue of direct
 * association and indirect group inheritance.
 *
 * @param  {String}       principalId       The principal id. This can be a user or a group
 * @param  {String}       resourceId        The resource id. This can be a group as well.
 * @param  {Function}     callback          Standard callback function
 * @param  {Object}       callback.err      An error that occurred, if any
 * @param  {String[]}     callback.roles    An array containing all the roles the principal has on the resource.
 */
var getAllRoles = module.exports.getAllRoles = function(principalId, resourceId, callback) {
    var validator = new Validator();
    validator.check(principalId, {'code': 400, 'msg': 'Invalid principal id provided.'}).isPrincipalId();
    validator.check(resourceId, {'code': 400, 'msg': 'Invalid resource id provided.'}).isResourceId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _getAllRoles(principalId, resourceId, callback);
};

/**
 * Given a principal and a resource, determine all the roles that the principal has on the resource, by virtue of direct association and
 * indirect group inheritance.
 *
 * @param   {String}       principalId     The principal id. This can be a user or a group
 * @param   {String}       resourceId      The resource id. This can be a group as well.
 * @param   {Function}     callback        Standard callback function
 * @param   {Object}       callback.err    An error that occurred, if any
 * @param   {String[]}     callback.roles  An array containing all the roles the principal has on the resource.
 * @api private
 */
var _getAllRoles = function(principalId, resourceId, callback) {
    var allRoles = [];
    // Get the direct role of the user
    _getDirectRole(principalId, resourceId, function(err, directRole) {
        if (err) {
            return callback(err);
        } else if (directRole) {
            allRoles.push(directRole);
        }

        // Get the groups that are directly and indirectly associated to the resource
        _getResourceGroupHierarchy(resourceId, function(err, groups) {
            if (err) {
                return callback(err);
            }
            
            // Check whether any of these are part of the user's direct memberships
            var groupIds = _.keys(groups);
            if (groupIds.length === 0) {
                return callback(null, allRoles);
            } else {
                Cassandra.runQuery('SELECT ? FROM AuthzRoles USING CONSISTENCY QUORUM WHERE principalId = ?', [groupIds, principalId], function(err, rows) {
                    if (err) {
                        return callback(err);
                    }
                    
                    // Add the roles of the matching groups
                    var memberships = Cassandra.rowToHash(rows[0]);
                    for (var m in memberships) {
                        if (memberships[m]) {
                            // Groups can have multiple roles on a resource (through indirect memberships),
                            // so we have to add all of them
                            for (var r = 0; r < groups[m].roles.length; r++) {
                                if (!_.contains(allRoles, groups[m].roles[r])) {
                                    allRoles.push(groups[m].roles[r]);
                                }
                            }
                        }
                    }
                    return callback(null, allRoles);
                });
            }
        });
    });
};

/**
 * Get all of the groups that are directly and indirectly associated to a resource.
 * 
 * @param   {String}       resourceId      The resource id. This can be a group as well.
 * @param   {Function}     callback        Standard callback function
 * @param   {Object}       callback.err    An error that occurred, if any
 * @param   {Object}       callback.roles  A JSON object where the keys are the group ids of the directly and indirectly associated groups and the values are arrays specifying that groups direct and indirect roles on the resouce
 * @api private
 */
var _getResourceGroupHierarchy = function(resourceId, callback) {
    // Get the groups that are directly associated to the resource
    var start = AuthzConstants.principalTypes.GROUP + ':';
    var end = start + '|';
    Cassandra.runQuery('SELECT ? .. ? FROM AuthzMembers USING CONSISTENCY QUORUM WHERE resourceId = ?', [start, end, resourceId], function(err, rows) {
        if (err) {
            return callback(err);
        }

        // If no groups are associated to the content, we can just return
        var groupHierarchy = Cassandra.rowToHash(rows[0]);
        var groupHierarchyIds = _.keys(groupHierarchy);
        if (groupHierarchyIds.length === 0) {
            callback(null, {});
        } else {
            
            // Explode the group memberships downwards
            var explodeGroupMemberships = function(groupIds) {
                Cassandra.runQuery('SELECT ? .. ? FROM AuthzMembers USING CONSISTENCY QUORUM WHERE resourceId IN (?)', [start, end, groupIds], function(err, rows) {
                    if (err) {
                        callback(err);
                    }

                    // Check if any of the returned groups need to be exploded further
                    var newToExplode = [];
                    for (var i = 0; i < rows.length; i++) {
                        var groups = Cassandra.rowToHash(rows[i]);
                        for (var g in groups) {
                            // Fill in the children of the parent group
                            if (groupHierarchy[rows[i].key].children.indexOf(g) === -1) {
                                groupHierarchy[rows[i].key].children.push(g);
                            }
                            // The group isn't part of the groupHierarchy yet.
                            if (!groupHierarchy[g]) {
                                newToExplode.push(g);
                                groupHierarchy[g] = {
                                    'roles': groupHierarchy[rows[i].key].roles.slice(),
                                    'children': []
                                }
                            // If the group is already part of the groupHierarchy, we add the roles of the parent group. If any of them weren't in the
                            // roles list yet, we recursively loop over all of the group children to add the roles to all of its children
                            } else {
                                var rolesNotOnChild = _.difference(groupHierarchy[rows[i].key].roles, groupHierarchy[g].roles);
                                // If a new role has been found
                                if (rolesNotOnChild.length > 0) {
                                    
                                    var addRolesToChild = function(groupId, rolesToAdd) {
                                        // Add the roles
                                        groupHierarchy[groupId].roles = groupHierarchy[groupId].roles.concat(rolesToAdd);
                                        // Add the roles to all of its children as well
                                        for (var c = 0; c < groupHierarchy[groupId].children.length; c++) {
                                            var rolesNotOnChild = _.difference(groupHierarchy[groupId].roles, groupHierarchy[groupHierarchy[groupId].children[c]].roles);
                                            if (rolesNotOnChild.length > 0) {
                                                addRolesToChild(groupHierarchy[groupId].children[c], rolesNotOnChild);
                                            }
                                        }
                                    }
                                    addRolesToChild(g, rolesNotOnChild);
                                }
                            }
                        }
                    }

                    // If no groups need to be exploded, we can return
                    if (newToExplode.length === 0) {
                        callback(null, groupHierarchy);
                    } else {
                        explodeGroupMemberships(newToExplode);
                    }
                });
            }
            
            // Convert the direct group members to be able to take multiple roles (which will be filled
            // in because of indirect memberships) and the list of child groups
            for (var g in groupHierarchy) {
                groupHierarchy[g] = {
                    'roles': [groupHierarchy[g]],
                    'children': []
                }
            }
            // Start by exploding the groups directly associated to the resource
            explodeGroupMemberships(groupHierarchyIds);
            
        }
    });
};

/**
 * Determines whether or not a principal has the specified role directly or indirectly on a given resource.
 * 
 * @param  {String}      principalId         The principal id. This can be a user or a group.
 * @param  {String}      resourceId          The resource id. This can be a group as well
 * @param  {String}      role                The role to check
 * @param  {Function}    callback            A function specifying whether or not the user has the role on the resource
 * @param  {Object}      callback.err        An error that occurred, if any
 * @param  {Boolean}     callback.hasRole    Whether or not the principal has the specified role on the resource
 */
var hasRole = module.exports.hasRole = function(principalId, resourceId, role, callback) {
    var validator = new Validator();
    validator.check(principalId, {'code': 400, 'msg': 'Invalid principal id provided.'}).isPrincipalId();
    validator.check(resourceId, {'code': 400, 'msg': 'Invalid resource id provided.'}).isResourceId();
    validator.check(role, {'code': 400, 'msg': 'Invalid role provided.'}).isValidRole();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _hasRole(principalId, resourceId, role, callback);
};

/**
 * Determines whether or not a principal has any role directly or indirectly on a given resource.
 * 
 * @param  {String}      principalId         The principal id. This can be a user or a group.
 * @param  {String}      resourceId          The resource id. This can be a group as well
 * @param  {Function}    callback            A function specifying whether or not the user has the role on the resource
 * @param  {Object}      callback.err        An error that occurred, if any
 * @param  {Boolean}     callback.hasRole    Whether or not the principal has a role on the resource
 */
var hasAnyRole = module.exports.hasAnyRole = function(principalId, resourceId, callback) {
    var validator = new Validator();
    validator.check(principalId, {'code': 400, 'msg': 'Invalid principal id provided.'}).isPrincipalId();
    validator.check(resourceId, {'code': 400, 'msg': 'Invalid resource id provided.'}).isResourceId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _hasRole(principalId, resourceId, null, callback);
};

/**
 * Determines whether or not a principal has the specified role directly or indirectly on a given resource.
 * 
 * @param  {String}       principalId         The principal id. This can be a user or a group.
 * @param  {String}       resourceId          The resource id. This can be a group as well
 * @param  {String}       role                The role to check. If the role is null, we check for any role
 * @param  {Function}     callback            A function specifying whether or not the user has the role on the resource
 * @param  {Object}       callback.err        An error that occurred, if any
 * @param  {Boolean}      callback.hasRole    Whether or not the principal has the specified role on the resource
 * @api private
 */
var _hasRole = function(principalId, resourceId, role, callback) {
    // Check for a direct role first
    _getDirectRole(principalId, resourceId, function(err, directRole) {
        if (err) {
            return callback(err);
        } else if (directRole && (role === null || directRole === role)) {
            return callback(null, true);
        }
        
        // If no direct role assignment is found, we try to find a role through an indirect membership
        getAllRoles(principalId, resourceId, function(err, roles) {
            if (err) {
                return callback(err);
            }
            
            // If a role is found and we are just looking for any role
            if (roles.length > 0 && role === null) {
                return callback(null, true);
            // If we are looking for a specific role and that specific role is present
            } else if (_.contains(roles, role)) {
                return callback(null, true);
            // If the specified role cannot be found
            } else {
                callback(null, false);
            }
        });
    });
};

/**
 * Assign one or multiple principals a role on a resource instance. If the user already has a role, it will simply be updated. When
 * false is passed in as a role, the role for that principal will be removed.
 *
 * @param  {String}          resourceId      The resource id.
 * @param  {Object}          changes         JSON object where the keys are principal ids and the values are role values
 * @param  {Function}        callback        Standard callback function
 * @param  {Object}          callback.err    Error object, containing the error message
 */
var updateRoles = module.exports.updateRoles = function(resourceId, changes, callback) {
    var validator = new Validator();
    validator.check(resourceId, {'code': 400, 'msg': 'Invalid resource id provided.'}).isResourceId();
    var roleChanges = _.keys(changes);
    validator.check(roleChanges.length, {'code': 400, 'msg': 'At least one role change needs to be applied'}).min(1);
    for (var i = 0; i < roleChanges.length; i++) {
        var principalId = roleChanges[i];
        validator.check(principalId, {'code': 400, 'msg': 'Invalid principal id specified: ' + principalId}).isPrincipalId();
        validator.check(changes[principalId], {'code': 400, 'msg': 'Invalid role provided.'}).isValidRoleChange();
    }
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _updateRoles(resourceId, changes, callback);
};

/**
 * Assign multiple principals a role on a resource instance. If the user already has a role, it will simply be updated. When
 * false is passed in as a role, the role for that principal will be removed.
 *
 * @param  {String}        resourceId      The resource id.
 * @param  {Object}        changes         JSON object where the keys are principal ids and the values are role values
 * @param  {Function}      callback        Standard callback function
 * @param  {Object}        callback.err    Error object, containing the error message
 * @api private
 */
var _updateRoles = function(resourceId, changes, callback) {
    var queries = [];
    var roleChanges = _.keys(changes);
    for (var i = 0; i < roleChanges.length; i++) {
        var principalId = roleChanges[i];
        if (changes[principalId]) {
            queries.push({
                'query': 'UPDATE AuthzRoles SET ? = ? WHERE principalId = ?',
                'parameters': [resourceId, changes[principalId], principalId]
            });
            queries.push({
                'query': 'UPDATE AuthzMembers SET ? = ? WHERE resourceId = ?',
                'parameters': [principalId, changes[principalId], resourceId]
            });
        } else if (changes[principalId] === false) {
            queries.push({
                'query': 'DELETE ? FROM AuthzRoles WHERE principalId = ?',
                'parameters': [resourceId, principalId]
            });
            queries.push({
                'query': 'DELETE ? FROM AuthzMembers WHERE resourceId = ?',
                'parameters': [principalId, resourceId]
            });
        } else {
            return callback({'code': 400, 'msg': 'An invalid role has been passed in (undefined)'});
        }
    }
    Cassandra.runBatchQuery(queries, 'QUORUM', callback);
};


///////////////////
// AUTHZ MEMBERS //
///////////////////

/**
 * Get all the direct members of a resource and their role on the resource.
 *
 * @param  {String}      resourceId              A unique identifier for a resource. ex: g:cam-oae-team or c:cam:XCDSasD
 * @param  {String}      start                   The principal id that comes just before the first principal you wish to have in your results.
 * @param  {Number}      limit                   The number of members you wish to retrieve.
 * @param  {Function}    callback                Standard callback function
 * @param  {Object}      callback.err            An error that occurred, if any
 * @param  {Object[]}    callback.members        Array of objects for each of the direct member of the resource. Each object has an 'id' key containing the principal id of the member and a 'role' key containing the role of that principal
 */
var getAuthzMembers = module.exports.getAuthzMembers = function(resourceId, start, limit, callback) {
    limit = limit || 10;

    var validator = new Validator();
    validator.check(resourceId, {'code': 400, 'msg': 'Invalid resource id provided'}).isResourceId();
    validator.check(limit, {'code': 400, 'msg': 'The limit parameter needs to be a number'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Page the query.
    var paging = Cassandra.getPagingParameters(start, limit);
    Cassandra.runQuery('SELECT FIRST ' + paging.limit + ' ?..\'\' FROM AuthzMembers USING CONSISTENCY QUORUM WHERE resourceId = ?', [paging.start, resourceId], function (err, rows) {
        if (err) {
            return callback(err);
        }

        var members = [];
        var row = rows[0];
        for (var i = paging.startIndex, j = row.count; i < j; i++) {
            if (row[i].name !== "groupId") {
                members.push({
                    'id': row[i].name,
                    'role': row[i].value
                });
            }
        }

        return callback(false, members);
    });
};

/**
 * Gets all the Authz groups of which a principal (either user or group) is a member. This includes all group ancestors to 
 * which the user is indirectly a member.
 *
 * @param  {String}           principalId             The principal id for which to retrieve all the group memberships
 * @param  {Function}         callback                Standard callback function
 * @param  {Object}           callback.err            An error that occured, if any
 * @param  {String[]}         callback.groups         An (inconsistently ordered) array of groups to which the user belongs, either directly or indirectly
 */
var getPrincipalMemberships = module.exports.getPrincipalMemberships = function(principalId, callback) {
    var validator = new Validator();
    validator.check(principalId, {'code': 400, 'msg': 'Invalid principal id provided'}).isPrincipalId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _getAuthzGroupMembershipAncestry(principalId, function(err, groups) {
        if (err) {
            return callback(err);
        }
        callback(null, _.keys(groups));
    });
};

/**
 * Given some principal, get their entire group membership ancestry. In other words, get all the groups to which they belong,
 * and the groups to all those groups belong, and so on.
 *
 * @param   {String}        principalId         The principal id for which to retrieve the authz group membership ancestry
 * @param   {Function}      callback            Standard callback function
 * @param   {Object}        callback.err        The err that occured, if any
 * @param   {Object}        callback.ancestry   A hash keyed by the group id of each group of which the user is indirectly a member. The value of each entry is `true`.
 * @api private
 */
var _getAuthzGroupMembershipAncestry = function(principalId, callback) {
    var results = {};

    var _multiGetAuthzGroupMembershipAncestry = function(principalIds, results) {
        getRolesForPrincipalsAndResourceType(principalIds, AuthzConstants.resourceTypes.GROUP, 10000, function(err, entries) {
            if (err) {
                return callback(err);
            }

            var nextPrincipalBatch = [];
            var ancestors = Object.keys(entries);
            // for each group parent, determine if we've already recorded their membership. if not, we record it, and search it's parents
            for (var i = 0; i < ancestors.length; i++) {
                var ancestorId = ancestors[i];
                if (!results[ancestorId]) {
                    results[ancestorId] = true;
                    nextPrincipalBatch.push(ancestorId);
                }
            }

            // if we found new ancestors, we need to recurse to find their ancestors. Otherwise we're done.
            if (nextPrincipalBatch.length > 0) {
                return _multiGetAuthzGroupMembershipAncestry(nextPrincipalBatch, results);
            } else {
                return callback(false, results);
            }
        });
    };

    _multiGetAuthzGroupMembershipAncestry([principalId], results, callback);
};

/**
 * Get all principal roles associated to the resourceType for all the principals in the array of principal ids. This can be performed on 
 * multiple principals at once.
 * 
 * The structure of the resulting entries is structured a hash, indicating all resource ids that were returned and what the
 * roles associated to those resources were. For example:
 *
 *     {
 *      'u:cam:simong': { 'member': true },
 *      'g:cam:group-b': { 'member': true }
 *     }
 *
 * or:
 *
 *     {
 *      'c:gat:Foo.docx': { 'viewer': true, 'manager': true},
 *      'c:cam:Bar.pdf':  { 'viewer': true }
 *     }
 *
 * In the latter case, 'c:gat:Foo.docx' has both 'viewer' and 'manager' role associated to it. This indicates that some principal
 * had a role of 'viewer' on the resource, while some other principal had the 'manager' role.
 *
 * @param  {Array}          principalIds        The array of principal ids to query for
 * @param  {String}         resourceType        The resource type of the resources to search for, as determined by Resource.resourceType
 * @param  {Number}         limit               The maximum number of resources to return per user
 * @param  {Function}       callback            A function providing all the {resource id} -> {roles} mappings aggregated for all the provided principals queried
 * @param  {Object}         callback.err        An error that occurred, if any
 * @param  {Object}         callback.entries    A JSON Object representing the roles associated to the users
 */
var getRolesForPrincipalsAndResourceType = module.exports.getRolesForPrincipalsAndResourceType = function(principalIds, resourceType, limit, callback) {
    limit = limit || 1000;
    principalIds = principalIds || [];

    var validator = new Validator();
    validator.check(limit, {'code': 400, 'msg': '"limit" parameter must be a number when searching for active roles'});
    validator.check(resourceType, {'code': 400, 'msg': 'A resourceType needs to be provided'}).notEmpty();
    validator.check(principalIds.length, {'code': 400, 'msg': 'At least one principal Id needs to be passed in'}).min(1);
    for (var i = 0; i < principalIds.length; i++) {
        var principalId = principalIds[i];
        validator.check(principalId, {'code': 400, 'msg': 'Invalid principal id specified: ' + principalId}).isPrincipalId();
    }
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // we append a '|' to the "end" range, as | has a high ASCII alphabetical ordering. This may not suffice if resourceIds have
    // multi-byte characters, which is technically possible. Unfortunately, I don't think there is a better way to do this with
    // CQL.
    var start = resourceType + ':';
    var end = start + '|';

    var cql = 'SELECT FIRST ' + limit + ' ? .. ? FROM AuthzRoles USING CONSISTENCY QUORUM WHERE principalId IN (?)';
    Cassandra.runQuery(cql, [start, end, principalIds], function(err, rows) {
        if (err) {
            return callback(err);
        }

        var entries = {};

        // Aggregate all resources from all the rows (principals) into the entries hash
        var expandRow = function(name, value) {
            if (!entries[name]) {
                entries[name] = {};
            }
            entries[name][value] = true;
        };
        for (var i = 0; i < rows.length; i++) {
            rows[i].forEach(expandRow);
        }
        
        callback(null, entries);
    });
};
