// oae-authz/lib/util.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');

var AuthzConstants = require('oae-authz/lib/constants').AuthzConstants;
var AuthzDelete = require('oae-authz/lib/delete');
var AuthzModel = require('oae-authz/lib/model');

/**
 * Construct a resource based on the given id.
 *
 * @param  {String}      id     A resouce id
 * @return {Resource}           A AuthzModel.Resource object derived from this id
 */
var getResourceFromId = module.exports.getResourceFromId = function(id) {
    return unpackId(AuthzModel.Resource, id);
};

/**
 * Construct a Principal based on the given id.
 *
 * @param  {String}     id      A principal id
 * @return {Principal}          A AuthzModel.Principal object derived from this id
 */
var getPrincipalFromId = module.exports.getPrincipalFromId = function(id) {
    return unpackId(AuthzModel.Principal, id);
};

/**
 * Construct an id based on the given type, tenantAlias and instanceId of a resource or principal.
 *
 * @param  {String}     type            The type of object, as defined by either AuthzConstants.principalTypes or AuthzConstants.resourceTypes
 * @param  {String}     tenantAlias     The alias of the tenant
 * @param  {String}     instanceId      The ID of the principal or resource instance
 * @return {String}                     The ID constructed from the given parameters, i.e., "type:tenantAlias:instanceId"
 */
var toId = module.exports.toId = function(type, tenantAlias, instanceId) {
    return type + ':' + tenantAlias + ':' + instanceId;
};

/**
 * Determine whether or not the given id represents a group
 *
 * @param  {String}     groupId     The group id to check
 * @return {Boolean}                Whether or not the id is a group id
 */
var isGroupId = module.exports.isGroupId = function(groupId) {
    return (isResourceId(groupId) && groupId.indexOf(AuthzConstants.resourceTypes.GROUP + ':') === 0);
};

/**
 * Determine whether or not the given id represents a user principal.
 *
 * @param  {String}     userId  The id of the principal
 * @return {Boolean}            Whether or not the id is a user id
 */
var isUserId = module.exports.isUserId = function(userId) {
    return (isResourceId(userId) && userId.indexOf(AuthzConstants.principalTypes.USER + ':') === 0);
};

/**
 * Determine if the given string is a valid email
 *
 * @param  {String}     resourceId      The string to validate
 * @return {Boolean}                    Whether or not the string is a valid email
 */
var isEmail = module.exports.isEmail = function(email) {
    return (_.isString(email) && email.split('@').length === 2);
};

/**
 * Parse the given share target from it's 3 known forms:
 *
 *  * Principal id of the form: `u:oae:abc123` or `g:oae:abc123`
 *  * Email of the form: `username@domain.com`
 *  * User Email of the form: `username@domain.com:u:oae:abc123`
 *
 * The user-email form indicates a user id is requested, however we grant a user the ability to
 * interact with another user if they demonstrate that they know their email address. Therefore it
 * acts as a credential to bypass interaction checks with a user.
 *
 * That said the result of this function has 3 permutations:
 *
 *  * Only `principalId` is set: The share target referenced a user or group, and no email was provided. Interaction checks are needed on the principal
 *  * Only `email` is set: The share target is an email invitation, not a direct share
 *  * Both `principalId` and `email` are set: The share target is referencing a user, and the email must match that of the user. If it doesn't match, then sharing should fail
 *
 * @param  {String}     shareTargetStr          The string share target to parse
 * @return {Object}     [target]                An object representing the target. If unspecified, then it is not a valid share target
 * @return {String}     [target.principalId]    If this target referenced a principal, then this is the principal id
 * @return {String}     [target.email]          The email that was referenced
 */
var parseShareTarget = module.exports.parseShareTarget = function(shareTargetStr) {
    if (!_.isString(shareTargetStr)) {
        return null;
    }

    var shareTargetSplit = shareTargetStr.split(':');
    if (isEmail(shareTargetSplit[0])) {
        var email = shareTargetSplit[0].toLowerCase();

        // If the first entry is an email, it's either an email or an email+user
        if (shareTargetSplit.length === 1) {
            return {'email': email};
        } else {
            var userId = shareTargetSplit.slice(1).join(':');
            if (isUserId(userId)) {
                return {'email': email, 'principalId': userId};
            }
        }
    } else if (isPrincipalId(shareTargetStr)) {
        return {'principalId': shareTargetStr};
    }

    return null;
};

/**
 * Determine if the given object is a valid resource object
 *
 * @param  {Object}     resource    The resource to validate
 * @return {Boolean}                Whether or not the object is a valid resource
 */
var isResource = module.exports.isResource = function(resource) {
    return (
        _.isObject(resource) &&
        (isResourceId(resource.id) || _.isString(resource.email)) &&
        _.isObject(resource.tenant) &&
        _.isString(resource.tenant.alias) &&
        isVisibility(resource.visibility)
    );
};

/**
 * Determine whether or not the given id represents a principal.
 *
 * @param  {String}     principalId     The id of the principal
 * @return {Boolean}                    Whether or not the id is a principal id
 */
var isPrincipalId = module.exports.isPrincipalId = function(principalId) {
    return (isGroupId(principalId) || isUserId(principalId));
};

/**
 * Determine if the given string is a valid resource id
 *
 * @param  {String}     resourceId      The string to validate
 * @return {Boolean}                    Whether or not the string is a valid resource id
 */
var isResourceId = module.exports.isResourceId = function(resourceId) {
    return (_.isString(resourceId) && resourceId.split(':').length >= 3);
};

/**
 * Determine if the given string is a valid role
 *
 * @param  {String}     resourceId      The string to validate
 * @return {Boolean}                    Whether or not the string is a valid role
 */
var isRole = module.exports.isRole = function(role) {
    return (_.contains(AuthzConstants.role.ALL_PRIORITY, role));
};

/**
 * Determine if the given string is a valid visibility
 *
 * @param  {String}     resourceId      The string to validate
 * @return {Boolean}                    Whether or not the string is a valid visibility
 */
var isVisibility = module.exports.isVisibility = function(visibility) {
    return (_.contains(AuthzConstants.visibility.ALL_PRIORITY), visibility);
};

/**
 * Convenience method to combine the similar operation of splitting an id
 *
 * @api private
 */
var unpackId = function(type, id) {
    var split = id.split(':');
    return new type(split[0], split[1], split.slice(2).join(':'));
};

/**
 * Extract the group ids from a set of principals
 *
 * @param  {Group|User}     principals      The principals from which to find the group ids
 * @return {String[]}                       The ids of the groups in the given set of principals
 */
var getGroupIds = module.exports.getGroupIds = function(principals) {
    return _.chain(principals).pluck('id').filter(isGroupId).value();
};

/**
 * Given a list of ids, filter them down to those that aren't deleted
 *
 * @param  {String[]}   ids             The list of resource ids to filter
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occurred, if any
 * @param  {String[]}   callback.ids    The ids from the original list that aren't deleted
 */
var filterDeletedIds = module.exports.filterDeletedIds = function(ids, callback) {
    AuthzDelete.isDeleted(ids, function(err, deletedIds) {
        if (err) {
            return callback(err);
        }

        return callback(null, _.filter(ids, function(id) {
            return (!deletedIds[id]);
        }));
    });
};

/**
 * Compute a standard `IdChangeInfo` object based on the given roles change information. The
 * resulting object is similar to the ChangeInfo objects, however since it does not work with full
 * resource objects, it will simply result in the specified ids in place of resources. Therefore
 * rather than an "members"/"emails" object to enumerate those added, udpated and removed, it has an
 * "ids" object instead
 *
 * @param  {IdRoles}        rolesBefore         Describes the base set of roles to work with
 * @param  {IdRoles}        roleChanges         Describes the role changes to compute on the base set of roles
 * @param  {Object}         [opts]              Optional arguments describing how to compute the changes
 * @param  {Boolean}        [opts.promoteOnly]  When `true`, indicates that only changes that signify a role addition or promotion should be acknowledged. Any changes that are not promotions would then simply be dropped. Default: `false`
 * @return {IdChangeInfo}                       The computed change information
 */
var computeRoleChanges = module.exports.computeRoleChanges = function(rolesBefore, roleChanges, opts) {
    opts = opts || {};
    opts.promoteOnly = opts.promoteOnly || false;

    var result = {
        'changes': _.extend({}, roleChanges),
        'roles': {
            'before': rolesBefore,
            'after': _.extend({}, rolesBefore)
        },
        'ids': {
            'added': [],
            'updated': [],
            'removed': []
        }
    };

    // If we're only taking into consideration promotions, remove any role changes that are removals
    // or already exist as an inferior role in the `rolesBefore` collection
    var rolesPriority = AuthzConstants.role.ALL_PRIORITY;
    if (opts.promoteOnly) {
        _.each(result.changes, function(role, id) {
            if (role === false) {
                delete result.changes[id];
            } else if (result.roles.before[id]) {
                var beforeIndex = _.indexOf(rolesPriority, result.roles.before[id]);
                var changeIndex = _.indexOf(rolesPriority, role);
                if (changeIndex <= beforeIndex) {
                    delete result.changes[id];
                }
            }
        });
    }

    // Apply all of the role changes in-memory
    _.each(result.changes, function(change, id) {
        if (change === false && result.roles.after[id]) {
            // The member `id` will be removed
            delete result.roles.after[id];
            result.ids.removed.push(id);
        } else if (_.isString(change) && !result.roles.after[id]) {
            // The member `id` will be added
            result.roles.after[id] = change;
            result.ids.added.push(id);
        } else if (_.isString(change) && result.roles.after[id] !== change) {
            // The member `id` will have their role updated
            result.roles.after[id] = change;
            result.ids.updated.push(id);
        } else {
            // This situation is where either a non-member user was removed, or a member was specified with a role
            // they already had, remove it from the applicable changes result
            delete result.changes[id];
        }
    });

    return result;
};

/**
 * Given a resource, get the resource id it uses to index authz membership information
 *
 * @param  {Resource}   resource    The resource whose authz id to get
 * @return {String}                 The resource id to use for authz membership look-ups
 */
var getAuthzId = module.exports.getAuthzId = function(resource) {
    return (resource.groupId || resource.id);
};
