// oae-transfer/lib/test/util.js
/*!
 * Copyright 2017 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var assert = require('assert');
var shortid = require('shortid');
var fs = require('fs');
var util = require('util');

var ContentTestUtil = require('oae-content/lib/test/util');
var DiscussionsTestUtil = require('oae-discussions/lib/test/util');
var MeetingAPI = require('oae-jitsi/lib/api.meetings');
var RestAPI = require('oae-rest');

var Transfer = require('oae-transfer/lib/model').Transfer;
var TransferDAO = require('oae-transfer/lib/internal/dao');


/* ================ CREATE TRANSFER ================ */

/**
 * Create a transfer
 *
 * @param  {RestContext}    restContext             The REST context to use for making requests
 * @param  {String}         originalEmail             The email of the user who wants to create a transfer
 * @param  {String}         targetEmail             The targetEmail of the transfer to create
 * @param  {String}         originalUserId            The id of the user who wants to create a transfer
 * @param  {Function}       callback                Standard callback function
 * @param  {Folder}         callback.transfer       Return transfer
 * @throws {AssertionError}                         Thrown if an error occurred generating the tranfer
 */
var assertCreateTransferSucceeds = module.exports.assertCreateTransferSucceeds = function(restContext, originalEmail, targetEmail, originalUserId, callback) {
    RestAPI.Transfer.createTransfer(restContext, originalUserId, originalEmail, targetEmail, function(err, createdTransfer) {
        assert.ok(createdTransfer);
        assert.equal(createdTransfer.originalEmail, originalEmail);
        assert.equal(createdTransfer.targetEmail, targetEmail);
        assert.equal(createdTransfer.originalUserId, originalUserId);
        return callback(createdTransfer);
    });
};

/* ================ GET TRANSFER ================ */

/**
 * Get a transfer, ensuring that it fails in a specified way
 *
 * @param  {RestContext}        restContext     The REST context to use when getting the transfer
 * @param  {String}             originalUserId    The id of the user who wants to make the transfer
 * @param  {Number}             httpCode        The expected failure HTTP code of the request
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request did not fail in the expected manner
 */
var assertGetTransferFails = module.exports.assertGetTransferFails = function(restContext, originalUserId, httpCode, callback) {
    RestAPI.Transfer.getTransferById(restContext, originalUserId, function(err, transfer) {
        assert.ok(err);
        assert.strictEqual(err.code, httpCode);
        assert.ok(!transfer);
        return callback(null);
    });
};

/**
 * Get a transfer, ensuring that the request is successful
 *
 * @param  {RestContext}        restContext     The REST context to use when getting the transfer
 * @param  {String}             originalUserId    The id of the user who wants to make the transfer
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request did not fail in the expected manner
 */
var assertGetTransferSucceeds = module.exports.assertGetTransferSucceeds = function(restContext, originalUserId, callback) {
    RestAPI.Transfer.getTransferById(restContext, originalUserId, function(err, transfer) {
        assert.ok(!err);
        assert.ok(transfer);
        assert.strictEqual(transfer.originalUserId, originalUserId);
        return callback(transfer);
    });
};

/* ================ DELETE TRANSFER ================ */

/**
 * Delete a transfer, ensuring that it fails in a specified way
 *
 * @param  {RestContext}        restContext     The REST context to use when deleting the transfer
 * @param  {String}             originalEmail     The originalEmail of the transfer to delete
 * @param  {String}             code            The transfer code of the transfer to delete
 * @param  {String}             originalUserId    The originalUserId of the transfer to delete
 * @param  {Number}             httpCode        The expected failure HTTP code of the request
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request did not fail in the expected manner
 */
var assertDeleteTransferFails = module.exports.assertDeleteTransferFails = function(restContext, originalEmail, code, originalUserId, httpCode, callback) {
    RestAPI.Transfer.deleteTransfer(restContext, originalEmail, code, originalUserId, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, httpCode);
        return callback();
    });
};

/**
 * Delete a transfer, ensuring that the request is successful
 *
 * @param  {RestContext}        restContext     The REST context to use when deleting the transfer
 * @param  {String}             originalEmail     The originalEmail of the transfer to delete
 * @param  {String}             code            The transfer code of the transfer to delete
 * @param  {String}             originalUserId    The originalUserId of the transfer to delete
 * @param  {Number}             httpCode        The expected failure HTTP code of the request
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request did not fail in the expected manner
 */
var assertDeleteTransferSucceeds = module.exports.assertDeleteTransferSucceeds = function(restContext, originalEmail, code, originalUserId, callback) {
    RestAPI.Transfer.deleteTransfer(restContext, originalEmail, code, originalUserId, function(err) {
        assert.ok(!err);
        return callback();
    });
};

/* ================ MAKE TRANSFER ================ */

/**
 * Execute transfer, ensuring that the request is successful
 *
 * @param  {RestContext}        restContext     The REST context to use when deleting the transfer
 * @param  {String}             originalEmail     The originalEmail of the transfer 
 * @param  {String}             code            The code write by the user
 * @param  {String}             targetEmail     The targetEmail of the transfer
 * @param  {String}             targetUserId    The id of the user target
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var assertMakeTransferSucceeds = module.exports.assertMakeTransferSucceeds = function(restContext, originalEmail, code, targetEmail, targetUserId, callback) {
    RestAPI.Transfer.makeTransfer(restContext, originalEmail, code, targetEmail, targetUserId, function(err, managers) {
        assert.ok(!err);
        return callback(err, managers);
    });
};

/**
 * Execute transfer, ensuring that it fails in a specified way
 *
 * @param  {RestContext}        restContext     The REST context to use when deleting the transfer
 * @param  {String}             originalEmail     The originalEmail of the transfer 
 * @param  {String}             code            The code write by the user
 * @param  {String}             targetEmail     The targetEmail of the transfer
 * @param  {String}             targetUserId    The id of the user target
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var assertMakeTransferFails = module.exports.assertMakeTransferFails = function(restContext, originalEmail, code, targetEmail, targetUserId, httpCode, callback) {
    RestAPI.Transfer.makeTransfer(restContext, originalEmail, code, targetEmail, targetUserId, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, httpCode);
        return callback();
    });
};

/* ================ GET RECORDED TRANSFER ================ */

/**
 * Get recorded transfer, ensuring that the request is successful
 *
 * @param  {String}             originalUserId  The id of the original user
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var assertGetRecordedTransferSucceeds = module.exports.assertGetRecordedTransferSucceeds = function(originalUserId, callback) {
    TransferDAO.getRecordedTransfer(originalUserId, function(err, recordedTransfer) {
        assert.ok(!err);
        return callback(err, recordedTransfer);
    });
};

/* ================ UTIL ================ */

/**
 * Generate discussions
 *
 * @param  {ctx}                context         The REST context to use when ti create the discussions 
 * @param  {String}             privacy         The privacy of the discussions
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateDiscussions = module.exports.generateDiscussions = function(ctx, privacy, numToCreate, callback, _created) {
    _created = _created || [];
    if (_created.length === numToCreate) {
        return callback(null, _created);
    }

    RestAPI.Discussions.createDiscussion(ctx, 'name', 'description', privacy, null, null, function(err, discussion) {
        assert.ok(!err);
        _created.push(discussion);
        return generateDiscussions(ctx, privacy, numToCreate, callback, _created);
    });
};

/**
 * Generate links
 *
 * @param  {ctx}                ctx             The REST context to use when create the links 
 * @param  {String}             privacy         The privacy of the links
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateLinks = module.exports.generateLinks = function(ctx, privacy, numToCreate, callback, _created) {
    _created = _created || [];
    if (_created.length === numToCreate) {
        return callback(null, _created);
    }

    RestAPI.Content.createLink(ctx, 'name', 'description', privacy, 'google.com', null, null, null, function(err, link) {
        assert.ok(!err);
        _created.push(link);
        return generateLinks(ctx, privacy, numToCreate, callback, _created);
    });
};

/**
 * Generate files
 *
 * @param  {ctx}                ctx             The REST context to use when create the files 
 * @param  {String}             privacy         The privacy of the files
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateFiles = module.exports.generateFiles = function(ctx, privacy, numToCreate, callback, _created) {
    _created = _created || [];
    if (_created.length === numToCreate) {
        return callback(null, _created);
    }

    RestAPI.Content.createFile(ctx, 'name', 'description', privacy, getFileStream, null, null, null, function(err, file) {
        assert.ok(!err);
        _created.push(file);
        return generateFiles(ctx, privacy, numToCreate, callback, _created);
    });
};

/**
 * Generate a file
 *
 * @param  {ctx}                ctx             The REST context to use when create the file
 * @param  {String}             privacy         The privacy of the file
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateFileOnFolder = module.exports.generateFileOnFolder = function(ctx, privacy, idFolder, idGroup, callback) {
    RestAPI.Content.createFile(ctx, 'name', 'description', privacy, getFileStream, [idGroup], null, [idFolder], function(err, file) {
        assert.ok(!err);
        callback(null, file);
    });
};

/**
 * Generate collabdocs
 *
 * @param  {ctx}                ctx             The REST context to use when create the collabdoc
 * @param  {String}             privacy         The privacy of the collabdoc
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
 var generateCollabdocs = module.exports.generateCollabdocs = function(ctx, privacy, numToCreate, callback, _created) {
    _created = _created || [];
    if (_created.length === numToCreate) {
        return callback(null, _created);
    }

    RestAPI.Content.createCollabDoc(ctx, 'name', 'description', privacy, [], [], [], [], function(err, collabdoc) {
        assert.ok(!err);
        _created.push(collabdoc);
        return generateCollabdocs(ctx, privacy, numToCreate, callback, _created);
    });
};

/**
 * Generate meetings
 *
 * @param  {ctx}                ctx             The REST context to use when create the meetings 
 * @param  {String}             privacy         The privacy of the meetings
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateMeetings = module.exports.generateMeetings = function(restCtx, manager, privacy, numToCreate, callback, _created) {
    _created = _created || [];
    if (_created.length === numToCreate) {
        return callback(null, _created);
    }

    RestAPI.MeetingsJitsi.createMeeting(restCtx, 'name', 'description', false, false, privacy, null, null, function (err, meeting) {
        assert.ok(!err);
        _created.push(meeting);
        return generateMeetings(restCtx, manager, privacy, numToCreate, callback, _created);
    });
};

/**
 * Generate right for a content
 *
 * @param  {Object}             owner           The ower of the content
 * @param  {String}             contributor     The user
 * @param  {String}             right           The right to attribute to the user
 * @param  {String}             content         The content to update
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateRightContent = module.exports.generateRightContent = function(owner, contributor, right, content, callback) {
    var memberUpdates = {};
    memberUpdates[contributor.user.id] = right;
    ContentTestUtil.assertUpdateContentMembersSucceeds(owner.restContext, owner.restContext, content.id, memberUpdates, function(err) {
        assert.ok(!err);
        callback(null, content);
    });
};

/**
 * Generate right for a discussion
 *
 * @param  {Object}             owner           The ower of the discussion
 * @param  {String}             contributor     The user
 * @param  {String}             right           The right to attribute to the user
 * @param  {String}             discussion      The discussion to update
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var generateRightDiscussion = module.exports.generateRightDiscussion = function(owner, contributor, right, discussion, callback) {
    var memberUpdates = {};
    memberUpdates[contributor.user.id] = right;
    DiscussionsTestUtil.assertUpdateDiscussionMembersSucceeds(owner.restContext, owner.restContext, discussion.id, memberUpdates, function(err) {
        assert.ok(!err);
        callback(null, discussion);
    });
};

/**
 * Get the path
 *
 * @param  {Function}           callback        Standard callback function
 * @throws {AssertionError}                     Thrown if the request failed
 */
var getFileStream = module.exports.getFileStream = function() {
    var file = __dirname + '/data/profilepic.jpg';
    return fs.createReadStream(file);
};
