// oae-activity/lib/test/util.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var assert = require('assert');
var cheerio = require('cheerio');
var clone = require('clone');
var ShortId = require('shortid');
var sjsc = require('sockjs-client-ws');

var EmitterAPI = require('oae-emitter');
var MqTestsUtil = require('oae-util/lib/test/mq-util');
var OaeUtil = require('oae-util/lib/util');
var RestAPI = require('oae-rest');

var ActivityAggregator = require('oae-activity/lib/internal/aggregator');
var ActivityAPI = require('oae-activity');
var ActivityConstants = require('oae-activity/lib/constants').ActivityConstants;

/**
 * Refresh the activity module's configuration, keeping in mind default test configuration. This is
 * important to ensure test defaults, particularly the fact that aggregation collection should
 * always be disabled during tests to avoid intermittent test failures.
 *
 * @param  {Object}     [config]        The configuration with which to reset configuration
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occurred, if any
 */
var refreshConfiguration = module.exports.refreshConfiguration = function(config, callback) {
    config = _.extend({
        'collectionPollingFrequency': -1,
        'numberOfProcessingBuckets': 1
    }, config);
    ActivityAPI.refreshConfiguration(config, callback);
};

/**
 * Get the `activity` activity stream by its resource id, but delay and perform
 * an activity collection first to ensure the stream is updated.
 *
 * @param  {RestContext}    restCtx             The context of the current request
 * @param  {String}         [resourceId]        The id of the resource's activity stream to get. If `null`, will get the stream of the current user
 * @param  {Object}         [opts]              @see RestAPI.Activity#getActivityStream for available options
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error that occurred, if any
 *
 * @see RestAPI.Activity#getActivityStream for more information.
 */
var collectAndGetActivityStream = module.exports.collectAndGetActivityStream = function(restCtx, resourceId, opts, callback) {
    MqTestsUtil.whenTasksEmpty(ActivityConstants.mq.TASK_ACTIVITY, function() {
        ActivityAggregator.collectAllBuckets(function() {
            if (resourceId) {
                RestAPI.Activity.getActivityStream(restCtx, resourceId, opts, callback);
            } else {
                RestAPI.Activity.getCurrentUserActivityStream(restCtx, opts, callback);
            }
        });
    });
};

/**
 * Get the notification stream, but delay and perform an activity collection first to ensure the stream is
 * updated.
 *
 * @param  {RestContext}    restCtx             The context of the current request
 * @param  {Object}         [opts]              @see RestAPI.Activity#getNotificationStream for available options
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error that occurred, if any
 *
 * @see RestAPI.Activity#getNotificationStream for more information.
 */
var collectAndGetNotificationStream = module.exports.collectAndGetNotificationStream = function(restCtx, opts, callback) {
    MqTestsUtil.whenTasksEmpty(ActivityConstants.mq.TASK_ACTIVITY, function() {
        ActivityAggregator.collectAllBuckets(function() {
            RestAPI.Activity.getNotificationStream(restCtx, opts, callback);
        });
    });
};

/**
 * Assert that retrieving an activity stream fails in an expected manner
 *
 * @param  {RestContext}    restCtx             The context of the current request
 * @param  {String}         resourceId          The id of the resource's activity stream to get
 * @param  {Object}         [opts]              @see RestAPI.Activity#getActivityStream for available options
 * @param  {Number}         code                The expected http status code
 * @param  {Function}       callback            Standard callback function
 * @throws {AssertionError}                     Thrown if the request did not fail in the expected manner
 */
var assertGetActivityStreamFails = module.exports.assertGetActivityStreamFails = function(restCtx, resourceId, opts, code, callback) {
    RestAPI.Activity.getActivityStream(restCtx, resourceId, opts, function(err, activityStream) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!activityStream);
        return callback();
    });
};

/**
 * Marks all the notifications for a user as read and waits till the activity aggregation for that user's notification
 * stream has been reset.
 *
 * @param  {RestContext}    restContext     The context of the current request
 * @param  {Function}       callback        Standard callback function
 * @throws {Error}                          An assertion error is thrown if the notifications could not be marked as read or the response was invalid
 */
var markNotificationsAsRead = module.exports.markNotificationsAsRead = function(restContext, callback) {
    var result = null;

    ActivityAPI.once(ActivityConstants.events.RESET_AGGREGATION, function(activityStreams) {
        return callback(result);
    });

    RestAPI.Activity.markNotificationsRead(restContext, function(err, _result) {
        assert.ok(!err);

        // Assert we're getting back a number
        result = _result;
        var lastReadTime = _result.lastReadTime;
        assert.strictEqual(lastReadTime, OaeUtil.getNumberParam(lastReadTime));
    });
};

/**
 * Assert an activity stream contains an activity specified by the given `activityType`
 * and verb. Also assert that the activity contains the expected entities
 *
 * @param  {RestContext}    restContext         The rest context to use to retrieve the activity stream
 * @param  {String}         activityStreamId    The id of the activity stream to check
 * @param  {String}         activityType        The expected type of the activity that should be in the activity stream
 * @param  {String}         verb                The expected verb of the activity
 * @param  {String}         actorId             The expected actor id, or an array of expected actor ids if the entity is an oae:collection aggregate
 * @param  {String}         [objectId]          The expected object id, or an array of expected object ids if the entity is an oae:collection aggregate
 * @param  {String}         [targetId]          The expected target id, or an array of expected target ids if the entity is an oae:collection aggregate
 * @param  {Function}       callback            Standard callback function
 * @throws {Error}                              An assertion error gets thrown if the activity was not found or some of its properties are not what was expected
 */
var assertFeedContainsActivity = module.exports.assertFeedContainsActivity = function(restContext, activityStreamId, activityType, verb, actorId, objectId, targetId, callback) {
    collectAndGetActivityStream(restContext, activityStreamId, null, function(err, response) {
        assert.ok(!err);
        var activity = _.findWhere(response.items, {'oae:activityType': activityType});
        assert.ok(activity);
        assertActivity(activity, activityType, verb, actorId, objectId, targetId);
        return callback();
    });
};

/**
 * Assert an activity stream does not contain an activity of a specified activity type
 *
 * @param  {RestContext}    restContext         The rest context to use to retrieve the activity stream
 * @param  {String}         activityType        The type of the activity that should not be in the activity stream
 * @param  {Function}       callback            Standard callback function
 * @throws {Error}                              An assertion error gets thrown if the activity was found
 */
var assertFeedDoesNotContainActivity = module.exports.assertFeedDoesNotContainActivity = function(restContext, activityStreamId, activityType, callback) {
    collectAndGetActivityStream(restContext, activityStreamId, null, function(err, response) {
        assert.ok(!err);
        var activity = _.findWhere(response.items, {'oae:activityType': activityType});
        assert.ok(!activity);
        return callback();
    });
};

/**
 * Assert a notification stream contains an activity specified by the given `activityType`
 * and verb. Also assert that the activity contains the expected entities
 *
 * @param  {RestContext}    restContext         The rest context to use to retrieve the notification stream
 * @param  {String}         activityStreamId    The id of the activity stream to check
 * @param  {String}         activityType        The expected type of the activity that should be in the notification stream
 * @param  {String}         verb                The expected verb of the activity
 * @param  {String}         actorId             The expected actor id, or an array of expected actor ids if the entity is an oae:collection aggregate
 * @param  {String}         [objectId]          The expected object id, or an array of expected object ids if the entity is an oae:collection aggregate
 * @param  {String}         [targetId]          The expected target id, or an array of expected target ids if the entity is an oae:collection aggregate
 * @param  {Function}       callback            Standard callback function
 * @throws {Error}                              An assertion error gets thrown if the activity was not found or some of its properties are not what was expected
 */
var _assertNotificationStreamContainsActivity = module.exports.assertNotificationStreamContainsActivity = function(restContext, activityType, verb, actorId, objectId, targetId, callback) {
    collectAndGetNotificationStream(restContext, null, function(err, notificationStream) {
        assert.ok(!err);
        var activity = _.findWhere(notificationStream.items, {'oae:activityType': activityType});
        assert.ok(activity);
        assertActivity(activity, activityType, verb, actorId, objectId, targetId);
        return callback();
    });
};

/**
 * Assert a notification stream does not contain an activity of a specified activity type
 *
 * @param  {RestContext}    restContext         The rest context to use to retrieve the notification stream
 * @param  {String}         activityType        The type of the activity that should not be in the notification stream
 * @param  {Function}       callback            Standard callback function
 * @throws {Error}                              An assertion error gets thrown if the activity was found
 */
var _assertNotificationStreamDoesNotContainActivity = module.exports.assertNotificationStreamDoesNotContainActivity = function(restContext, activityType, callback) {
    collectAndGetNotificationStream(restContext, null, function(err, notificationStream) {
        assert.ok(!err);
        var activity = _.findWhere(notificationStream.items, {'oae:activityType': activityType});
        assert.ok(!activity);
        return callback();
    });
};

/**
 * Perform basic assertions of an activity and the entities that it contains
 *
 * @param  {Activity}           activity            The activity whose contents to verify
 * @param  {String}             activityType        The activity type that should be associated to the activity
 * @param  {String}             verb                The verb that should be associated to the activity
 * @param  {String|String[]}    actorEntityId       The id of the entity that should be the actor, or an array of expected actor entity ids if the entity is expected to be an oae:collection aggregate
 * @param  {String|String[]}    [objectEntityId]    The id of the entity that should be the object, or an array of expected object entity ids if the entity is expected to be an oae:collection aggregate. If not specified, an assertion will be performed that the object does not exist
 * @param  {String|String[]}    [targetEntityId]    The id of the entity taht should be the target, or an array of expected target entity ids if the entity is expected to be an oae:collection aggregate. If not specified, an assertion will be performed that the target does not exist
 */
var assertActivity = module.exports.assertActivity = function(activity, activityType, verb, actorEntityId, objectEntityId, targetEntityId) {
    assert.ok(activity);
    assert.equal(activity[ActivityConstants.properties.OAE_ACTIVITY_TYPE], activityType);
    assert.equal(activity.verb, verb);
    _assertActivityEntity(activity.actor, actorEntityId);
    _assertActivityEntity(activity.object, objectEntityId);
    _assertActivityEntity(activity.target, targetEntityId);
};

/**
 * Verify the id(s) of the provided activity entity against the provided entity id
 *
 * @param  {ActivityEntity}     [activityEntity]    The activity entity to validate
 * @param  {String|String[]}    [entityId]          The expected id of the entity, or an array of expected entity ids if the entity is expected to be an oae:collection aggregate. If not specified, an assertion will be performed that the entity does not exist (i.e., it's unspecified)
 * @api private
 */
var _assertActivityEntity = function(activityEntity, entityId) {
    if (!entityId) {
        assert.ok(!activityEntity);
    } else if (_.isString(entityId)) {
        assert.ok(activityEntity);
        assert.equal(activityEntity[ActivityConstants.properties.OAE_ID], entityId);
    } else if (_.isArray(entityId)) {
        // The entity id is an array, so we need to make sure every id in this list is on the activity entity collection
        var entityIds = entityId;
        assert.ok(activityEntity);

        // Ensure it is a collection with the same amount of ids as the given list of entity ids
        assert.equal(activityEntity.objectType, 'collection');
        assert.ok(activityEntity[ActivityConstants.properties.OAE_COLLECTION]);
        assert.strictEqual(activityEntity[ActivityConstants.properties.OAE_COLLECTION].length, entityIds.length);

        // Ensure every id in the list is in the entity collection
        _.each(activityEntity[ActivityConstants.properties.OAE_COLLECTION], function(activityEntity) {
            assert.ok(_.contains(entityIds, activityEntity[ActivityConstants.properties.OAE_ID]));
        });
    }
};

/**
 * Parse the given activity email HTML and extract the activity data from the template
 *
 * @param  {String}     html                                The activity email html
 * @return {Object[]}   activities                          The activity info that were parsed
 * @return {String}     activities[i].id                    The id of the activity
 * @return {Object}     activities[i].summary               The summary info of the activity
 * @return {String}     activities[i].summary.html          The full html of the summary
 * @return {String}     activities[i].summary.text          The text version of the summary
 * @return {Object[]}   activities[i].summary.links         The link info for each link that was in the summary
 * @return {String}     activities[i].summary.link[i].href  The href attribute for the link
 * @return {String}     activities[i].summary.link[i].html  The inner html of the link
 * @return {String}     activities[i].summary.link[i].text  The inner text of the link
 */
var parseActivityHtml = module.exports.parseActivityHtml = function(html) {
    var $ = cheerio.load(html);

    var activities = [];
    $('tr.activity-row').each(function() {
        var $activity = $(this);
        var $summary = $activity.find('.activity-summary-container > div:first-child');

        // Get the id and the summary of this activity
        var activityId = $activity.attr('data-activity-id');
        var summary = {
            'html': $summary.html(),
            'text': $summary.text(),
            'links': []
        };

        // Get all the links located in the summary and extract their information
        $summary.find('a').each(function() {
            var $a = $(this);
            summary.links.push({
                'href': $a.attr('href'),
                'html': $a.html(),
                'text': $a.text()
            });
        });

        // Append this activity data to the list
        activities.push({
            'id': activityId,
            'summary': summary,
        });
    });

    return activities;
};

/**
 * Get a WebSocket that is connected to the `localhost` tenant.
 *
 * This client emits the following events:
 *  * `message(msg)`    -   A new message arrived on the socket and it's not a response to an earlier sent message
 *  * `close()`         -   The underlying socket has been closed
 *
 * @param  {Function}   callback                Standard callback function
 * @param  {Client}     callback.client         The connected client
 */
var getPushClient = module.exports.getPushClient = function(callback) {
    // The client that we'll return to the caller
    var client = new EmitterAPI.EventEmitter();

    // A dictionary that maps message IDs to callbacks
    var messageCallbacks = {};

    // Set up a websocket connection to the localhost tenant
    var socket = sjsc.create('http://localhost:2001/api/push');

    socket.on('error', function (e) {
        assert.fail(e, null, 'Did not expect an error on the websocket');
    });

    socket.on('connection', function() {
        // The socket has been connected and is ready to transmit messages
        callback(client);
    });

    socket.on('data', function (msg) {
        // We ignore 'open', 'heartbeat' or 'close' messages
        if (msg === 'o' || msg === 'h' || msg === 'c') {
            return;
        }

        // Everything we receive at this point, should be sent too us by OAE. This should be proper JSON
        try {
            msg = JSON.parse(msg);
        } catch (err) {
            assert.fail(err, null, 'Did not expect malformed JSON as a response from the server');
        }

        // If we registered a callback for this message, we execute it
        if (msg.replyTo && messageCallbacks[msg.replyTo]) {
            messageCallbacks[msg.replyTo](msg.error, msg);

            // Since we should only get one message from the server with that ID, we change the callback
            // so that if we get another message with that ID, we'll throw an error
            messageCallbacks[msg.replyTo] = function(error, msg) {
                assert.fail(msg, null, 'A message ID can only be responded to once');
            };

        // Otherwise we simply emit the message
        } else {
            client.emit('message', msg);
        }
    });

    socket.on('close', function() {
        client.emit('close');
    });

    /**
     * Returns the raw socket. This allows you to send custom messages
     *
     * @return {Socket}     The raw socket
     */
    client.getRawSocket = function() {
        return socket;
    };

    /**
     * Sends a message over the wire
     *
     * @param  {String}     name            A name for this frame. ex: `authentication` or `register-for-feed`
     * @param  {Object}     payload         A payload object
     * @param  {Function}   [callback]      A function that should be executed when a response is received
     */
    client.sendMessage = function(name, payload, callback) {
        callback = callback || function() {};

        // Generate a random ID for this message so we can identify the response later on
        var id = 'id' + ShortId.generate();
        messageCallbacks[id] = callback;

        // Send the message over the wire
        var msg = JSON.stringify({'id': id, 'name': name, 'payload': payload});
        socket.write(msg);
    };

    /**
     * Utility to help you authenticate a user on the websocket
     *
     * @param  {String}     userId       The ID of the user to authenticate
     * @param  {String}     tenantAlias  The alias of the tenant of which this user originates
     * @param  {Object}     signature    A signature object
     * @param  {Function}   callback     Standard callback function
     */
    client.authenticate = function(userId, tenantAlias, signature, callback) {
        var payload = {
            'userId': userId,
            'tenantAlias': tenantAlias,
            'signature': signature
        };
        client.sendMessage('authentication', payload, callback);
    };

    /**
     * Utility to register this WS for a stream
     *
     * @param  {String}    resourceId           The ID of the resource, ex: a userId
     * @param  {String}    streamType           The type of the activity stream, ex: `activity`
     * @param  {Object}    token                An object that can be used to authorize the subscription request
     * @param  {String}    format               The format in which the activity entities should be returned
     * @param  {Function}  callback             Standard callback function
     */
    client.subscribe = function(resourceId, streamType, token, format, callback) {
        var payload = {
            'stream': {
                'resourceId': resourceId,
                'streamType': streamType
            },
            'token': token,
            'format': format
        };
        client.sendMessage('subscribe', payload, callback);
    };

    /**
     * Closes the underlying socket.
     * A `close` event will also be emitted once the socket is closed.
     *
     * @param  {Function}   callback    Standard callback function
     */
    client.close = function(callback) {
        callback = callback || function() {};
        socket.on('close', callback);
        socket.close();
    };
};

/**
 * Get a WebSocket that is authenticated and has registered for 0, 1 or multiple feeds.
 *
 * @param  {Object}     data                                A data object that contains the authentication credentials and the feeds that should be registered
 * @param  {Object}     data.authentication                 The object containing the authentication data
 * @param  {String}     data.authentication.userId          The id of the user to authenticate
 * @param  {String}     data.authentication.tenantAlias     The alias of the tenant on which the signature was retrieved
 * @param  {Object}     data.authentication.signature       The signature which can be used to authenticate the client
 * @param  {Object[]}   [data.streams]                      The streams on which we should subscribe
 * @param  {String}     [data.streams[i].resourceId]        The id of the resource for which to retrieve an activity stream
 * @param  {String}     [data.streams[i].streamType]        The type of the stream that should be retrieved, ex: `notification` or `activity`
 * @param  {Object}     [data.streams[i].token]             A token that can be used to authorize the client on the stream
 * @param  {Object}     [data.streams[i].format]            The format in which the activity entities should be returned
 * @param  {Function}   callback                            Standard callback function
 * @param  {Client}     callback.client                     A connected WS client that is authenticated and registered for all the desired feeds
 * @throws {Error}                                          If the client could not be authenticated or registered for a feed
 */
var getFullySetupPushClient = module.exports.getFullySetupPushClient = function(data, callback) {
    getPushClient(function(client) {
        client.authenticate(data.authentication.userId, data.authentication.tenantAlias, data.authentication.signature, function(err) {
            assert.ok(!err, 'Failed to authenticate');

            if (_.isEmpty(data.streams)) {
                return callback(client);
            }

            var allRegisteredCallback = _.after(data.streams.length, callback);
            _.each(data.streams, function(stream) {
                client.subscribe(stream.resourceId, stream.streamType, stream.token, stream.format, function(err) {
                    assert.ok(!err, 'Failed to register for feed');
                    allRegisteredCallback(client);
                });
            });
        });
    });
};

/**
 * Wait for a message to come through the activity that matches the given criteria
 *
 * @param  {PushClient}     client              The WS client on which to listen for messages
 * @param  {String}         activityType        The type of activity to wait for
 * @param  {String}         verb                The verb of the activity to wait for
 * @param  {String}         actorId             The id of the actor who performed the activity to wait for
 * @param  {String}         [objectId]          The id of the object of the activity to wait for. If unspecified, indicates that you don't care what the object id is
 * @param  {String}         [targetId]          The id of the target of the activity to wait for. If unspecified, indicates that you don't care what the target id is
 * @param  {Function}       callback            Invoked when a message with the specified activity has been received
 * @param  {Activity}       callback.activity   The activity object that was contained in the message
 */
var waitForPushActivity = module.exports.waitForPushActivity = function(client, activityType, verb, actorId, objectId, targetId, callback) {

    /*!
     * Listener function to wait for messages, perform the activity filter, and unbind itself from
     * the client when the activity has been found. When found, the callback will be invoked with
     * the activity
     */
    var _onMessage = function(message) {
        var targetActivity = _.find(message.activities, function(activity) {
            if (activity[ActivityConstants.properties.OAE_ACTIVITY_TYPE] !== activityType) {
                return false;
            } else if (activity.verb !== verb) {
                return false;
            } else if (activity.actor[ActivityConstants.properties.OAE_ID] !== actorId) {
                return false;
            } else if (objectId && activity.object[ActivityConstants.properties.OAE_ID] !== objectId) {
                return false;
            } else if (targetId) {
                if (!activity.target) {
                    return false;
                } else if (activity.target[ActivityConstants.properties.OAE_ID] !== targetId) {
                    return false;
                }
            }

            return true;
        });

        if (targetActivity) {
            client.removeListener('message', _onMessage);
            return callback(targetActivity);
        }
    };

    // Begin waiting for messages
    return client.on('message', _onMessage);
};
