// oae-principals/lib/definitiveDeletion.js
/*!
 * Copyright 2017 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the 'License'); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an 'AS IS'
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var async = require('async');
var fs = require('fs');
var shortId = require('shortId');
var util = require('util');

var AuthenticationConstants = require('oae-authentication/lib/constants').AuthenticationConstants;
var AuthzAPI = require('oae-authz');
var AuthzConstants = require('oae-authz/lib/constants').AuthzConstants;
var AuthzDelete = require('oae-authz/lib/delete');
var Cassandra = require('oae-util/lib/cassandra');
var ConfigAPI = require('oae-config');
var ContentAPI = require('oae-content');
var ContentConfig = require('oae-config').config('oae-content');
var ContentConstants = require('oae-content/lib/constants').ContentConstants;
var ContentUtil = require('oae-content/lib/internal/util');
var Context = require('oae-context').Context;
var DiscussionAPI = require('oae-discussions/lib/api.discussions');
var DiscussionsDAO = require('oae-discussions/lib/internal/dao');
var EmailAPI = require('oae-email');
var FolderAPI = require('oae-folders');
var FolderLibraryDAO = require('oae-folders/lib/internal/foldersLibrary');
var FoldersConstants = require('oae-folders/lib/constants').FoldersConstants;
var FoldersDAO = require('oae-folders/lib/internal/dao');
var FollowingAPI = require('oae-following');
var FollowingConstants = require('oae-following/lib/constants').FollowingConstants;
var FollowingDAO = require('oae-following/lib/internal/dao');
var GroupAPI = require('./api.group');
var LoginId = require('oae-authentication/lib/model').LoginId;
var MeetingsAPI = require('oae-jitsi');
var MeetingsConstants = require('oae-jitsi/lib/constants').MeetingsConstants;
var MembersLibraryDAO = require('oae-content/lib/internal/membersLibrary');
var OaeUtil = require('oae-util/lib/util');
var PrincipalsUtil = require('oae-principals/lib/util');
var ResourceActions = require('oae-resource/lib/actions');
var TenantsAPI = require('oae-tenants');
var TenantsUtil = require('oae-tenants/lib/util');

var PrincipalsAPI = require('./api.user');
var PrincipalsConfig = ConfigAPI.config('oae-principals');
var PrincipalsConstants = require('./constants').PrincipalsConstants;
var PrincipalsDAO = require('./internal/dao');
var PrincipalsEmitter = require('./internal/emitter');

var log = require('oae-logger').logger('oae-principals');

const FOLDER = 'folder';
const CONTENT = 'content';
const DISCUSSION = 'discussion';
const MEETING = 'meeting';
const GROUP = 'group';

const FOLDER_PREFIX = 'f';
const CONTENT_PREFIX = 'c';
const DISCUSSION_PREFIX = 'd';
const MEETING_PREFIX = 'm';
const GROUP_PREFIX = 'g';

/**
 * Get or create user archive
 *
 * @param  {Context}    ctx                     Standard context object containing the current user and the current tenant
 * @param  {Object}     user                    User to be archived
 * @param  {Function}   callback                Standard callback function
 * @param  {Object}     callback.err            An error that occured, if any
 * @param  {Object}     callback.userArchive    Return user archive
 */
var getOrCreateArchiveUser = module.exports.getOrCreateArchiveUser = function(ctx, user, callback) {
    
    // Get and/or create archiveUser
    PrincipalsDAO.getArchivedUser(user.tenant.alias, function(err, userArchive) {
        if (err) {
            return callback(err);
        } else if (userArchive) {
            return callback(null, userArchive);
        } else {
            // Get email
            var emailDomains = user.tenant.emailDomains;

            // Persist the user object
            var userOpts = {
                'tenant': {'alias': user.tenant.alias},
                'visibility': 'public',
                'emailPreference': 'never',
                'locale':  ctx.locale(),
                'acceptedTC': null, 
                'isUserArchive' : 'true'
            };

            var displayName = user.tenant.alias + ' archive';

            // Create a record in the principals table
            PrincipalsAPI.createUser(ctx, user.tenant.alias, displayName, userOpts, function(err, createdUser) {
                if (err) {
                    return callback(err);
                }

                // Create a user archive in the table archiveByTenant
                PrincipalsDAO.createArchivedUser(user.tenant.alias, createdUser.id, function(err) {
                    if (err) {
                        return callback(err);
                    } 

                    // Get and return the userArchive
                    PrincipalsDAO.getArchivedUser(user.tenant.alias, function(err, userArchive) {
                        if (err) {
                            return callback(err);
                        }
                        return callback(null, userArchive);
                    });
                });
            });
        }
    });
};    

/**
 * Delete rights on a user + update roles on editors && on user archive
 *
 * @param  {Context}    ctx                     Standard context object containing the current user and the current tenant
 * @param  {String}     user                    User that will be deleted
 * @param  {String}     archiveUsers            Users Archive
 * @param  {Function}   callback                Standard callback function
 * @param  {Object}     callback.err            An error that occured, if any
 * @param  {Object}     callback.listEmail      Array of users to email
 */
var transferDataToArchive = module.exports.transferDataToArchive = function (ctx, user, archiveUsers, callback) {
    var callback = callback || function() {};
    var listEmail = [];
    var listOfIdElementToBeTransferred = [];

    var types = [CONTENT, FOLDER, DISCUSSION, MEETING, GROUP];
    var listElementByMember = [];

    async.eachSeries(types, function iterator(type, callback) {

        // Delete/update roles
        _transferRight(ctx, user, archiveUsers, listEmail, listOfIdElementToBeTransferred, type, function(err, listEmail, listOfIdElementToBeTransferred) {
            if (err) {
                return callback(err);
            }
            return callback();
        });

    }, function done() {

        // Add element to archive
        _addToArchive(archiveUsers.archiveId, user.id, listOfIdElementToBeTransferred, function(err) {
            if (err) {
                return callback(err);
            }

            // Send all emails
            _sendEmail(ctx, listEmail, archiveUsers, user, function(err, listEmail) {
                if (err) {
                    return callback(err);
                }

                return callback(null, listEmail);
            });
        });
    });          
};

/**
 * Create a member list with corresponding resources 
 *
 * values of deleted :
 *          contentWillBeDeleted : list of resources - corresponding to a user - which they will be delete into X months
 *          userJustLeaving : list of resources - corresponding to a user - where the user deleted is leaving these resources
 * 
 * @param  {Array}      list            List of resources corresponding to a user
 * @param  {Object}     resource        The resource to add to the list
 * @param  {Array}      memberList      The list of memeber to add to the list
 * @param  {Array}      action          The action can be 'true' if the content will be deleted into X month or 'false' if 
 *                                          the deleted user is just leaving the resource.
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occurred, if any
 */
var _addMemberToList = function(list, resource, memberList, action, callback) {
    
    if (_.isEmpty(memberList)) {
        return callback();
    }

    var list = list || [];

    async.eachSeries(memberList, function(member, innerCallback) {

        _parseMember(list, resource, member, action, function(err, newList) {
            list = newList;
            return innerCallback();
        });

    }, function() {
        return callback(null, list);
    });
};

var _parseMember = function(list, resource, user, action, callback) {

    if (_.isEmpty(list)) {
        list.push({'contentWillBeDeleted': [], 'userJustLeaving': [], 'profile': user.profile});
    }

    async.eachSeries(list, function(member, innerCallback) {

        if (user.profile.id == member.profile.id) {
            if (action === true) {
                member.contentWillBeDeleted.push(resource);
            } else {
                member.userJustLeaving.push(resource);
            }
            return innerCallback(true);
        }
        return innerCallback();

    }, function(isUserFound) {
        if (!isUserFound) {
            if (action === true) {
                list.push({'contentWillBeDeleted': [resource], 'userJustLeaving': [], 'profile': user.profile});
            } else {
                list.push({'contentWillBeDeleted': [], 'userJustLeaving': [resource], 'profile': user.profile});
            }
        }
        return callback(null, list);
    });
};

/**
 * Send an email token to a user that have a resource shared with the deleted user
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {Object}     data            User to send mail + resource
 * @param  {Object}     archiveUser     The archive user
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occurred, if any
 */
var _sendEmail = function(ctx, data, archiveUser, userDeleted, callback) {

    var callback = callback || function(err) {
        if (err) {
            log().error({'err': err, 'user': user.id}, 'Unable to send a user a verification email');
        }
    };

    if (_.isEmpty(data)) {
        return callback();
    }

    PrincipalsDAO.getPrincipal(archiveUser.archiveId, function(err, archiveUser) {
        if (err) {
            return callback(err);
        }

        // Grab the configuration field. This will return the number of months 
        var month = PrincipalsConfig.getValue(ctx.tenant().alias, 'user', 'delete');

        if (_.isEmpty(data)) {
            return callback();
        }

        var listEmail = [];

        async.eachSeries(data, function(user, innerCallback) {

            // If there is no data, don't send mail
            if (_.isEmpty(user.contentWillBeDeleted) && _.isEmpty(user.userJustLeaving)) {
                return innerCallback();

            // Send an unique email to each user
            } else {

                // Generate a token if none was specified
                var token = shortId.generate();

                // Store the token
                PrincipalsDAO.storeEmailToken(user.profile.id, user.profile.email, token, function(err) {
                    if (err) {
                        return callback(err);
                    }

                    var resource = _.omit(user,'profile');
                    user = _.omit(user,['contentWillBeDeleted', 'userJustLeaving']);

                    // The EmailAPI expects a user to have a verified email address. As this is not the case
                    // when sending an email token, we send in a patched user object
                    var userToEmail = _.extend({}, user, {'email': user.profile.email});
                    var tenant = TenantsAPI.getTenant(user.profile.tenant.alias);
                    
                    // Send an email to the specified e-mail address
                    var dataEmail = {
                        'tenant': ctx.tenant(),
                        'userDeletedName': userDeleted.displayName,
                        'resource': resource,
                        'user': user,
                        'month': month,
                        'baseUrl': TenantsUtil.getBaseUrl(ctx.tenant()),
                        'skinVariables': require('oae-ui').getTenantSkinVariables(ctx.tenant().alias),
                        'token': token,
                        'archiveEmail': archiveUser.email
                    };

                    // We pass the current date in as the 'hashCode' for this email. We need to be able to send
                    // a copy of the same email for the 'resend email token' functionality. As we don't expect
                    // that this logic will get stuck in a loop this is probably OK
                    EmailAPI.sendEmail('oae-principals', 'notify', userToEmail.profile, dataEmail, {'hash': Date.now()});
                    listEmail.push(user.profile.email);

                    return innerCallback();
                });
            }
        }, function() {
            return callback(null, listEmail);
        });
    });
};

/**
 * Delete rights on element + update roles on editors && on user archive
 *
 * @param  {Context}    ctx                                          Standard context object containing the current user and the current tenant
 * @param  {String}     user                                         The user to delete
 * @param  {String}     archiveUser                                  User Archive
 * @param  {String}     listEmail                                    Array of users to email
 * @param  {String}     listOfIdElementToBeTransferred               List of element to remove from deleted user
 * @param  {String}     type                                         Type can be : 'content', 'folder', 'discussion', 'meeting' or 'group'
 * @param  {Function}   callback                                     Standard callback function
 * @param  {Object}     callback.err                                 An error that occured, if any
 * @param  {Array}      callback.listEmail                           Array of users to email
 * @param  {Array}      callback.listOfIdElementToBeTransferred      List of element to remove from deleted user
 * @api private
 */
var _transferRight = function(ctx, user, archiveUser, listElementByMember, listOfIdElementToBeTransferred, type, callback) {

    // Get all documents 
    _getLibrary(ctx, user.id, type, function(err, elements) {
        if (err) {
            return callback(err);
        }
        if (_.isEmpty(elements)) {
            return callback(null, listElementByMember, listOfIdElementToBeTransferred);
        }

        async.mapSeries(elements, function(element, innerCallback) {

            // Search if other user have right on this document
            _getMembers(ctx, element.id, type, function(err, memberList) {
                if (err) {
                    return callback(err);
                }

                // If he's not a manager, do nothing
                _doesUserHaveRoleManager(user.id, element, type, function(err, isManager) {
                    if (err) {
                        return callback(err);
                    } 

                    // Remove the deleted user from the member list
                    _removeDeletedUserFromArray(memberList, user.id, function(err, newMemberList) {
                        if (err) {
                            return callback(err);
                        }

                        if (isManager) {

                            // If member list is empty
                            if (_.isEmpty(newMemberList)) { 

                                listOfIdElementToBeTransferred.push(element.id);

                                // Make user archive a manager of the resource
                                _updateRoles(ctx, element, archiveUser, type, function(err) {
                                    if (err) {
                                        return callback(err);
                                    }
                                    // Inner callback
                                    _removeFromLibrary(ctx, user.id, element, type, function(err) {
                                        if (err) {
                                            return callback(err);
                                        }
                                        innerCallback()
                                    });
                                });

                            } else {

                                var hasAnotherManager = _.find(newMemberList, function(member) {
                                    return member.role === AuthzConstants.role.MANAGER;
                                });

                                // If there is another manager on the resource, send a notification email
                                if (hasAnotherManager) {
                                    // We will just notify the members that the user will remove his account
                                    _addMemberToList(listElementByMember, element, newMemberList, false, function(err, listElementByMember) {
                                        // Inner callback
                                        _removeFromLibrary(ctx, user.id, element, type, function(err) {
                                            if (err) {
                                                return callback(err);
                                            }
                                            innerCallback()
                                        });
                                    }); 

                                // Has no manager
                                } else {
                                    listOfIdElementToBeTransferred.push(element.id);

                                    // Make user archive a manager of the resource
                                    _updateRoles(ctx, element, archiveUser, type, function(err) {
                                        if (err) {
                                            return callback(err);
                                        }
                                        // We will notify the members that the user will remove his account and make the user archive manager of the resource
                                        _addMemberToList(listElementByMember, element, newMemberList, true, function(err, listElementByMember) {
                                            // Inner callback
                                            _removeFromLibrary(ctx, user.id, element, type, function(err) {
                                                if (err) {
                                                    return callback(err);
                                                }
                                                innerCallback()
                                            });
                                        });
                                    });
                                }
                            }
                        } else {
                            // We will just notify the members that the user will remove his account
                            _addMemberToList(listElementByMember, element, newMemberList, false, function(err, listElementByMember) {
                                // Inner callback
                                _removeFromLibrary(ctx, user.id, element, type, function(err) {
                                    if (err) {
                                        return callback(err);
                                    }
                                    innerCallback()
                                });
                            }); 
                        }
                    });
                });
            });
        }, function() {
            return callback(null, listElementByMember, listOfIdElementToBeTransferred);
        });
    });
};

/**
 * Clear the list by removing the deleted user
 *
 * @param  {Array}      memberList                  Array of user
 * @param  {String}     userId                      User to delete from the array
 * @param  {Function}   callback                    Standard callback function
 * @param  {Object}     callback.err                An error that occured, if any
 * @param  {Object}     callback.newMemberList      Array of user without the deleted user
 * @api private
 */
var _removeDeletedUserFromArray = function(memberList, userId, callback) {
    return callback(null, _.reject(memberList, function(element) {
        return element.profile.id === userId;
    }));
};

/**
 * Add elements to archive
 *
 * @param  {String}     userArchiveId   The user archive of the tenant
 * @param  {String}     principalId     The id user to delete 
 * @param  {String}     elementId       The id element to delete
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 * @api private
 */
var _addToArchive = function(userArchiveId, principalId, elementId, callback) {

    if (!userArchiveId || !principalId) {
        return callback(err);
    }
    if (!elementId) {
        return callback();
    }

    // Return list of ids after removing duplicate elements
    var duplicationRemoved = elementId.filter( (elem, index, self) => {
        return index === self.indexOf(elem);
    });

    // Add the element to data archive
    PrincipalsDAO.addDataToArchive(userArchiveId, principalId, duplicationRemoved, function(err) {
        if (err) {
            return callback(err);
        }
        return callback();  
    });  
};

/**
 * Check if the current user is manager a content
 *
 * @param  {String}     userId              The user id 
 * @param  {Object}     element             The resource 
 * @param  {String}     type                Type can be : 'content', 'folder', 'discussion', 'meeting' or 'group'
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occured, if any
 * @param  {Boolean}    callback.hasRole    True if the user are manager of the content
 * @api private
 */
var _doesUserHaveRoleManager = function(userId, element, type, callback) {

    var elementId = element.id;

    if (type === FOLDER) {
        elementId = element.groupId;
    }

    AuthzAPI.hasRole(userId, elementId, AuthzConstants.role.MANAGER, function(err, hasRole) {
        if (err) {
            return callback(err);
        }
        return callback(null, hasRole);
    });
};

/**
 * Remove elements from library
 *
 * @param  {Context}    ctx                 Standard context object containing the current user and the current tenant
 * @param  {String}     userId              The user id 
 * @param  {String}     element             Element to remove from library
 * @param  {String}     type                Type can be : 'content', 'folder', 'discussion', 'meeting' or 'group'
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occured, if any
 * @api private
 */
var _removeFromLibrary = function(ctx, userId, element, type, callback) {
    if (!element) {
        return callback();
    }

    switch (type) {
        case CONTENT:
            ContentAPI.removeContentFromLibrary(ctx, userId, element.id, function(err) {
                if (err) {
                    return callback(err);
                }
                return callback();
            });
            break;

        case FOLDER:
            FolderAPI.removeFolderFromLibrary(ctx, userId, element.id, function(err) {
                if (err) {
                    return callback(err);
                }
                return callback();
            });
            break;

        case DISCUSSION:
            DiscussionAPI.removeDiscussionFromLibrary(ctx, userId, element.id, function(err) {
                if (err) {
                    return callback(err);
                }
                return callback();
            }); 
            break;

        case MEETING:
            MeetingsAPI.Meetings.removeMeetingFromLibrary(ctx, userId, element.id, function(err) {
                if (err) {
                    return callback(err);
                }
                return callback();
            });
            break;

        case GROUP:
            var changes = {}; 
            changes[userId] = false;

            AuthzAPI.updateRoles(element.id, changes, function(err) {
                if (err) {
                    return callback(err);
                }
                return callback();
            });
            break;
    }
};

/**
 * Get members of a resource
 *
 * @param  {Context}    ctx                     Standard context object containing the current user and the current tenant
 * @param  {String}     elementId               The element id 
 * @param  {String}     type                    Type can be : 'content', 'folder', 'discussion', 'meeting' or 'group'
 * @param  {Function}   callback                Standard callback function
 * @param  {Object}     callback.err            An error that occured, if any
 * @param  {Object}     callback.memberList     List of members
 * @api private
 */
var _getMembers = function(ctx, elementId, type, callback) {

    switch (type) {
        case CONTENT:
            ContentAPI.getContentMembersLibrary(ctx, elementId, null, null, function(err, memberList) {
                if (err) {
                    return callback(err);
                }
                return callback(null, memberList);
            });
            break;

        case FOLDER:
            FolderAPI.getFolderMembers(ctx, elementId, null, null, function(err, memberList) {
                if (err) {
                    return callback(err);
                }
                return callback(null, memberList);
            });
            break;

        case DISCUSSION:
            DiscussionAPI.getDiscussionMembers(ctx, elementId, null, null, function(err, memberList) {
                if (err) {
                    return callback(err);
                }
                return callback(null, memberList);
            });
            break;

        case MEETING:
           MeetingsAPI.Meetings.getMeetingMembers(ctx, elementId, null, null, function(err, memberList) {
                if (err) {
                    return callback(err);
                }
                return callback(null, memberList);
            });
            break;

        case GROUP:
            GroupAPI.getMembersLibrary(ctx, elementId, null, null, function(err, memberList) {
                if (err) {
                    return callback(err);
                }
                return callback(null, memberList);
            });
            break;
    }
};

/**
 * Get members of a resource
 *
 * @param  {Context}    ctx                     Standard context object containing the current user and the current tenant
 * @param  {String}     userId                  The user id 
 * @param  {String}     type                    Type can be : 'content', 'folder', 'discussion', 'meeting' or 'group'
 * @param  {Function}   callback                Standard callback function
 * @param  {Object}     callback.err            An error that occured, if any
 * @param  {Object}     callback.resourceList   List of resources
 * @api private
 */
var _getLibrary = function(ctx, userId, type, callback) {

    switch (type) {
        case CONTENT:
            ContentAPI.getContentLibraryItems(ctx, userId, null, null, function(err, contents) {
                if (err) {
                    return callback(err);
                }
                return callback(null, contents);
            });
            break;

        case FOLDER:
            FolderAPI.getFoldersLibrary(ctx, userId, null, null, function(err, folders) {
                if (err) {
                    return callback(err);
                }
                return callback(null, folders);
            });
            break;

        case DISCUSSION:
            DiscussionAPI.getDiscussionsLibrary(ctx, userId, null, null, function(err, discussions) {
                if (err) {
                    return callback(err);
                }
                return callback(null, discussions);
            }); 
            break;

        case MEETING:
            MeetingsAPI.Meetings.getMeetingsLibrary(ctx, userId, null, null, function(err, meetings) {
                if (err) {
                    return callback(err);
                }
                return callback(null, meetings);
            });
            break;

        case GROUP:
            GroupAPI.getMembershipsLibrary(ctx, userId, null, null, function(err, groups) {
                if (err) {
                    return callback(err);
                }
                return callback(null, groups);
            });
            break;
    }
};

/**
 * Update roles and make the archive user manager of the resource
 *
 * @param  {Context}    ctx                 Standard context object containing the current user and the current tenant
 * @param  {Object}     element             The resource 
 * @param  {String}     archiveUser         User Archive
 * @param  {String}     type                Type can be : 'content', 'folder', 'discussion', 'meeting' or 'group'
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occured, if any
 * @api private
 */
var _updateRoles = function(ctx, element, archiveUser, type, callback) {

    // Set update details
    var update = {}; 
    update[archiveUser.archiveId] = AuthzConstants.role.MANAGER;

    // if (type === FOLDER) {
    //     AuthzAPI.updateRoles(element.groupId, update, function(err, update) {
    //         if (err) {
    //             return callback(err);
    //         }
    //         return callback();

    //     });
    // } else {
    //     AuthzAPI.updateRoles(element.id, update, function(err, update) {
    //         if (err) {
    //             return callback(err);
    //         }
    //         return callback();
    //     });
    // }
    switch (type) {
        
        case 'content':
            ContentAPI.setContentPermissions(ctx, element.id, update, function(err) {
                if (err) {
                    return callback(err);
                }
                return callback();
            });
            break;

        case 'discussion':
            DiscussionAPI.setDiscussionPermissions(ctx, element.id, update, function(err) {
                if (err) {
                    return callback(err);
                }
                return callback();
            }); 
            break;

        case 'meeting':
            ResourceActions.setRoles(ctx, element, update, function (err, memberChangeInfo) {
                if (err) {
                    return callback(err);
                }

                MeetingsAPI.emit(MeetingsConstants.events.UPDATED_MEETING_MEMBERS, ctx, element, memberChangeInfo, {}, function(errs) {
                    if (errs) {
                        return callback(_.first(errs));
                    }
                    return callback();
                });
            });
            break;

        case 'group':
            ResourceActions.setRoles(ctx, element, update, function(err, memberChangeInfo) {
                if (err) {
                    return callback(err)
                } 

                PrincipalsUtil.touchLastModified(element, function(err, updatedGroup) {
                    if (err) {
                        return callback(err);
                    }

                    PrincipalsEmitter.emit(PrincipalsConstants.events.UPDATED_GROUP_MEMBERS, ctx, updatedGroup, element, memberChangeInfo, {}, function(errs) {
                        if (errs) {
                            return callback(_.first(errs));
                        }
                        return callback();
                    });
                });
            });
            break;

        case 'folder':
            AuthzAPI.updateRoles(element.groupId, update, function(err) {
                if (err) {
                    return callback(err);
                }
                return callback();
            });
            break;
    }
};

/** =============================================================== **/
/** ===================== Definitive deletion ===================== **/
/** =============================================================== **/

/**
 * Definitive delete of user
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {String}     user            User that will be deleted
 * @param  {String}     alias           Tenant alias
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 */
var definitiveDeletion = module.exports.definitiveDeletion = function(ctx, user, alias, callback) {
    var callback = callback || function() {};

    // Get userArchive
    PrincipalsDAO.getArchivedUser(alias, function(err, userArchive) {
        if (err) {
            return callback(err);
        }   

        // Delete all rights on resources
        _deleteResourceRights(ctx, user, userArchive.archiveId, function(err) {
            if (err) {
                return callback(err);
            } 

            // Remove profile picture from file system
            removeProfilePicture(ctx, user, function(err) {
                if (err) {
                    return callback(err);
                }

                // Delete followers and following
                FollowingAPI.deleteFollowers(ctx, user, function(err) {
                    if (err) {
                        return callback(err);
                    }
                    FollowingAPI.deleteFollowing(ctx, user, function(err) {
                        if (err) {
                            return callback(err);
                        }

                        // Determine if the principal was already deleted in the authz index before we set them and flip
                        // the principals deleted flag
                        AuthzDelete.isDeleted([user.id], function(err, wasDeleted) {
                            if (err) {
                                return callback(err);
                            }

                            // Get the login to delete the user form the table AuthenticationLoginId
                            require('oae-authentication').getUserLoginIds(ctx, user.id, function(err, login) {
                                if (err) {
                                    return callback(err);
                                }

                                // Set (or re-Set) the principal as deleted in the authz index
                                AuthzDelete.setDeleted(user.id, function(err) {
                                    if (err) {
                                        return callback(err);
                                    }

                                    // Delete a user from the database
                                    PrincipalsDAO.fullyDeletePrincipal(user, login, function(err) {
                                        if (err) {
                                            return callback(err);
                                        }

                                        // Notify that a user has been deleted
                                        PrincipalsEmitter.emit(PrincipalsConstants.events.DELETED_USER, ctx, user, function(err) {
                                            if (err) {
                                                return callback(err);
                                            }

                                            // Notify consumers that a user has been deleted
                                            log().info({'userId': user.id, 'name': 'oae-principals'}, 'Definitive deletion user with a mapped login id');
                                            
                                            return callback(null, true);
                                        });
                                    });
                                });
                            });
                        });
                    });
                });
            });
        });
    });
};

/**
 * Remove profile picture from file system
 *
 * @param  {String}     user            The user to delete
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 */
var removeProfilePicture = module.exports.removeProfilePicture = function(ctx, user, callback) {
    PrincipalsDAO.getPrincipal(user.id, function (err, principal) {
        if (err) {
            return callback(err);
        }

        if (_.isEmpty(principal.picture)) {
            return callback();
        } else {
            // Remove profile picture
            var pathSmallPicture = principal.picture.smallUri.split(':');
            var pathMediumPicture = principal.picture.mediumUri.split(':');
            var pathLargePicture = principal.picture.largeUri.split(':');

            var path = ContentUtil.getStorageBackend(ctx, principal.picture.largeUri).getRootDirectory();

            fs.unlink(path + '/' + pathSmallPicture[1]);
            fs.unlink(path + '/' + pathMediumPicture[1]);
            fs.unlink(path + '/' + pathLargePicture[1]);

            return callback();
        }
    });
};

/**
 * Remove all his rights on resources
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {String}     user            The user to delete
 * @param  {String}     archiveId       The user archive of the tenant
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 * @api private
 */
var _deleteResourceRights = function(ctx, user, archiveId, callback) {

    // Get all data from user archive where data belonged to the user removed
    PrincipalsDAO.getDataFromArchive(archiveId, user.id, function(err, data) {
        if (err) {
            return callback(err);
        } 
        if (_.isEmpty(data)) {
            return callback();
        } else {

            var ids = data.resourceId.split(',');

            // Remove the data from the table 'DataArchive'
            PrincipalsDAO.removeDataInArchive(archiveId, user.id, function(err) {
                if (err) {
                    return callback(err);
                }    

                // If there is no ids, return callback
                if (_.isEmpty(ids)) {
                    return callback();
                }

                async.eachSeries(ids, function(id, innerCallback) {

                    // if the data belonged to the removed user, delete the resource
                    if (data.principalId === user.id) {
                        
                        // Get type of resource
                        var idResource = id;
                        var splitId = idResource.split(':');

                        // If it's a folder get the idGroup
                        _ifFolderGetIdGroup(ctx, idResource, splitId[0], function(err, idFolder) {
                            if (err) {
                                return callback(err);
                            }
                            if (idFolder) {
                                idResource = idFolder;
                            }
                            
                            // Get role
                            AuthzAPI.getAllAuthzMembers(idResource, function(err, memberIdRoles) {

                                var managers = 0;
                                var rolesMember = _.pluck(memberIdRoles, 'role');

                                _.each(rolesMember, function(role) {
                                    if (role === AuthzConstants.role.MANAGER) {
                                        managers = managers + 1;
                                    }
                                });
                                
                                var del = managers > 1 ? false : true;

                                // Remove the resource with the appropriate method
                                switch(splitId[0]) {
                                    case CONTENT_PREFIX:
                                        _deleteRightToContent(ctx, del, archiveId, idResource, function(err) {
                                            if (err) {
                                                return callback(err);
                                            }
                                        });
                                        break;
                                    case DISCUSSION_PREFIX:
                                        _deleteRightToDiscussion(ctx, del, archiveId, idResource, function(err) {
                                            if (err) {
                                                return callback(err);
                                            }
                                        });
                                        break;
                                    case FOLDER_PREFIX:
                                        _deleteRightToFolder(ctx, del, archiveId, id, function(err) {
                                            if (err) {
                                                return callback(err);
                                            }
                                        });
                                        break;
                                    case GROUP_PREFIX:
                                        _deleteRightToGroup(ctx, del, archiveId, idResource, function(err) {
                                            if (err) {
                                                return callback(err);
                                            }
                                        });
                                        break;
                                    case MEETING_PREFIX:
                                        _deleteRightToMeeting(ctx, del, archiveId, idResource, function(err) {
                                            if (err) {
                                                return callback(err);
                                            }
                                        });
                                        break;
                                }
                                return innerCallback();
                            });
                        });
                    }
                }, function() {
                    return callback();
                });
            });
        }
    });
};

/**
 * Remove right on content or remove it if there is no manager 
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {String}     idResource      The id of the resource to delete
 * @param  {String}     splitId        The first part of a resource id (e.g. f:test:ryfQL_D4b, it will be 'f') who define the type of resource
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 * @api private
 */
var _ifFolderGetIdGroup = function(ctx, idResource, splitId, callback) {
    if (splitId === FOLDER_PREFIX) {
        FolderAPI.getFolder(ctx, idResource, function(err, folder) { 
            if (err) {
                return callback(err);
            }
            return callback(null, folder.groupId);
        });
    } else {
        return callback();
    }
};

/**
 * Remove right on content or remove it if there is no manager 
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {Boolean}    del             Boolean who determine if the element should be remove or just removed from the library
 * @param  {String}     archiveId       The id user archive of the tenant
 * @param  {String}     idResource      The id of the resource to delete
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 * @api private
 */
var _deleteRightToContent = function(ctx, del, archiveId, idResource, callback) {
    if (del) {
        ContentAPI.deleteContent(ctx, idResource, function(err) {
            if (err) {
                return callback(err);
            }
            return callback();
        });
    } else {
        // If there is another manager on the content, remove it from the library
        ContentAPI.removeContentFromLibrary(ctx, archiveId, idResource, function(err) {
            if (err) {
                return callback(err);
            }
            return callback();
        });
    }
};

/**
 * Remove right on discussion or remove it if there is no manager 
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {Boolean}    del             Boolean who determine if the element should be remove or just removed from the library
 * @param  {String}     archiveId       The id user archive of the tenant
 * @param  {String}     idResource      The id of the resource to delete
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 * @api private
 */
var _deleteRightToDiscussion = function(ctx, del, archiveId, idResource, callback) {
    if (del) {
        // Remove the actual discussion profile
        DiscussionAPI.deleteDiscussion(ctx, idResource, function(err) {
            if (err) {
                return callback(err);
            }
            return callback();
        });
    } else {
        // If there is another manager on the discussion, remove it from the library
        DiscussionAPI.removeDiscussionFromLibrary(ctx, archiveId, idResource, function(err) {
            if (err) {
                return callback(err);
            }
            return callback();
        });        
    }
};

/**
 * Remove right on folder or remove it if there is no manager 
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {Boolean}    del             Boolean who determine if the element should be remove or just removed from the library
 * @param  {String}     archiveId       The id user archive of the tenant
 * @param  {String}     idResource      The id of the resource to delete
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 * @api private
 */
var _deleteRightToFolder = function(ctx, del, archiveId, idResource, callback) {
    if (del) {
        // Remove folder
        FolderAPI.deleteFolder(ctx, idResource, false, function(err, content) {
            if (err) {
                return callback(err);
            }
            return callback();
        });
    } else {
        // If there is another manager on the folder, remove it from the library
        FolderAPI.removeFolderFromLibrary(ctx, archiveId, idResource, function(err) {
            if (err) {
                return callback(err);
            }
            return callback();
        });
    }
};

/**
 * Remove right on group or remove it if there is no manager 
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {Boolean}    del             Boolean who determine if the element should be remove or just removed from the library
 * @param  {String}     archiveId       The id user archive of the tenant
 * @param  {String}     idResource      The id of the resource to delete
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 * @api private
 */
var _deleteRightToGroup = function(ctx, del, archiveId, idResource, callback) {
    if (del) {
        // Remove group
        GroupAPI.deleteGroup(ctx, idResource, function(err) {
            if (err) {
                return callback(err);
            }
            // Remove roles
            var update = {};
            update[archiveId] = false;
            AuthzAPI.updateRoles(idResource, update, function(err, usersToInvalidate) {
                if (err) {
                    return callback(err);
                }
                return callback();                    
            });
        });
    } else {
        // If there is another manager on the group, remove it from the library
        GroupAPI.leaveGroup(ctx, idResource, function(err) {
            if (err) {
                return callback(err);
            }
            return callback();
        });
    }
};

/**
 * Remove right on meeting or remove it if there is no manager 
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {Boolean}    del             Boolean who determine if the element should be remove or just removed from the library
 * @param  {String}     archiveId       The id user archive of the tenant
 * @param  {String}     idResource      The id of the resource to delete
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occured, if any
 * @api private
 */
var _deleteRightToMeeting = function(ctx, del, archiveId, idResource, callback) {
    if (del) {
        // Remove meeting
        MeetingsAPI.Meetings.deleteMeeting(ctx, idResource, function(err) {
            if (err) {
                return callback(err);
            }
            return callback();
        });
    } else {
        // If there is another manager on the meeting, remove it from the library
        MeetingsAPI.Meetings.removeMeetingFromLibrary(ctx, archiveId, idResource, function(err) {
            if (err) {
                return callback(err);
            }
            return callback();
        });
    }
};
