// oae-principals/lib/search.js
/*!
 * Copyright 2012 Sakai Foundation (SF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://www.osedu.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
var _ = require('underscore');
var AuthzUtil = require('oae-authz/lib/util');
var SearchAPI = require('oae-search');

/**
 * Index a group.
 *
 * @param {User}        group           The Group object to index.
 * @param {String[]}    members         An array of principal id's representing the members of the group
 * @param {Function}    callback        Invoked when the process completes
 * @param {Object}      callback.err    An error that occurred, if any
 */
module.exports.indexGroup = function(group, members, callback) {
    members = members || [];

    var id = group.id;
    var tenantId = AuthzUtil.getPrincipalFromUuid(id).tenantId;
    var visibility = group.visibility;
    var general = '';

    if (group.name) {
        general += ' ' + group.name;
    }

    if (group.description) {
        general += ' ' + group.description;
    }

    if (group.alias) {
        general += ' ' + group.alias;
    }

    general = general.trim();
    sort = general;

    var groupDoc = {
        id: id,
        tenantId: tenantId,
        name: group.name,
        description: group.description,
        visibility: visibility,
        general: general,
        sort: sort,
        access: members
    };

    SearchAPI.index('group', groupDoc, callback);
}

/**
 * Index a user.
 *
 * @param {User}        user            The User object to index.
 * @param {Function}    callback        Invoked when the process completes
 * @param {Object}      callback.err    An error that occurred, if any
 */
module.exports.indexUser = function(user, callback) {
    var id = user.id;
    var tenantId = AuthzUtil.getPrincipalFromUuid(id).tenantId;
    var visibility = user.visibility;
    var general = '';
    var displayName = user.displayName;

    if (user.displayName) {
        general += ' ' + user.displayName;
    }

    if (user.username) {
        general += ' ' + user.username;
    }

    general = general.trim();
    sort = general;

    var userDoc = {
        id: id,
        tenantId: tenantId,
        name: user.displayName,
        visibility: visibility,
        general: general,
        sort: sort
    };

    SearchAPI.index('user', userDoc, callback);
}
