// oae-content/lib/test/util.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var assert = require('assert');
var ShortId = require('shortid');
var util = require('util');

var AuthzTestUtil = require('oae-authz/lib/test/util');
var ConfigTestsUtil = require('oae-config/lib/test/util');
var MqTestUtil = require('oae-util/lib/test/mq-util');
var LibraryAPI = require('oae-library');
var LibraryTestUtil = require('oae-library/lib/test/util');
var RestAPI = require('oae-rest');
var TaskQueue = require('oae-util/lib/taskqueue');
var TestsUtil = require('oae-tests/lib/util');

var ContentConstants = require('oae-content/lib/constants').ContentConstants;

/**
 * Set up 2 public tenants and 2 private tenants, each with a public, loggedin, private set of users, groups and
 * content. The resulting model looks like this:
 *
 * ```
 *  {
 *      "publicTenant": {
 *          "tenant": <Tenant>,
 *          "anonymousRestContext": <RestContext>,
 *          "adminRestContext": <RestContext>,
 *          "publicGroup": <Group>,
 *          "loggedinGroup": <Group>,
 *          "privateGroup": <Group>,
 *          "publicContent": <Content>,
 *          "loggedinContent": <Content>,
 *          "privateContent": <Content>,
 *          "publicUser": {
 *              "user": <User>,
 *              "restContext": <RestContext>
 *          },
 *          "loggedinUser": { ... }
 *          "privateUser": { ... }
 *      },
 *      "publicTenant1": { ... },
 *      "privateTenant": { ... },
 *      "privateTenant1": { ... }
 *  }
 * ```
 *
 * @param  {Function}   Invoked when all the entities are set up
 * @throws {Error}      An assertion error is thrown if something does not get created properly
 */
var setupMultiTenantPrivacyEntities = module.exports.setupMultiTenantPrivacyEntities = function(callback) {
    // Create the tenants and users
    TestsUtil.setupMultiTenantPrivacyEntities(function(publicTenant, publicTenant1, privateTenant, privateTenant1) {
        // Create the content
        _setupTenant(publicTenant, function() {
            _setupTenant(publicTenant1, function() {
                _setupTenant(privateTenant, function() {
                    _setupTenant(privateTenant1, function() {
                        return callback(publicTenant, publicTenant1, privateTenant, privateTenant1);
                    });
                });
            });
        });
    });
};

/**
 * Ensure that the provided etherpad content matches the expected content that was persisted. This
 * assertion abstracts things such as the `<html></html>` pre-amble of an etherpad document so
 * the consumer only needs to focus on the content it persisted
 *
 * @param  {String}         actualContent       The actual content that was read from the Etherpad API
 * @param  {String}         expectedContent     The expected contents of the etherpad document
 * @throws {AssertionError}                     Thrown if the actual content did not match the expected content
 */
var assertEtherpadContentEquals = module.exports.assertEtherpadContentEquals = function(actualContent, expectedContent) {
    // Wrap the expected content into the etherpad document structure. Also add a line break at the
    // end of the provided content. This is not extremely robust and will likely only be valid for
    // asserting simple single-line content, so it's possible if we have more complex test cases
    // we'll need to revisit this logic
    expectedContent =
        '<!DOCTYPE HTML>' +
        '<html>' +
            '<body>' +
                expectedContent + '<br>' +
            '</body>' +
        '</html>';

    assert.strictEqual(actualContent, expectedContent);
};

/**
 * Try and get the members library of a content item, ensuring the request fails in a particular way
 *
 * @param  {RestContext}    restContext     The context to use to issue the request
 * @param  {String}         [contentId]     The id of the content item whose members library to get
 * @param  {String}         [start]         From where in the list to start listing results
 * @param  {Number}         [limit]         The maximum number of member results to return
 * @param  {Number}         httpCode        The expected failure HTTP code of the get content members library request
 * @param  {Function}       callback        Invoked when the assertions have completed
 * @throws {AssertionError}                 Thrown if the the request did not fail in the expected manner
 */
var assertGetContentMembersFails = module.exports.assertGetContentMembersFails = function(restContext, contentId, start, limit, httpCode, callback) {
    RestAPI.Content.getMembers(restContext, contentId, start, limit, function(err, result) {
        assert.ok(err);
        assert.strictEqual(err.code, httpCode);
        return callback();
    });
};

/**
 * Get the members library of a content item, ensuring the request is successful
 *
 * @param  {RestContext}    restContext                         The context to use to issue the request
 * @param  {String}         contentId                           The id of the content item whose members library to get
 * @param  {String}         [start]                             From where in the list to start listing results
 * @param  {Number}         [limit]                             The maximum number of member results to return. Default: 25
 * @param  {Function}       callback                            Invoked when the assertions have completed
 * @param  {Object}         callback.result                     The member results
 * @param  {Object[]}       callback.result.results             An array of members library results
 * @param  {User|Group}     callback.result.results[i].profile  The profile of the user or group member
 * @param  {String}         callback.result.results[i].role     The role the user or group has on the content item
 * @param  {String}         callback.result.nextToken           The token to use for the next members library request to get the next page of members
 * @throws {AssertionError}                                     Thrown if there is an error getting the content members library
 */
var assertGetContentMembersSucceeds = module.exports.assertGetContentMembersSucceeds = function(restContext, contentId, start, limit, callback) {
    RestAPI.Content.getMembers(restContext, contentId, start, limit, function(err, result) {
        assert.ok(!err);
        assert.ok(result);
        assert.ok(_.isArray(result.results));
        assert.ok(_.isString(result.nextToken) || _.isNull(result.nextToken));

        // If a valid limit was specified (valid meaning above 0 and below the maximum amount of
        // 25), ensure the `nextToken` is shown if there were less than the expected amount of
        // results
        if (_.isNumber(limit) && limit > 0 && result.results.length < limit && limit <= 25) {
            assert.strictEqual(result.nextToken, null);
        }

        // Ensure each result has an id and a valid role
        _.each(result.results, function(result) {
            assert.ok(result);
            assert.ok(result.profile);
            assert.ok(result.profile.id);
            assert.ok(_.contains(['manager', 'viewer'], result.role));
        });

        return callback(result);
    });
};

/**
 * Try and update the members of a content item, ensuring that the request fails in a specified manner
 *
 * @param  {RestContext}    managerRestContext  The rest context of a manager of the content item. This is needed to ensure that the membership is not impacted by the failure
 * @param  {RestContext}    actorRestContext    The rest context of the user who should perform the update members action
 * @param  {String}         [contentId]         The id of the content item whose members to try and update
 * @param  {Object}         [roleChanges]       A hash keyed by user id, whose values are the role to set for each member
 * @param  {Number}         httpCode            The expected failure HTTP code of the update members request
 * @param  {Function}       callback            Invoked when the assertions have completed
 * @throws {AssertionError}                     Thrown if there is an error ensuring that the request fails in the specified manner
 */
var assertUpdateContentMembersFails = module.exports.assertUpdateContentMembersFails = function(managerRestContext, actorRestContext, contentId, roleChanges, httpCode, callback) {
    // Get the members library so we can ensure it does not change after the failure
    getAllContentMembers(managerRestContext, contentId, null, function(membersBeforeUpdate) {
        membersBeforeUpdate = AuthzTestUtil.getMemberRolesFromResults(membersBeforeUpdate);

        // Perform the update and ensure it fails as expected
        RestAPI.Content.updateMembers(actorRestContext, contentId, roleChanges, function(err) {
            assert.ok(err);
            assert.strictEqual(err.code, httpCode);

            // Ensure the library members did not change
            getAllContentMembers(managerRestContext, contentId, null, function(membersAfterUpdate) {
                AuthzTestUtil.assertMembershipEquals(membersBeforeUpdate, null, AuthzTestUtil.getMemberRolesFromResults(membersAfterUpdate));

                // Test once more that the library did not change by purging and rebuilding it
                _purgeMembersLibrary(contentId, function() {

                    // Ensure the library members still did not change
                    getAllContentMembers(managerRestContext, contentId, null, function(membersAfterUpdate) {
                        AuthzTestUtil.assertMembershipEquals(membersBeforeUpdate, null, AuthzTestUtil.getMemberRolesFromResults(membersAfterUpdate));
                        return callback();
                    });
                });
            });
        });
    });
};

/**
 * Update the members of a content item, ensuring that the request fails in a specified manner
 *
 * @param  {RestContext}    managerRestContext  The rest context of a manager of the content item. This is needed to ensure that the membership is impacted by the role changes in the expected manner
 * @param  {RestContext}    actorRestContext    The rest context of the user who should perform the update members action
 * @param  {String}         contentId           The id of the content item whose members to update
 * @param  {Object}         roleChanges         A hash keyed by user id, whose values are the role to set for each member
 * @param  {Function}       callback            Invoked when the assertions have completed
 * @throws {AssertionError}                     Thrown if there is an error ensuring that the update members operation was successful
 */
var assertUpdateContentMembersSucceeds = module.exports.assertUpdateContentMembersSucceeds = function(managerRestContext, actorRestContext, contentId, roleChanges, callback) {
    // Ensure the members library is currently built
    getAllContentMembers(managerRestContext, contentId, null, function(membersBeforeUpdate) {
        membersBeforeUpdate = AuthzTestUtil.getMemberRolesFromResults(membersBeforeUpdate);

        // Perform the update, causing the library to update on-the-fly
        RestAPI.Content.updateMembers(actorRestContext, contentId, roleChanges, function(err) {
            assert.ok(!err);

            // Get the content members again and ensure they have the expected members and roles
            getAllContentMembers(managerRestContext, contentId, null, function(membersAfterUpdate) {
                AuthzTestUtil.assertMembershipEquals(membersBeforeUpdate, roleChanges, AuthzTestUtil.getMemberRolesFromResults(membersAfterUpdate));

                // Test the library once more by purging and building the library from scratch
                _purgeMembersLibrary(contentId, function() {

                    // Now that we're running with a fresh library, ensure that the members we receive are still what we expect
                    getAllContentMembers(managerRestContext, contentId, null, function(membersAfterUpdate) {
                        AuthzTestUtil.assertMembershipEquals(membersBeforeUpdate, roleChanges, AuthzTestUtil.getMemberRolesFromResults(membersAfterUpdate));

                        return callback();
                    });
                });
            });
        });
    });
};

/**
 * Try and share a content item, ensuring that the request fails in a specified manner
 *
 * @param  {RestContext}    managerRestContext  The rest context of a manager of the content item. This is needed to ensure that the membership is not impacted by the failure
 * @param  {RestContext}    actorRestContext    The rest context of the user who should perform the share action
 * @param  {String}         [contentId]         The id of the content item to try and share
 * @param  {String[]}       [memberIds]         The ids of the members with which to share the content item
 * @param  {Number}         httpCode            The expected failure HTTP code of the share content request
 * @param  {Function}       callback            Invoked when the assertions have completed
 * @throws {AssertionError}                     Thrown if there is an error ensuring that the request fails in the specified manner
 */
var assertShareContentFails = module.exports.assertShareContentFails = function(managerRestContext, actorRestContext, contentId, memberIds, httpCode, callback) {
    // Get the members library so we can ensure it does not change after the failure
    getAllContentMembers(managerRestContext, contentId, null, function(membersBeforeUpdate) {
        membersBeforeUpdate = AuthzTestUtil.getMemberRolesFromResults(membersBeforeUpdate);

        // Perform the update and ensure it fails as expected
        RestAPI.Content.shareContent(actorRestContext, contentId, memberIds, function(err) {
            assert.ok(err);
            assert.strictEqual(err.code, httpCode);

            // Ensure the library members did not change
            getAllContentMembers(managerRestContext, contentId, null, function(membersAfterUpdate) {
                AuthzTestUtil.assertMembershipEquals(membersBeforeUpdate, null, AuthzTestUtil.getMemberRolesFromResults(membersAfterUpdate));

                // Test once more that the library did not change by purging and rebuilding it
                _purgeMembersLibrary(contentId, function() {

                    // Ensure the library members still did not change
                    getAllContentMembers(managerRestContext, contentId, null, function(membersAfterUpdate) {
                        AuthzTestUtil.assertMembershipEquals(membersBeforeUpdate, null, AuthzTestUtil.getMemberRolesFromResults(membersAfterUpdate));
                        return callback();
                    });
                });
            });
        });
    });
};

/**
 * Share the content item with the given list of users, ensuring it succeeds and the content item is
 * added to the content item members library
 *
 * @param  {RestContext}    managerRestContext  The rest context of a manager of the content item. This is needed to ensure that the membership is impacted by the sharing in the expected manner
 * @param  {RestContext}    actorRestContext    The rest context of the user who should perform the share action
 * @param  {String}         contentId           The id of the content item to share
 * @param  {String[]}       memberIds           The ids of the members with which to share the content item
 * @param  {Function}       callback            Invoked when the sharing has completed
 * @throws {AssertionError}                     Thrown if there is an error verifying that the content item is successfully shared
 */
var assertShareContentSucceeds = module.exports.assertShareContentSucceeds = function(managerRestContext, actorRestContext, contentId, memberIds, callback) {
    // Ensure the members library is currently built
    getAllContentMembers(managerRestContext, contentId, null, function(membersBeforeUpdate) {
        membersBeforeUpdate = AuthzTestUtil.getMemberRolesFromResults(membersBeforeUpdate);

        // Build a role update object that represents the change that should occur in the share
        // operation
        var roleChange = {};
        _.each(memberIds, function(memberId) {
            if (!membersBeforeUpdate[memberId]) {
                membersBeforeUpdate[memberId] = 'viewer';
            }
        });

        // Perform the share action, causing the library to update on-the-fly
        RestAPI.Content.shareContent(actorRestContext, contentId, memberIds, function(err) {
            assert.ok(!err);

            // Get the content members again and ensure they have the expected members and roles
            getAllContentMembers(managerRestContext, contentId, null, function(membersAfterUpdate) {
                AuthzTestUtil.assertMembershipEquals(membersBeforeUpdate, roleChange, AuthzTestUtil.getMemberRolesFromResults(membersAfterUpdate));

                // Test the library once more by purging and building the library from scratch
                _purgeMembersLibrary(contentId, function() {

                    // Now that we're running with a fresh library, ensure that the members we receive are what we expect
                    getAllContentMembers(managerRestContext, contentId, null, function(membersAfterUpdate) {
                        AuthzTestUtil.assertMembershipEquals(membersBeforeUpdate, roleChange, AuthzTestUtil.getMemberRolesFromResults(membersAfterUpdate));

                        return callback();
                    });
                });
            });
        });
    });
};

/**
 * Get the full content members library of a content item
 *
 * @param  {RestContext}    restContext         The rest context to use to get the content members library
 * @param  {String}         contentId           The id of the content item whose members to get
 * @param  {Object}         [opts]              Optional arguments for getting the content members library
 * @param  {Number}         [opts.batchSize]    The size of the pages to use when paging through the content members library
 * @param  {Function}       callback            Invoked when all the members in the library have been fetched
 * @param  {Object[]}       callback.members    An array of users and groups that were fetched from the library
 * @param  {Object[][]}     callback.responses  The raw response objects for each page request that was made to get the content members library
 * @throws {AssertionError}                     Thrown if an error occurrs while paging through the content members library
 */
var getAllContentMembers = module.exports.getAllContentMembers = function(restContext, contentId, opts, callback, _members, _responses, _nextToken) {
    _members = _members || [];
    _responses = _responses || [];
    if (_nextToken === null) {
        return callback(_members, _responses);
    }

    opts = opts || {};
    opts.batchSize = opts.batchSize || 25;
    assertGetContentMembersSucceeds(restContext, contentId, _nextToken, opts.batchSize, function(result) {
        _responses.push(result);
        return getAllContentMembers(restContext, contentId, opts, callback, _.union(_members, result.results), _responses, result.nextToken);
    });
};

/**
 * Create the content within a tenant
 *
 * @param  {Tenant}     tenant          The tenant to setup
 * @param  {Function}   callback        Invoked when all the entities are set up
 * @throws {Error}                      An assertion error is thrown if something does not get created properly
 * @api private
 */
var _setupTenant = function(tenant, callback) {
    _createMultiPrivacyContent(tenant.adminRestContext, function(publicContent, loggedinContent, privateContent) {
        tenant.publicContent = publicContent;
        tenant.loggedinContent = loggedinContent;
        tenant.privateContent = privateContent;
        return callback();
    });
};

/**
 * Set up content of all privacies using the given rest context
 *
 * @param  {RestContext}    restCtx         The rest context to use
 * @param  {Function}       callback        Invoked when all the entities are set up
 * @throws {Error}                          An assertion error is thrown if something does not get created properly
 * @api private
 */
var _createMultiPrivacyContent = function(restCtx, callback) {
    _createContentWithVisibility(restCtx, 'public', function(publicContent) {
        _createContentWithVisibility(restCtx, 'loggedin', function(loggedinContent) {
            _createContentWithVisibility(restCtx, 'private', function(privateContent) {
                return callback(publicContent, loggedinContent, privateContent);
            });
        });
    });
};

/**
 * Create a piece of content with the specified visibility
 *
 * @param  {RestContext}    restCtx             The rest context to use
 * @param  {String}         visibility          The visibility of the user
 * @param  {Function}       callback            Invoked when all the entities are set up
 * @param  {Content}        callback.content    The piece of content that was created
 * @throws {Error}                              An assertion error is thrown if something does not get created properly
 * @api private
 */
var _createContentWithVisibility = function(restCtx, visibility, callback) {
    var randomId = util.format('%s-%s', visibility, ShortId.generate());
    RestAPI.Content.createLink(restCtx, 'displayName-' + randomId, 'description-' + randomId, visibility, 'http://www.oaeproject.org', null, null, function(err, content) {
        assert.ok(!err);
        return callback(content);
    });
};

/**
 * Create a set of test users and a collaborative document.
 * The `nrOfJoinedUsers` specifies how many users should join the document
 *
 * @param  {RestContext}    adminRestContext        An administrator rest context that can be used to create the users
 * @param  {Number}         nrOfUsers               The number of users that should be created
 * @param  {Number}         nrOfJoinedUsers         The number of users that should be joined in the document. These will be the first `nrOfJoinedUsers` of the users hash
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            Standard error object, if any
 * @param  {Content}        callback.contentObj     The created collaborative document
 * @param  {Object}         callback.users          The created users
 * @param  {Object}         callback.user1          An object containing the user profile and a rest context for the first user of the set of users that was created
 * @param  {Object}         callback.user..         An object containing the user profile and a rest context for the next user of the set of users that was created
 */
var createCollabDoc = module.exports.createCollabDoc = function(adminRestContext, nrOfUsers, nrOfJoinedUsers, callback) {
    TestsUtil.generateTestUsers(adminRestContext, nrOfUsers, function(err, users) {
        assert.ok(!err);

        var userIds = _.keys(users);
        var userValues = _.values(users);

        // Create a collaborative document where all the users are managers
        var name = TestsUtil.generateTestUserId('collabdoc');
        RestAPI.Content.createCollabDoc(userValues[0].restContext, name, 'description', 'public', userIds, [], function(err, contentObj) {
            assert.ok(!err);

            // Create a function that will get executed once each user has joined the document
            var callCallback = _.after(nrOfJoinedUsers, function() {
                var callbackArgs = _.union([contentObj, users], userValues);
                return callback.apply(callback, callbackArgs);
            });

            // If no user should join the document we can return immediately
            if (nrOfJoinedUsers === 0) {
                return callCallback();
            }


            // Join the collab doc for `nrOfJoinedUsers` users
            var joinCollabDoc = function(i) {
                var restCtx = userValues[i].restContext;
                RestAPI.Content.joinCollabDoc(restCtx, contentObj.id, function(err, data) {
                    assert.ok(!err);
                    callCallback();
                });
            };
            for (var i = 0; i < nrOfJoinedUsers; i++) {
                joinCollabDoc(i);
            }
        });
    });
};

/**
 * Publish a collaborative document. This function will mimick Etherpad's
 * publishing behaviour by sending the appropriate message to RabbitMQ
 *
 * @param  {String}     contentId   The ID of the content item
 * @param  {String}     userId      The ID of the user who will be publishing the collaborative document
 * @param  {Function}   callback    Standard callback function
 * @throws {Error}                  An assertion error is thrown when something unexpected occurs
 */
var publishCollabDoc = module.exports.publishCollabDoc = function(contentId, userId, callback) {
    var data = {
        'contentId': contentId,
        'userId': userId
    };
    TaskQueue.submit(ContentConstants.queue.ETHERPAD_PUBLISH, data, function(err) {
        assert.ok(!err);
    });

    MqTestUtil.whenTasksEmpty(ContentConstants.queue.ETHERPAD_PUBLISH, callback);
};

/**
 * Purge the members library. See @LibraryTestUtil.assertPurgeFreshLibrary for more details
 *
 * @param  {String}     contentId   The id of the content item whose members library to purge
 * @param  {Function}   callback    Invoked when the library has been purged
 * @throws {AssertionError}         Thrown if there is an issue purging the library
 * @api private
 */
var _purgeMembersLibrary = function(contentId, callback) {
    LibraryTestUtil.assertPurgeFreshLibraries(ContentConstants.library.MEMBERS_LIBRARY_INDEX_NAME, [contentId], callback);
};
