// oae-tenants/lib/api.js
var OAE = require('oae-util/lib/OAE');
var model = require('oae-tenants/lib/model');

/**
 * Get all the tenants in the datastore.
 *
 * @param {Function} callback A function(tenants, err) specifying the array of tenants in the
 * datastore.
 */ 
var getAllTenants = module.exports.getAllTenants = function(callback) {
    OAE.runQuery('select * from Tenant', false, function(err, rows) {
        if (!err) {
            var tenants = [];
            if (hasTenant(rows)) {
                for (var i = 0; i < rows.rowCount(); i++) {
                    tenants.push(mapToTenant(rows[i]));
                }
            }
            callback(tenants);
        } else {
            callback(null, err);
        }
    });
};

/**
 * Get the unique tenant by their mapped port.
 * 
 * @param {Number}   port The port number on which the tenant listens
 * @param {Function} callback A function(tenant, err) specifying the tenant requested
 */
var getTenantByPort = module.exports.getTenantByPort = function(port, callback) {
    OAE.runQuery('select * from Tenant where port = ?', [ port ], function(err, rows) {
        if (!err) {
            if (hasTenant(rows)) {
                var tenant = mapToTenant(rows[0]);
                callback(tenant);
            } else {
                callback(null);
            }
        } else {
            callback(null, err);
        }
    });
};

/**
 * Get the unique tenant by their alias.
 * 
 * @param {String}   alias The alias assigned to the tenant
 * @param {Function} callback A function(tenant, err) specifying the tenant requested
 */
var getTenantByAlias = module.exports.getTenantByAlias = function(alias, callback) {
    getAllTenants(function(tenants, err) {
        if (!err) {
            var tenant = null;
            if (tenants !== null) {
                for (var i = 0; i < tenants.length; i++) {
                    if (tenants[i].alias === alias) {
                        tenant = tenants[i];
                        break;
                    }
                }
            }
            callback(tenant);
        } else {
            callback(null, err);
        }
    });
};

/**
 * Create a tenant using the provided information.
 * 
 * @param {String}   alias The unique alias assigned to the tenant
 * @param {String}   name A descriptive short name for the tenant
 * @param {String}   description A long description of the tenant
 * @param {Number}   port The port on which the tenant should listen
 * @param {Function} callback A function(tenant, err) specifying the tenant information that was persisted.
 */
var createTenant = module.exports.createTenant = function(alias, name, description, port, callback) {
    getTenantByPort(port, function(tenant, err) {
        if (!err) {
            if (!tenant) {
                getTenantByAlias(alias, function(tenant, err) {
                    if (!err) {
                        if (!tenant) {
                            tenant = new model.Tenant(alias, name, description, port);
                            // no tenant by this alias or port exist, create one
                            OAE.runQuery('update Tenant set alias = ?, name = ?, description = ? where port = ?',
                                [tenant.alias, tenant.name, tenant.description, tenant.port], function(err) {
                                    if (!err) {
                                        callback(tenant);
                                    } else {
                                        callback(null, err);
                                    }
                                });
                        } else {
                            callback(null, {why: 'Tenant by alias "'+alias+'" alread exists.'});
                        }
                    } else {
                        callback(null, err);
                    }
                });
            } else {
                callback(null, {why: 'Tenant by port "'+port+'" already exists.'});
            }
        } else {
            callback(null, err);
        }
    });
};

/**
 * Determines whether or not this set of rows represents a result set with a tenant.
 */
function hasTenant(rows) {
    return (rows && rows[0] && rows[0].colHash && rows[0].colHash['alias']);
}

/**
 * Map a cassandra row element to a tenant object.
 */
function mapToTenant(row) {
    return new model.Tenant(row.colHash['alias'], row.colHash['name'], row.colHash['description'],
        row.colHash['port']);
}
