// oae-meetups/lib/internal/dao.js
/*!
 * Copyright 2017 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var util = require('util');

var AuthzUtil = require('oae-authz/lib/util');
var Cassandra = require('oae-util/lib/cassandra');
var log = require('oae-logger').logger('meetings-dao');
var OaeUtil = require('oae-util/lib/util');
var TenantsAPI = require('oae-tenants');

var Meetup = require('oae-meetups/lib/model').Meetup;
var Recording = require('oae-meetups/lib/model').Recording;

/** ####################################### **/
/** ############### MEETUPS ############### **/
/** ####################################### **/

/**
 * Create a new meeting.
 *
 * @param  {String}     meetingId           The id of the meeting
 * @param  {String}     createdBy           The id of the user creating the meeting
 * @param  {String}     displayName         The display name of the meeting
 * @param  {String}     [record]            Flag indicating that the meeting may be recorded
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Meetup} callback.meeting        The meeting object that was created
 */
var createMeetup = module.exports.createMeetup = function(meetingId, createdBy, displayName, record, callback) {
    var opts = {};

    var created = opts.created || Date.now();
    created = created.toString();

    var tenantAlias = AuthzUtil.getPrincipalFromId(createdBy).tenantAlias;

    var storageHash = {
        'tenantAlias': tenantAlias,
        'createdBy': createdBy,
        'displayName': displayName,
        'record': record.toString(),
        'created': created,
    };

    var query = Cassandra.constructUpsertCQL('Meetups', 'id', meetingId, storageHash);
    Cassandra.runQuery(query.query, query.parameters, function(err) {
        if (err) {
            log().error;
            return callback(err);
        }

        if (callback) {
            return callback(null, _storageHashToMeetup(meetingId, storageHash));
        }
    });
};

/**
 * Get a meetup basic profile by its id.
 *
 * @param  {String}     meetingId        The id of the meetup to get
 * @param  {Function}   callback         Standard callback function
 * @param  {Object}     callback.err     An error that occurred, if any
 * @param  {Meeting} callback.meeting The meeting object requested
 */
var getMeetupById = module.exports.getMeetupById = function(meetingId, callback) {

    Cassandra.runQuery('SELECT * FROM "Meetups" WHERE "id" = ?', meetingId, function(err, meeting) {
        if (err) {
            return callback(err);
        }

        return callback(null, meeting);
    });
};

/**
* @function end
* @param  {String} groupId  The id of the group to get
* @param  {function} callback Standard callback function
*/
var end = module.exports.end = function(groupId, callback) {
    log().info({'groupId': groupId}, 'Meetup ended');
    callback(null);
    // Update the table structure to add a field "ended" ???
    // Cassandra.runQuery('UPDATE "Meetups" FROM WHERE "id" = ?', [recordingId], callback);
};

/**
 * Get meetup by id
 *
 * @param  {String[]}       meetingIds           The id of the meetings to get
 * @param  {String[]}       [fields]             The meeting fields to select. If unspecified, selects all of them
 * @param  {Function}       callback             Standard callback function
 * @param  {Object}         callback.err         An error that occurred, if any
 * @param  {Meeting[]}   callback.meetings       The meeting objects requested, in the same order as the meeting ids
 */
var getMeetupsById = module.exports.getMeetupsById = function(meetingIds, fields, callback) {
    if (_.isEmpty(meetingIds)) {
        return callback(null, []);
    }

    var query;
    var parameters = [];

    // If `fields` was specified, we select only the fields specified. Otherwise we select all (i.e., *)
    if (fields) {
        var columns = _.map(fields, function(field) {
            return util.format('"%s"', field);
        });

        query = util.format('SELECT %s FROM "Meetups" WHERE "id" IN (?)', columns.join(','));
    } else {
        query = 'SELECT * FROM "Meetups" WHERE "id" IN (?)';
    }

    parameters.push(meetingIds);

    Cassandra.runQuery(query, parameters, function(err, rows) {
        if (err) {
            return callback(err);
        }

        // Convert the retrieved storage hashes into the Meeting model
        var meetings = {};
        _.chain(rows).map(Cassandra.rowToHash).each(function(row) {
            meetings[row.id] = _storageHashToMeetup(row.id, row);
        });

        // Order the meetings according to the array of meeting ids
        var orderedMeetings = _.map(meetingIds, function(meetingId) {
            return meetings[meetingId];
        });

        return callback(null, orderedMeetings);
    });
};

/**
* @function _storageHashToMeetup
* @param  {String} meetingId The id of meeting
* @param  {Object} hash      Hash of meetup info/settings
* @return {Meetup} The new created Meetup object
*/
var _storageHashToMeetup = function(meetingId, hash) {
    return new Meetup(
        TenantsAPI.getTenant(hash.tenantAlias),
        meetingId,
        hash.createdBy,
        hash.displayName,
        hash.record,
        OaeUtil.getNumberParam(hash.created)
    );
};

/**
 * Create a new recording link.
 *
 * @param  {String}     recordId            The id of the recording
 * @param  {String}     contentId           The id of the recording link object
 * @param  {String}     createdBy           The id of the user creating the recording link
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Recording} callback.recording   The recording object that was created
 */
var createRecording = module.exports.createRecording = function(recordId, contentId, createdBy, callback) {
    var opts = {};

    var created = opts.created || Date.now();
    created = created.toString();

    var recordId = recordId;
    var storageHash = {
        'contentId': contentId,
        'createdBy': createdBy,
        'created': created,
    };

    var query = Cassandra.constructUpsertCQL('Recordings', 'id', recordId, storageHash);
    Cassandra.runQuery(query.query, query.parameters, function(err) {
        if (err) {
            log().error;
            return callback(err);
        }

        if (callback) {
            return callback(null, _storageHashToRecording(recordId, storageHash));
        }
    });
};

/**
* @function _storageHashToRecording
* @param  {String} recordId The id of recording
* @param  {Object} hash     Hash of recording info/settings
* @return {Recording} The new Recording object
*/
var _storageHashToRecording = function(recordId, hash) {
    return new Recording(
        recordId,
        hash.contentID,
        hash.createdBy,
        OaeUtil.getNumberParam(hash.created)
    );
};

/**
 * Get a recording basic profile by its id.
 *
 * @param  {String}     recordId          The id of the recording to get
 * @param  {Function}   callback          Standard callback function
 * @param  {Object}     callback.err      An error that occurred, if any
 * @param  {Recording} callback.recording The recording object requested
 */
var getRecordingById = module.exports.getRecordingById = function(recordId, callback) {

    Cassandra.runQuery('SELECT * FROM "Recordings" WHERE "id" = ?', recordId, function(err, recording) {
        if (err) {
            return callback(err);
        }

        return callback(null, recording);
    });
};

/**
 * Get recordings by id
 *
 * @param  {String[]}       recordIds           The id of the recordings to get
 * @param  {String[]}       [fields]            The recording fields to select. If unspecified, selects all of them
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error that occurred, if any
 * @param  {Recording[]}   callback.recordings  The recording objects requested, in the same order as the record ids
 */
var getRecordingsById = module.exports.getRecordingsById = function(recordIds, fields, callback) {
    if (_.isEmpty(recordIds)) {
        return callback(null, []);
    }

    var query;
    var parameters = [];

    // If `fields` was specified, we select only the fields specified. Otherwise we select all (i.e., *)
    if (fields) {
        var columns = _.map(fields, function(field) {
            return util.format('"%s"', field);
        });

        query = util.format('SELECT %s FROM "Recordings" WHERE "id" IN (?)', columns.join(','));
    } else {
        query = 'SELECT * FROM "Recordings" WHERE "id" IN (?)';
    }

    parameters.push(recordIds);

    Cassandra.runQuery(query, parameters, function(err, rows) {
        if (err) {
            return callback(err);
        }

        // Convert the retrieved storage hashes into the Meeting model
        var recordings = {};
        _.chain(rows).map(Cassandra.rowToHash).each(function(row) {
            recordings[row.id] = _storageHashToRecording(row.id, row);
        });

        // Order the meetings according to the array of meeting ids
        var orderedRecordings = _.map(recordIds, function(recordId) {
            return recordings[recordId];
        });

        return callback(null, orderedRecordings);
    });
};

/**
 * Delete a recording by its id.
 * This will *NOT* remove the recording from libraries.
 *
 * @param  {String}     recordingId        The id of the recording to delete
 * @param  {Function}   callback           Standard callback function
 * @param  {Object}     callback.err       An error that occurred, if any
 */
var deleteRecording = module.exports.deleteRecording = function(recordingId, callback) {
    log().info({'recordingId': recordingId}, 'Recording deleted');
    Cassandra.runQuery('DELETE FROM "Recordings" WHERE "id" = ?', [recordingId], callback);
};
