// oae-discussions/lib/api.discussions.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');

var AuthzAPI = require('oae-authz');
var AuthzConstants = require('oae-authz/lib/constants').AuthzConstants;
var AuthzInvitations = require('oae-authz/lib/invitations');
var AuthzPermissions = require('oae-authz/lib/permissions');
var AuthzUtil = require('oae-authz/lib/util');
var LibraryAPI = require('oae-library');
var log = require('oae-logger').logger('discussions-api');
var MessageBoxAPI = require('oae-messagebox');
var MessageBoxConstants = require('oae-messagebox/lib/constants').MessageBoxConstants;
var OaeUtil = require('oae-util/lib/util');
var PrincipalsUtil = require('oae-principals/lib/util');
var PrincipalsDAO = require('oae-principals/lib/internal/dao');
var ResourceActions = require('oae-resource/lib/actions');
var Signature = require('oae-util/lib/signature');
var Validator = require('oae-authz/lib/validator').Validator;

var DiscussionsAPI = require('oae-discussions');
var DiscussionsConfig = require('oae-config').config('oae-discussions');
var DiscussionsConstants = require('./constants').DiscussionsConstants;
var DiscussionsDAO = require('./internal/dao');

// Discussion fields that are allowed to be updated
var DISCUSSION_UPDATE_FIELDS = ['displayName', 'description', 'visibility'];

/**
 * Create a new discussion
 *
 * @param  {Context}    ctx                 Standard context object containing the current user and the current tenant
 * @param  {String}     displayName         The display name of the discussion
 * @param  {String}     [description]       A longer description for the discussion
 * @param  {String}     [visibility]        The visibility of the discussion. One of public, loggedin, private. Defaults to the configured tenant default
 * @param  {Object}     [roles]             The initial membership of the discussion (the user in context will be a manager regardless of this parameter)
 * @param  {Object}     [opts]              Additional optional parameters
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Discussion} callback.discussion The discussion object that was created
 */
var createDiscussion = module.exports.createDiscussion = function(ctx, displayName, description, visibility, roles, opts, callback) {
    visibility = visibility || DiscussionsConfig.getValue(ctx.tenant().alias, 'visibility', 'discussion');
    roles = roles || {};
    opts = opts || {};

    var allVisibilities = _.values(AuthzConstants.visibility);

    // Verify basic properties
    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'Anonymous users cannot create a discussion'}).isLoggedInUser(ctx);
    validator.check(displayName, {'code': 400, 'msg': 'Must provide a display name for the discussion'}).notEmpty();
    validator.check(displayName, {'code': 400, 'msg': 'A display name can be at most 1000 characters long'}).isShortString();
    validator.check(description, {'code': 400, 'msg': 'Must provide a description for the discussion'}).notEmpty();
    validator.check(description, {'code': 400, 'msg': 'A description can be at most 10000 characters long'}).isMediumString();
    validator.check(visibility, {'code': 400, 'msg': 'An invalid discussion visibility option has been provided. Must be one of: ' + allVisibilities.join(', ')}).isIn(allVisibilities);

    // Verify each role is valid
    _.each(roles, function(role, memberId) {
        validator.check(role, {'code': 400, 'msg': 'The role: ' + role + ' is not a valid member role for a discussion'}).isIn(DiscussionsConstants.role.ALL_PRIORITY);
    });

    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // The current user is always a manager
    roles[ctx.user().id] = AuthzConstants.role.MANAGER;

    var createFn = _.partial(DiscussionsDAO.createDiscussion, ctx.user().id, displayName, description, visibility, null);
    ResourceActions.create(ctx, roles, createFn, function(err, discussion, memberChangeInfo) {
        if (err) {
            return callback(err);
        }

        DiscussionsAPI.emit(DiscussionsConstants.events.CREATED_DISCUSSION, ctx, discussion, memberChangeInfo, function(errs) {
            if (errs) {
                return callback(_.first(errs));
            }

            return callback(null, discussion);
        });
    });
};

/**
 * Update a discussion
 *
 * @param  {Context}    ctx                 Standard context object containing the current user and the current tenant
 * @param  {String}     discussionId        The id of the discussion to update
 * @param  {Object}     profileFields       An object whose keys are profile field names, and the value is the value to which you wish the field to change. Keys must be one of: displayName, visibility, discription
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Discussion} callback.discussion The updated discussion object
 */
var updateDiscussion = module.exports.updateDiscussion = function(ctx, discussionId, profileFields, callback) {
    var allVisibilities = _.values(AuthzConstants.visibility);

    var validator = new Validator();
    validator.check(discussionId, {'code': 400, 'msg': 'A discussion id must be provided'}).isResourceId();
    validator.check(null, {'code': 401, 'msg': 'You must be authenticated to update a discussion'}).isLoggedInUser(ctx);
    validator.check(_.keys(profileFields).length, {'code': 400, 'msg': 'You should at least one profile field to update'}).min(1);
    _.each(profileFields, function(value, field) {
        validator.check(field, {'code': 400, 'msg': 'The field \'' + field + '\' is not a valid field. Must be one of: ' + DISCUSSION_UPDATE_FIELDS.join(', ')}).isIn(DISCUSSION_UPDATE_FIELDS);
        if (field === 'visibility') {
            validator.check(value, {'code': 400, 'msg': 'An invalid visibility was specified. Must be one of: ' + allVisibilities.join(', ')}).isIn(allVisibilities);
        } else if (field === 'displayName') {
            validator.check(value, {'code': 400, 'msg': 'A display name cannot be empty'}).notEmpty();
            validator.check(value, {'code': 400, 'msg': 'A display name can be at most 1000 characters long'}).isShortString();
        } else if (field === 'description') {
            validator.check(value, {'code': 400, 'msg': 'A description cannot be empty'}).notEmpty();
            validator.check(value, {'code': 400, 'msg': 'A description can only be 10000 characters long'}).isMediumString();
        }
    });

    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        AuthzPermissions.canManage(ctx, discussion, function(err) {
            if (err) {
                return callback(err);
            }

            DiscussionsDAO.updateDiscussion(discussion, profileFields, function(err, updatedDiscussion) {
                if (err) {
                    return callback(err);
                }

                // Fill in the full profile, the user has to have been a manager, so these are all true
                updatedDiscussion.isManager = true;
                updatedDiscussion.canPost = true;
                updatedDiscussion.canShare = true;

                DiscussionsAPI.emit(DiscussionsConstants.events.UPDATED_DISCUSSION, ctx, updatedDiscussion, discussion, function(errs) {
                    if (errs) {
                        return callback(_.first(errs));
                    }

                    return callback(null, updatedDiscussion);
                });
            });
        });
    });
};

/**
 * Deletes the specified discussion.
 *
 * @param  {Context}    ctx                 Standard context object containing the current user and the current tenant
 * @param  {String}     discussionId        The id of the discussion to delete
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occurred, if any
 */
var deleteDiscussion = module.exports.deleteDiscussion = function(ctx, discussionId, callback) {
    var validator = new Validator();
    validator.check(discussionId, {'code': 400, 'msg': 'A discussion id must be provided'}).isResourceId();
    validator.check(null, {'code': 401, 'msg': 'You must be authenticated to delete a discussion'}).isLoggedInUser(ctx);
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        AuthzPermissions.canManage(ctx, discussion, function(err) {
            if (err) {
                return callback(err);
            }

            // TODO: Use a "mark as deleted" approach for deleting discussions instead of hard
            // deleting it from the database. This approach wouldn't require accessing and deleting
            // all authz member roles from the discussion
            AuthzAPI.getAllAuthzMembers(discussion.id, function(err, memberIdRoles) {
                if (err) {
                    return callback(err);
                }

                var roleChanges = {};
                var removedMemberIds = _.pluck(memberIdRoles, 'id');
                _.each(removedMemberIds, function(memberId) {
                    roleChanges[memberId] = false;
                });

                // Update the authz associations
                AuthzAPI.updateRoles(discussion.id, roleChanges, function(err) {
                    if (err) {
                        return callback(err);
                    }

                    // Remove the actual discussion profile
                    DiscussionsDAO.deleteDiscussion(discussion.id, function(err) {
                        if (err) {
                            return callback(err);
                        }

                        DiscussionsAPI.emit(DiscussionsConstants.events.DELETED_DISCUSSION, ctx, discussion, removedMemberIds, function(errs) {
                            if (errs) {
                                return callback(_.first(errs));
                            }

                            return callback();
                        });
                    });
                });
            });
        });
    });
};

/**
 * Get the discussions library items for a user or group. Depending on the access of the principal in context,
 * either a library of public, loggedin, or all items will be returned.
 *
 * @param  {Context}        ctx                     Standard context object containing the current user and the current tenant
 * @param  {String}         principalId             The id of the principal whose discussion library to fetch
 * @param  {String}         [start]                 The discussion ordering token from which to start fetching discussions (see `nextToken` in callback params)
 * @param  {Number}         [limit]                 The maximum number of results to return. Default: 10
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            An error that occurred, if any
 * @param  {Discussion[]}   callback.discussions    The array of discussions fetched
 * @param  {String}         [callback.nextToken]    The token that can be used as the `start` parameter to fetch the next set of tokens (exclusively). If not specified, indicates that the query fetched all remaining results.
 */
var getDiscussionsLibrary = module.exports.getDiscussionsLibrary = function(ctx, principalId, start, limit, callback) {
    limit = OaeUtil.getNumberParam(limit, 10, 1);

    var validator = new Validator();
    validator.check(principalId, {'code': 400, 'msg': 'A user or group id must be provided'}).isPrincipalId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Get the principal
    PrincipalsDAO.getPrincipal(principalId, function(err, principal) {
        if (err) {
            return callback(err);
        }

        // Determine which library visibility the current user should receive
        LibraryAPI.Authz.resolveTargetLibraryAccess(ctx, principal.id, principal, function(err, hasAccess, visibility) {
            if (err) {
                return callback(err);
            } else if (!hasAccess) {
                return callback({'code': 401, 'msg': 'You do not have have access to this library'});
            }

            // Get the discussion ids from the library index
            LibraryAPI.Index.list(DiscussionsConstants.library.DISCUSSIONS_LIBRARY_INDEX_NAME, principalId, visibility, {'start': start, 'limit': limit}, function(err, entries, nextToken) {
                if (err) {
                    return callback(err);
                }

                // Get the discussion objects from the discussion ids
                var discussionIds = _.pluck(entries, 'resourceId');
                DiscussionsDAO.getDiscussionsById(discussionIds, null, function(err, discussions) {
                    if (err) {
                        return callback(err);
                    }

                    // Emit an event indicating that the discussion library has been retrieved
                    DiscussionsAPI.emit(DiscussionsConstants.events.GET_DISCUSSION_LIBRARY, ctx, principalId, visibility, start, limit, discussions);

                    return callback(null, discussions, nextToken);
                });
            });
        });
    });
};

/**
 * Get a discussion basic profile by its id.
 *
 * @param  {Context}    ctx                 Standard context object containing the current user and the current tenant
 * @param  {String}     discussionId        The id of the discussion to get
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Discussion} callback.discussion The discussion object requested
 */
var getDiscussion = module.exports.getDiscussion = function(ctx, discussionId, callback) {
    var validator = new Validator();
    validator.check(discussionId, {'code': 400, 'msg': 'discussionId must be a valid resource id'}).isResourceId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        AuthzPermissions.canView(ctx, discussion, function(err) {
            if (err) {
                return callback(err);
            }

            return callback(null, discussion);
        });
    });
};

/**
 * Get a full discussion profile. In addition to the basic profile, the full profile contains
 * the basic profile of the creator, and access information (see parameters)
 *
 * @param  {Context}    ctx                             Standard context object containing the current user and the current tenant
 * @param  {String}     discussionId                    The id of the discussion to get
 * @param  {Function}   callback                        Standard callback function
 * @param  {Object}     callback.err                    An error that occurred, if any
 * @param  {Discussion} callback.discussion             The discussion object requested
 * @param  {User}       callback.discussion.createdBy   The basic profile of the user who created the discussion
 * @param  {Boolean}    callback.discussion.isManager   Specifies if the current user in context is a manager of the discussion
 * @param  {Boolean}    callback.discussion.canShare    Specifies if the current user in context is allowed to share the discussion
 * @param  {Boolean}    callback.discussion.canPost     Specifies if the current user in context is allowed to post messages to the discussion
 */
var getFullDiscussionProfile = module.exports.getFullDiscussionProfile = function(ctx, discussionId, callback) {
    var validator = new Validator();
    validator.check(discussionId, {'code': 400, 'msg': 'discussionId must be a valid resource id'}).isResourceId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Get the discussion object, throwing an error if it does not exist but does not do permission checks
    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        // Resolve the full discussion access information for the current user
        AuthzPermissions.resolveEffectivePermissions(ctx, discussion, function(err, permissions) {
            if (err) {
                return callback(err);
            } else if (!permissions.canView) {
                // The user has no effective role, which means they are not allowed to view (this has already taken into
                // consideration implicit privacy rules, such as whether or not the discussion is public).
                return callback({'code': 401, 'msg': 'You are not authorized to view this discussion'});
            }

            discussion.isManager = permissions.canManage;
            discussion.canShare = permissions.canShare;
            discussion.canPost = permissions.canInteract;

            if (ctx.user()) {
                // Attach a signature that can be used to perform quick access checks
                discussion.signature = Signature.createExpiringResourceSignature(ctx, discussionId);
            }

            // Populate the creator of the discussion
            PrincipalsUtil.getPrincipal(ctx, discussion.createdBy, function(err, creator) {
                if (err) {
                    log().warn({
                        'err': err,
                        'userId': discussion.createdBy,
                        'discussionId': discussion.id
                    }, 'An error occurred getting the creator of a discussion. Proceeding with empty user for full profile');
                } else if (creator) {
                    discussion.createdBy = creator;
                }

                DiscussionsAPI.emit(DiscussionsConstants.events.GET_DISCUSSION_PROFILE, ctx, discussion);
                return callback(null, discussion);
            });
        });
    });
};

/**
 * Get the members of a discussion and their roles
 *
 * @param  {Context}        ctx                             Standard context object containing the current user and the current tenant
 * @param  {String}         discussionId                    The id of the discussion to get the members for
 * @param  {String}         [start]                         The id of the principal from which to begin the page of results (exclusively). By default, begins from the first in the list
 * @param  {Number}         [limit]                         The maximum number of results to return. Default: 10
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Object[]}       callback.members                Array that contains an object for each member
 * @param  {String}         callback.members[i].role        The role of the member at index `i`
 * @param  {User|Group}     callback.members[i].profile     The principal profile of the member at index `i`
 * @param  {String}         callback.nextToken              The value to provide in the `start` parameter to get the next set of results
 */
var getDiscussionMembers = module.exports.getDiscussionMembers = function(ctx, discussionId, start, limit, callback) {
    limit = OaeUtil.getNumberParam(limit, 10, 1);

    var validator = new Validator();
    validator.check(discussionId, {'code': 400, 'msg': 'A discussion id must be provided'}).isResourceId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    getDiscussion(ctx, discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        // Get the discussion members
        AuthzAPI.getAuthzMembers(discussionId, start, limit, function(err, memberRoles, nextToken) {
            if (err) {
                return callback(err);
            }

            // Get the basic profiles for all of these principals
            var memberIds = _.pluck(memberRoles, 'id');
            PrincipalsUtil.getPrincipals(ctx, memberIds, function(err, memberProfiles) {
                if (err) {
                    return callback(err);
                }

                // Merge the member profiles and roles into a single object
                var memberList = _.map(memberRoles, function(memberRole) {
                    return {
                        'profile': memberProfiles[memberRole.id],
                        'role': memberRole.role
                    };
                });

                return callback(null, memberList, nextToken);
            });
        });
    });
};

/**
 * Get the invitations for the specified discussion
 *
 * @param  {Context}        ctx                     Standard context object containing the current user and the current tenant
 * @param  {String}         discussionId            The id of the discussion to get the invitations for
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            An error that occurred, if any
 * @param  {Invitation[]}   callback.invitations    The invitations
 */
var getDiscussionInvitations = module.exports.getDiscussionInvitations = function(ctx, discussionId, callback) {
    var validator = new Validator();
    validator.check(discussionId, {'code': 400, 'msg': 'A valid resource id must be specified'}).isResourceId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        return AuthzInvitations.getAllInvitations(ctx, discussion, callback);
    });
};

/**
 * Resend an invitation email for the specified email and discussion
 *
 * @param  {Context}        ctx             Standard context object containing the current user and the current tenant
 * @param  {String}         discussionId    The id of the discussion to which the email was invited
 * @param  {String}         email           The email that was previously invited
 * @param  {Function}       callback        Standard callback function
 * @param  {Object}         callback.err    An error that occurred, if any
 */
var resendDiscussionInvitation = module.exports.resendDiscussionInvitation = function(ctx, discussionId, email, callback) {
    var validator = new Validator();
    validator.check(discussionId, {'code': 400, 'msg': 'A valid resource id must be specified'}).isResourceId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        return ResourceActions.resendInvitation(ctx, discussion, email, callback);
    });
};

/**
 * Share a discussion with a number of users and groups. The role of the target principals will be `member`. If
 * any principals in the list already have the discussion in their library, then this will have no impact for
 * that user with no error. Only those who do not have the discussion in their library will be impacted.
 *
 * @param  {Context}    ctx                 Standard context object containing the current user and the current tenant
 * @param  {String}     discussionId        The id of the discussion to share
 * @param  {String[]}   principalIds        The ids of the principals with which the discussion will be shared
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occurred, if any
 */
var shareDiscussion = module.exports.shareDiscussion = function(ctx, discussionId, principalIds, callback) {
    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'You have to be logged in to be able to share a discussion'}).isLoggedInUser(ctx);
    validator.check(discussionId, {'code': 400, 'msg': 'A valid discussion id must be provided'}).isResourceId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        ResourceActions.share(ctx, discussion, principalIds, AuthzConstants.role.MEMBER, function(err, memberChangeInfo) {
            if (err) {
                return callback(err);
            } else if (_.isEmpty(memberChangeInfo.changes)) {
                return callback();
            }

            DiscussionsAPI.emit(DiscussionsConstants.events.UPDATED_DISCUSSION_MEMBERS, ctx, discussion, memberChangeInfo, {}, function(errs) {
                if (errs) {
                    return callback(_.first(errs));
                }

                return callback();
            });
        });
    });
};

/**
 * Set the permissions of a discussion. This method will ensure that the current user in context has access to change the
 * permissions, as well as ensure the discussion does not end up with no manager members.
 *
 * @param  {Context}    ctx                     Standard context object containing the current user and the current tenant
 * @param  {String}     discussionId            The id of the discussion to share
 * @param  {Object}     changes                 An object that describes the permission changes to apply to the discussion. The key is the id of the principal to which to apply the change, and the value is the role to apply to the principal. If the value is `false`, the principal will be revoked access.
 * @param  {Function}   callback                Standard callback function
 * @param  {Object}     callback.err            An error that occurred, if any
 * @param  {Object}     callback.permissions    An object describing the permissions of the discussion after the change is applied. The key is the principal id and the value is the role that the principal has on the discussion
 */
var setDiscussionPermissions = module.exports.setDiscussionPermissions = function(ctx, discussionId, changes, callback) {
    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'You have to be logged in to be able to change discussion permissions'}).isLoggedInUser(ctx);
    validator.check(discussionId, {'code': 400, 'msg': 'A valid discussion id must be provided'}).isResourceId();
    _.each(changes, function(role, principalId) {
        validator.check(role, {'code': 400, 'msg': 'The role change: ' + role + ' is not a valid value. Must either be a string, or false'}).isValidRoleChange();
        if (role) {
            validator.check(role, {'code': 400, 'msg': 'The role: "' + role + '" is not a valid value. Must be one of: ' + DiscussionsConstants.role.ALL_PRIORITY.join(', ') + '; or false'}).isIn(DiscussionsConstants.role.ALL_PRIORITY);
        }
    });

    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Get the discussion object, throwing an error if it doesn't exist, but not applying permissions checks
    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        ResourceActions.setRoles(ctx, discussion, changes, function(err, memberChangeInfo) {
            if (err) {
                return callback(err);
            } else if (_.isEmpty(memberChangeInfo.changes)) {
                return callback();
            }

            DiscussionsAPI.emit(DiscussionsConstants.events.UPDATED_DISCUSSION_MEMBERS, ctx, discussion, memberChangeInfo, {}, function(errs) {
                if (errs) {
                    return callback(_.first(errs));
                }

                return callback(null, memberChangeInfo.roles.after);
            });
        });
    });
};

/**
 * Remove a discussion from a discussion library. This is its own API method due to special permission handling required, as the user
 * is effectively updating a discussions permissions (removing themselves, or removing it from a group they manage), and they might not
 * necessarily have access to update the permissions of the private discussion (e.g., they are only a member). Also, tenant privacy
 * rules do not come into play in this case.
 *
 * @param  {Context}    ctx             Standard context object containing the current user and the current tenant
 * @param  {String}     libraryOwnerId  The owner of the library, should be a principal id (either user or group id)
 * @param  {String}     discussionId    The id of the discussion to remove from the library
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occurred, if any
 */
var removeDiscussionFromLibrary = module.exports.removeDiscussionFromLibrary = function(ctx, libraryOwnerId, discussionId, callback) {
    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'You must be authenticated to remove a discussion from a library'}).isLoggedInUser(ctx);
    validator.check(libraryOwnerId, {'code': 400, 'msg': 'A user or group id must be provided'}).isPrincipalId();
    validator.check(discussionId, {'code': 400, 'msg': 'An invalid discussion id "' + discussionId + '" was provided'}).isResourceId();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Make sure the discussion exists
    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        PrincipalsDAO.getPrincipal(libraryOwnerId, function(err, principal) {
            if (err) {
                return callback(err);
            }

            AuthzPermissions.canRemoveRole(ctx, principal, discussion, function(err, memberChangeInfo) {
                if (err) {
                    return callback(err);
                }

                // All validation checks have passed, finally persist the role change and update the user library
                AuthzAPI.updateRoles(discussionId, memberChangeInfo.changes, function(err) {
                    if (err) {
                        return callback(err);
                    }

                    DiscussionsAPI.emit(DiscussionsConstants.events.UPDATED_DISCUSSION_MEMBERS, ctx, discussion, memberChangeInfo, {}, function(errs) {
                        if (errs) {
                            return callback(_.first(errs));
                        }

                        return callback();
                    });
                });
            });
        });
    });
};

/**
 * Create a new message in a discussion. If `replyToCreatedTimestamp` is specified, the message will be
 * a reply to the message in the discussion identified by that timestamp.
 *
 * @param  {Context}        ctx                         Standard context object containing the current user and the current tenant
 * @param  {String}         discussionId                The id of the discussion to which to post the message
 * @param  {String}         body                        The body of the message
 * @param  {String|Number}  [replyToCreatedTimestamp]   The timestamp of the message to which this message is a reply. Not specifying this will create a top level comment
 * @param  {Function}       callback                    Standard callback function
 * @param  {Object}         callback.err                An error that occurred, if any
 * @param  {Message}        callback.message            The created message
 */
var createMessage = module.exports.createMessage = function(ctx, discussionId, body, replyToCreatedTimestamp, callback) {
    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'Only authenticated users can post on discussions'}).isLoggedInUser(ctx);
    validator.check(discussionId, {'code': 400, 'msg': 'Invalid discussion id provided'}).isResourceId();
    validator.check(body, {'code': 400, 'msg': 'A discussion body must be provided'}).notEmpty();
    validator.check(body, {'code': 400, 'msg': 'A discussion body can only be 100000 characters long'}).isLongString();
    if (replyToCreatedTimestamp) {
        validator.check(replyToCreatedTimestamp, {'code': 400, 'msg': 'Invalid reply-to timestamp provided'}).isInt();
    }

    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Get the discussion, throwing an error if it doesn't exist, avoiding permission checks for now
    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        // Determine if the current user can post discussion messages to this discussion
        AuthzPermissions.canInteract(ctx, discussion, function(err) {
            if (err) {
                return callback(err);
            }

            // Create the message
            MessageBoxAPI.createMessage(discussionId, ctx.user().id, body, {'replyToCreated': replyToCreatedTimestamp}, function(err, message) {
                if (err) {
                    return callback(err);
                }

                // Get a UI-appropriate representation of the current user
                PrincipalsUtil.getPrincipal(ctx, ctx.user().id, function(err, createdBy) {
                    if (err) {
                        return callback(err);
                    }

                    message.createdBy = createdBy;

                    // The message has been created in the database so we can emit the `created-message` event
                    DiscussionsAPI.emit(DiscussionsConstants.events.CREATED_DISCUSSION_MESSAGE, ctx, message, discussion, function(errs) {
                        if (errs) {
                            return callback(_.first(errs));
                        }

                        return callback(null, message);
                    });
                });
            });
        });
    });
};

/**
 * Delete a message in a discussion. Managers of the discussion can delete all messages while people that have access
 * to the discussion can only delete their own messages. Therefore, anonymous users will never be able to delete messages.
 *
 * @param  {Context}    ctx                     Standard context object containing the current user and the current tenant
 * @param  {String}     discussionId            The id of the discussion from which to delete the message
 * @param  {Number}     messageCreatedDate      The timestamp of the message that should be deleted
 * @param  {Function}   callback                Standard callback function
 * @param  {Object}     callback.err            An error that occurred, if any
 * @param  {Comment}    [callback.softDeleted]  When the message has been soft deleted (because it has replies), a stripped down message object representing the deleted message will be returned, with the `deleted` parameter set to `false`. If the message has been deleted from the index, no message object will be returned
 */
var deleteMessage = module.exports.deleteMessage = function(ctx, discussionId, messageCreatedDate, callback) {
    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'Only authenticated users can delete messages'}).isLoggedInUser(ctx);
    validator.check(discussionId, {'code': 400, 'msg': 'A discussion id must be provided'}).isResourceId();
    validator.check(messageCreatedDate, {'code': 400, 'msg': 'A valid integer message created timestamp must be specified'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Get the discussion without permissions check
    _getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        // Ensure that the message exists. We also need it so we can make sure we have access to deleted it
        MessageBoxAPI.getMessages(discussionId, [messageCreatedDate], {'scrubDeleted': false}, function(err, messages) {
            if (err) {
                return callback(err);
            } else if (!messages[0]) {
                return callback({'code': 404, 'msg': 'The specified message does not exist'});
            }

            var message = messages[0];

            // Determine if we have access to delete the discussion message
            AuthzPermissions.canManageMessage(ctx, discussion, message, function(err) {
                if (err) {
                    return callback(err);
                }

                // Delete the message using the "leaf" method, which will SOFT delete if the message has replies, or HARD delete if it does not
                MessageBoxAPI.deleteMessage(discussionId, messageCreatedDate, {'deleteType': MessageBoxConstants.deleteTypes.LEAF}, function(err, deleteType, deletedMessage) {
                    if (err) {
                        return callback(err);
                    }

                    DiscussionsAPI.emit(DiscussionsConstants.events.DELETED_DISCUSSION_MESSAGE, ctx, message, discussion, deleteType);

                    // If a soft-delete occurred, we want to inform the consumer of the soft-delete message model
                    if (deleteType === MessageBoxConstants.deleteTypes.SOFT) {
                        return callback(null, deletedMessage);
                    } else {
                        return callback();
                    }
                });
            });
        });
    });
};

/**
 * Get the messages in a discussion
 *
 * @param  {Context}        ctx                     Standard context object containing the current user and the current tenant
 * @param  {String}         discussionId            The id of the discussion for which to get the messages
 * @param  {String}         [start]                 The `threadKey` of the message from which to start retrieving messages (exclusively). By default, will start fetching from the most recent message
 * @param  {Number}         [limit]                 The maximum number of results to return. Default: 10
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            An error that occurred, if any
 * @param  {Message[]}      callback.messages       The messages in the discussion. Of the type `MessageBoxModel#Message`
 * @param  {String}         callback.nextToken      The value to provide in the `start` parameter to get the next set of results
 */
var getMessages = module.exports.getMessages = function(ctx, discussionId, start, limit, callback) {
    limit = OaeUtil.getNumberParam(limit, 10, 1);

    var validator = new Validator();
    validator.check(discussionId, {'code': 400, 'msg': 'Must provide a valid discussion id'}).isResourceId();
    validator.check(limit, {'code': 400, 'msg': 'Must provide a valid limit'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Get the discussion, throwing an error if the user in context doesn't have view access or if it doesn't exist
    getDiscussion(ctx, discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        }

        // Fetch the messages from the message box
        MessageBoxAPI.getMessagesFromMessageBox(discussionId, start, limit, null, function(err, messages, nextToken) {
            if (err) {
                return callback(err);
            }

            var userIds = _.map(messages, function(message) {
                return message.createdBy;
            });

            // Remove falsey and duplicate userIds
            userIds = _.uniq(_.compact(userIds));

            // Get the basic principal profiles of the messagers to add to the messages as `createdBy`.
            PrincipalsUtil.getPrincipals(ctx, userIds, function(err, users) {
                if (err) {
                    return callback(err);
                }

                // Attach the user profiles to the message objects
                _.each(messages, function(message) {
                    if (users[message.createdBy]) {
                        message.createdBy = users[message.createdBy];
                    }
                });

                return callback(err, messages, nextToken);
            });
        });
    });
};

/**
 * Get the discussion with the specified id. If it doesn't exist, a 404 error will be thrown. No permission checks
 * will be performed.
 *
 * @param  {Context}    ctx                 Standard context object containing the current user and the current tenant
 * @param  {String}     discussionId        The id of the discussion to get
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Discussion} callback.discussion The discussion object requested
 * @api private
 */
var _getDiscussion = function(discussionId, callback) {
    DiscussionsDAO.getDiscussion(discussionId, function(err, discussion) {
        if (err) {
            return callback(err);
        } else if (!discussion) {
            return callback({'code': 404, 'msg': 'Could not find discussion: ' + discussionId});
        }

        return callback(null, discussion);
    });
};
