// oae-discussions/lib/api.authz.js
/*!
 * Copyright 2012 Sakai Foundation (SF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://www.osedu.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');

var AuthzAPI = require('oae-authz');
var AuthzUtil = require('oae-authz/lib/util');
var LibraryConstants = require('oae-library/lib/constants').LibraryConstants;
var PrincipalsUtil = require('oae-principals/lib/util');
var TenantsUtil = require('oae-tenants/lib/util');

var DiscussionsConstants = require('./constants').DiscussionsConstants;

/**
 * Determine whether or not the user in context can view the provided discussion.
 *
 * @param  {Context}    ctx                 The context of the current request
 * @param  {Discussion} discussion          The discussion to test for access
 * @param  {Function}   callback            Invoked when the process completes
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Boolean}    callback.canView    `true` if the user in context has the appropriate permission. `false` otherwise.
 */
var canViewDiscussion = module.exports.canViewDiscussion = function(ctx, discussion, callback) {
    var user = ctx.user();
    _resolveImplicitRole(ctx, discussion, function(err, implicitRole, canInteract) {
        if (err) {
            return callback(err);
        } else if (implicitRole) {
            // We have an implicit access, no reason to try and find an explicit access because we can atleast view
            return callback(null, true);
        } else if (!user) {
            // Anonymous user with no implicit access cannot view
            return callback(null, false);
        }

        // By this point, we only have access to view if we have a role on the item
        AuthzAPI.hasAnyRole(user.id, discussion.id, callback);
    });
};

/**
 * Determine whether or not the user in context can manage the provided discussion.
 *
 * @param  {Context}    ctx                 The context of the current request
 * @param  {Discussion} discussion          The discussion to test for access
 * @param  {Function}   callback            Invoked when the process completes
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Boolean}    callback.canManage  `true` if the user in context has the appropriate permission. `false` otherwise.
 */
var canManageDiscussion = module.exports.canManageDiscussion = function(ctx, discussion, callback) {
    var user = ctx.user();

    // Anonymous can never manage
    if (!user) {
        return callback(null, false);
    }

    _resolveImplicitRole(ctx, discussion, function(err, implicitRole, canInteract) {
        if (err) {
            return callback(err);
        } else if (implicitRole === DiscussionsConstants.roles.MANAGER) {
            // We have an implicit management role (e.g., we are a manager), return true
            return callback(null, true);
        }

        // By this point, we can only manage if we have explicit manager role
        AuthzAPI.hasRole(user.id, discussion.id, DiscussionsConstants.roles.MANAGER, callback);
    });
};

/**
 * Determine whether or not the user in context can share the provided discussion with the principalIds
 * provided. `canShare` can be `false` either if the user doesn't have appopriate permission to share the
 * discussion, or if sharing with some target principals violates tenant boundaries.
 *
 * @param  {Context}    ctx                             The context of the current request
 * @param  {Discussion} discussion                      The discussion to test for access
 * @param  {String[]}   principalIds                    The principalIds with which the user wishes to share the discussion
 * @param  {Function}   callback                        Invoked when the process completes
 * @param  {Object}     callback.err                    An error that occurred, if any
 * @param  {Boolean}    callback.canShare               `true` if the u ser in context is allowed to perform this share operation. `false` otherwise
 * @param  {String[]}   [callback.illegalPrincipalIds]  If the check failed because of a tenant boundary violation, this will be an array of principalIds that were in violation
 */
var canShareDiscussion = module.exports.canShareDiscussion = function(ctx, discussion, principalIds, callback) {
    var user = ctx.user();

    // Anonymous users can never share
    if (!user) {
        return callback(null, false);
    }

    // Verify tenant boundaries
    // Verify tenant boundaries
    _resolveTenantBoundaryViolations(ctx, discussion, principalIds, function(err, canShare, illegalPrincipalIds) {
        if (err) {
            return callback(err);
        } else if (illegalPrincipalIds && illegalPrincipalIds.length > 0) {
            return callback(null, false, illegalPrincipalIds);
        }

        _resolveImplicitRole(ctx, discussion, function(err, implicitRole, canInteract) {
            if (err) {
                return callback(err);
            } else if (implicitRole === DiscussionsConstants.roles.MANAGER) {
                // Managers can always share
                return callback(null, true);
            } else if (canInteract) {
                // If we can interact with the item, we can always share it
                return callback(null, true);
            }

            // If the discussion is private, only managers can share it
            if (discussion.visibility === LibraryConstants.visibility.PRIVATE) {
                return AuthzAPI.hasRole(user.id, discussion.id, DiscussionsConstants.roles.MANAGER, function(err, hasRole) {
                    return callback(err, hasRole);
                });
            }

            // At this point, we have to see if the user has any explicit role on the resource to see if they can share it
            AuthzAPI.hasAnyRole(user.id, discussion.id, function(err, hasAnyRole) {
                return callback(err, hasAnyRole);
            });
        });
    });
};

/**
 * Checks for tenant boundary violations between:
 *   * the discussion item and any of the principalIds
 *   * the current user and any of the principalIds
 *
 * The visibility settings for each user will also be checked in case they
 * are external users.
 *
 * @param  {Context}    ctx                             The context of the current request
 * @param  {Discussion} discussion                      The discussion to test for access
 * @param  {String[]}   principalIds                    The principalIds with which the user wishes to share the discussion
 * @param  {Function}   callback                        Invoked when the process completes
 * @param  {Object}     callback.err                    An error that occurred, if any
 * @param  {Boolean}    callback.canShare               `true` if the user in context is allowed to perform this share operation. `false` otherwise
 * @param  {String[]}   [callback.illegalPrincipalIds]  If the check failed because of a tenant boundary violation, this will be an array of principalIds that were in violation
 * @api private
 */
var _resolveTenantBoundaryViolations = function(ctx, discussion, principalIds, callback) {
    var user = ctx.user();
    // First test the easy stuff:
    //  * A direct violation between discussion.tenant and target.tenant
    //  * A direct violation between actor.tenant and target.tenant
    var illegalPrincipalIds = TenantsUtil.resolveTenantBoundaryViolations(user, discussion.tenant.alias, principalIds);
    if (illegalPrincipalIds.length > 0) {
        return callback(null, false, illegalPrincipalIds);
    }

    // Now, if any of the specified principalIds are part of an external tenant, we'll need to check their visibility.
    var externalPrincipals = _.filter(principalIds, function(principalId) {
        var targetTenant = AuthzUtil.getResourceFromId(principalId).tenantAlias;
        return targetTenant !== ctx.tenant().alias;
    });
    if (externalPrincipals.length > 0) {
        PrincipalsUtil.getPrincipals(ctx, externalPrincipals, function(err, principals) {
            if (err) {
                return callback(err);
            }

            var nonPublicUsers = _.filter(principals, function(principal) { return principal.visibility !== 'public'; });
            var canShare = (nonPublicUsers.length === 0);
            var illegalPrincipalIds = (canShare) ? null : _.map(nonPublicUsers, function(user) { return user.id; });
            callback(null, canShare, illegalPrincipalIds);
        });
    } else {
        // Otherwise we're good
        callback(null, true);
    }
};

/**
 * Determine whether or not the user in context can update the permissions of the provided discussion. Since tenant boundaries
 * allow that existing members of discussions can have their membership changed (even if their tenant has since become private)
 * and members may be removed even if they cross private tenant boundaries, the `addMemberIds` array should only contain ids of
 * members who don't already exist as members on the discussion.
 *
 * @param  {Context}    ctx                         The context of the current request
 * @param  {Discussion} discussion                  The discussion to test for access
 * @param  {String[]}   addMemberIds                An array of ids of members that are being **added** to the discussion
 * @param  {Function}   callback                    Invoked when the process completes
 * @param  {Object}     callback.err                An error that occurred, if any
 * @param  {Boolean}    callback.canSetPermissions  `true` if the user in context has the appropriate permission. `false` otherwise.
 */
var canSetDiscussionPermissions = module.exports.canSetDiscussionPermissions = function(ctx, discussion, addMembers, callback) {
    // Only managers can set the permissions of any discussion, regardless of privacy rules
    if (!ctx.user()) {
        return callback({'code': 401, 'msg': 'You must be authenticated to update permissions of a discussion'});
    }

    // Reject the operation if it will violate tenant privacy boundaries
    var illegalMemberIds = TenantsUtil.resolveTenantBoundaryViolations(ctx.user(), discussion.tenant.alias, addMembers);
    if (illegalMemberIds.length > 0) {
        return callback(null, false, illegalMemberIds);
    }

    // Boundaries are all good, perform the manager check
    canManageDiscussion(ctx, discussion, callback);
};

/**
 * Determine whether or not the user in context can post a message in the provided discussion.
 *
 * @param  {Context}    ctx                 The context of the current request
 * @param  {Discussion} discussion          The discussion to test for access
 * @param  {Function}   callback            Invoked when the process completes
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Boolean}    callback.canPost    `true` if the user in context has the appropriate permission. `false` otherwise.
 */
var canPostDiscussionMessage = module.exports.canPostDiscussionMessage = function(ctx, discussion, callback) {
    var user = ctx.user();

    if (!user) {
        // Anonymous can never post to a discussion
        return callback(null, false);
    }

    _resolveImplicitRole(ctx, discussion, function(err, implicitRole, canInteract) {
        if (err) {
            return callback(err);
        } else if (implicitRole === DiscussionsConstants.roles.MANAGER) {
            // Implicit managers can always post, regardless of tenant
            return callback(null, true);
        } else if (canInteract) {
            // If we can interact with the item, we can always post to it
            return callback(null, true);
        }

        // See if this user has any explicit role, if so they can post
        AuthzAPI.hasAnyRole(user.id, discussion.id, callback);
    });
};

/**
 * Determine whether or not the user in context can delete the provided message in the provided discussion.
 *
 * @param  {Context}    ctx                 The context of the current request
 * @param  {Discussion} discussion          The discussion to test for access
 * @param  {Message}    message             The message to test for access
 * @param  {Function}   callback            Invoked when the process completes
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Boolean}    callback.canDelete  `true` if the user in context has the appropriate permission. `false` otherwise.
 */
var canDeleteDiscussionMessage = module.exports.canDeleteDiscussionMessage = function(ctx, discussion, message, callback) {
    var user = ctx.user();

    if (!user) {
        return callback(null, false);
    }

    _resolveImplicitRole(ctx, discussion, function(err, implicitRole, canInteract) {
        if (err) {
            return callback(err);
        } else if (implicitRole === DiscussionsConstants.roles.MANAGER) {
            // Managers can always delete messages
            return callback(null, true);
        } else if (canInteract && message.createdBy === user.id) {
            // The user is still able to interact with this tenant, they can delete their own message
            return callback(null, true);
        }

        // It's not our message, or we cannot implicitly interact. We need explicit manager access to delete the message
        AuthzAPI.hasRole(user.id, discussion.id, DiscussionsConstants.roles.MANAGER, callback);
    });
};

/**
 * Determine the full set of access that the user in context has on the discussion
 *
 * @param  {Context}    ctx                 The context of the current request
 * @param  {Discussion} discussion          The discussion to test for access
 * @param  {Function}   callback            Invoked when the process completes
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Boolean}    callback.canView    `true` if the user in context can view the discussion. `false` otherwise.
 * @param  {Boolean}    callback.canManage  `true` if the user in context can manage the discussion. `false` otherwise.
 * @param  {Boolean}    callback.canShare   `true` if the user in context can share the discussion. `false` otherwise.
 * @param  {Boolean}    callback.canPost    `true` if the user in context can post a message in the discussion. `false` otherwise.
 */
var resolveEffectiveDiscussionAccess = module.exports.resolveEffectiveDiscussionAccess = function(ctx, discussion, callback) {
    _resolveEffectiveDiscussionRole(ctx, discussion, function(err, role, canInteract) {
        if (err) {
            return callback(err);
        }

        var canView = _.isString(role);
        var canManage = (role === DiscussionsConstants.roles.MANAGER);

        // Anybody who can "interact" with the discussion can post to it
        var canPost = canInteract;

        // Anyone who can interact can share, unless the discussion is private. In that case, only managers can share
        var canShare = canInteract;
        if (discussion.visibility !== LibraryConstants.visibility.PUBLIC && discussion.visibility !== LibraryConstants.visibility.LOGGEDIN) {
            canShare = canManage;
        }

        return callback(null, canView, canManage, canShare, canPost);
    });
};

/**
 * Determine the **effective** role a user has in a discussion. Though a user can have multiple roles on a discussion
 * by virtue of indirect group membership, this determines the highest level of access granted. This check is not only
 * implicit, but includes explicit role membership lookup. Therefore, its output alone can be used as an authoritative
 * source of access information.
 *
 * @param  {Context}    ctx                     The context of the user to check for access
 * @param  {Discussion} discussion              The discussion to check against for access
 * @param  {Function}   callback                Invoked when the process completes
 * @param  {Object}     callback.err            An error that occured, if any
 * @param  {String}     callback.effectiveRole  The effective role of the user in context
 * @param  {Boolean}    callback.canInteract    Whether or not the user can interact
 * @api private
 */
var _resolveEffectiveDiscussionRole = function(ctx, discussion, callback) {
    var user = ctx.user();
    _resolveImplicitRole(ctx, discussion, function(err, implicitRole, canInteract) {
        if (err) {
            return callback(err);
        } else if (implicitRole === DiscussionsConstants.roles.MANAGER) {
            // We already have the highest role, use it
            return callback(null, implicitRole, true);
        } else if (!user) {
            // We are anonymous so cannot have any explicit access or interact. Use only our implicitRole if we have one
            return callback(null, implicitRole, false);
        }

        AuthzAPI.getAllRoles(user.id, discussion.id, function(err, roles) {
            if (err) {
                return callback(err);
            } else if (roles.length === 0) {
                // We have no explicit role, so we fall back to the implicit access
                return callback(null, implicitRole, canInteract);
            }

            // The resolved role is the one in the highest index of the DiscussionsConstants.roles.ALL_PRIORITY array
            var highestIndex = _.indexOf(DiscussionsConstants.roles.ALL_PRIORITY, implicitRole);
            _.each(roles, function(role) {
                highestIndex = Math.max(highestIndex, _.indexOf(DiscussionsConstants.roles.ALL_PRIORITY, role));
            });

            // The `canInteract` parameter is true because we can always interact if we have an explicit role
            return callback(null, DiscussionsConstants.roles.ALL_PRIORITY[highestIndex], true);
        });
    });
};

/**
 * Determine the highest **implicit** role that the user in context has on a discussion. Implicit means that the
 * user is granted access based on simple privacy / visibility / tenant rules associated to the context. Some
 * examples:
 *
 *  *   If the user in context is an admin, they will implicitly have manager access of the discussion, even if they
 *      don't have an explicit manager role membership on the discussion;
 *  *   If the discussion has visibility "public", all contexts (anonymous or authenticated) have implicit "viewer"
 *      role on a discussion
 *  *   If the discussion has visibility "private", no user context (except administrator) can have implicit "viewer"
 *      as that can only be determined through an **explicit** role check
 *
 * Therefore, the output from this call is the **minimum** effective permissions the current context has on a discussion,
 * since there is no concept of a "deny" permission. If this method returns that there is no implicit role, it is still
 * possible that the user has been explicitly granted a membership role. If this method returns that there is an implicit
 * "viewer" role, it is possible that the user has been explicitly granted a role of "manager".
 *
 * The `canInteract` determines if the user can implicitly "interact" with the discussion through potential tenant privacy
 * boundaries (e.g., share it, add it to their library, post a message on it). Like `implicitRole`, it is also a minimal
 * implicit check. So if this results in `false`, it is still possible that a user has an explicit manager role on the
 * discussion which grants it ability to interact.
 *
 * To get a context's explicit effective role, please use #_resolveEffectiveDiscussionRole instead.
 *
 * @param  {Context}    ctx                     The context of the current request
 * @param  {Discussion} discussion              The discussion to test access against
 * @param  {Function}   callback                Invoked when the process completes
 * @param  {Object}     callback.err            An error that occurred, if any
 * @param  {String}     callback.implicitRole   The implicit (minimum) role the user in context has on the discussion
 * @param  {Boolean}    callback.canInteract    Determines whether or not the user in context can implicitly interact with the discussion
 */
var _resolveImplicitRole = function(ctx, discussion, callback) {
    var user = ctx.user();

    // Check admin access
    if (user && (user.isGlobalAdmin() || user.isTenantAdmin(discussion.tenant.alias))) {
        // The user is an administrator, give them the highest allowable access
        return callback(null, DiscussionsConstants.roles.MANAGER, true);
    }

    // If we aren't an explicit manager, ability to interact is based on tenant privacy boundaries
    var canInteract = TenantsUtil.canInteract(ctx.tenant().alias, discussion.tenant.alias);

    // Check implicit member access
    if (discussion.visibility === LibraryConstants.visibility.PUBLIC) {
        // Public resources can be seen by anyone
        return callback(null, DiscussionsConstants.roles.MEMBER, canInteract);
    } else if (discussion.visibility === LibraryConstants.visibility.LOGGEDIN && TenantsUtil.isLoggedIn(ctx, discussion.tenant.alias)) {
        // We are from the same tenant as the "loggedin" resource, so we can view it and interact
        return callback(null, DiscussionsConstants.roles.MEMBER, canInteract);
    }

    return callback();
};
