// oae-zencoder/lib/internal/util.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var util = require('util');
var Zencoder = require('zencoder');

var log = require('oae-logger').logger('oae-zencoder');
var PreviewConstants = require('oae-preview-processor/lib/constants');
var TenantsAPI = require('oae-tenants/lib/api');
var ZencoderConfig = require('oae-config').config('oae-zencoder');

/**
 * Get the Zencoder API values that have been configured in the Admin UI.
 *
 * @return {Object} The Zencoder configuration for the tenant or the generic config for the host
 */
var getConfig = module.exports.getConfig = function(config, tenantAlias) {
    if (ZencoderConfig.getValue(tenantAlias, 'zencoder', 'enabled') && ZencoderConfig.getValue(tenantAlias, 'zencoder', 'apiKey')) {
        return _.extend(config, {
            'apiKey': ZencoderConfig.getValue(tenantAlias, 'zencoder', 'apiKey'),
            'watermark': ZencoderConfig.getValue(tenantAlias, 'zencoder', 'watermark') || null,
            'amazons3-region': ZencoderConfig.getValue(tenantAlias, 'zencoder', 'amazons3-region') || config['amazons3-region'],
            'amazons3-bucket': ZencoderConfig.getValue(tenantAlias, 'zencoder', 'amazons3-bucket') || config['amazons3-bucket']
        });
    } else {
        return config;
    }
};

/**
 * Create an encoding job in Zencoder.
 *
 * @param  {Object}     config              The default configuration
 * @param  {String}     tenantAlias         The alias of the tenant whose configuration to use
 * @param  {String}     path                The path to video/audio file
 * @param  {Function}   callback            Standard callback function
 * @param  {Error}      callback.err        An error that occurred, if any
 * @param  {Response}   callback.response   The HTTP Response object
 * @param  {String}     callback.body       The body of the response
 */
var job = module.exports.job = function(config, tenantAlias, path, contentId, revisionId, isVideo, callback) {
    config = getConfig(config, tenantAlias);
    var zencoder = new Zencoder(config.apiKey);
    var tenant = TenantsAPI.getTenant(tenantAlias);
    var baseUrl = util.format('s3://s3-%s.amazonaws.com/%s/%s', config['amazons3-region'], config['amazons3-bucket'], contentId);
    var filename = revisionId ? revisionId.replace(/:/g, '_') : contentId.replace(/:/g, '_');

    var options = {
        input: util.format('https://%s%s', tenant.host, path),
        label: filename,
        pass_through: contentId,
        private: true,
        output: {
            notifications: util.format('https://%s/api/zencoder/encodingNotification', tenant.host),
            base_url: baseUrl,
            public: true
        },
    };

    if (isVideo) {
        _.extend(options.output, {
                filename: filename + '.mp4',
                format: 'mp4',
                thumbnails: [
                    {
                        label: 'thumb',
                        number: 1,
                        base_url: baseUrl,
                        height: PreviewConstants.SIZES.IMAGE.LARGE,
                        width: PreviewConstants.SIZES.IMAGE.LARGE,
                        filename: 'thumb'
                    }
                ]
            });
        if (config.watermark) {
            _.extend(options.output, {
                    watermarks: [
                        {
                            url: config.watermark,
                            opacity: 0.3
                        }
                    ]
            });
        }
    } else {
        _.extend(options, {
                filename: filename + '.mp3',
                skip_video: true,
                format: 'mp3'
        });
    }

    log().info('Creating Zencoder job');
    zencoder.Job.create(options, function(err, data, res) {
        if (err) {
            return callback(err);
        }

        return callback(null, res, data);
    });
};
