// oae-zencoder/lib/internal/util.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var util = require('util');
var Zencoder = require('zencoder');

var log = require('oae-logger').logger('oae-zencoder');
var PreviewConstants = require('oae-preview-processor/lib/constants');
var TenantsAPI = require('oae-tenants/lib/api');
var ZencoderConfig = require('oae-config').config('oae-zencoder');

/**
 * Get the Zencoder API values that have been configured in the Admin UI.
 *
 * @return {Object} The Zencoder configuration for the tenant or the generic config for the host
 */
var getConfig = module.exports.getConfig = function(config, tenantAlias) {
    log().info(config.apiKey);

    if (ZencoderConfig.getValue(tenantAlias, 'zencoder', 'enabled') && ZencoderConfig.getValue(tenantAlias, 'zencoder', 'apiKey')) {
        return _.extend(config, {
            'apiKey': ZencoderConfig.getValue(tenantAlias, 'zencoder', 'apiKey'),
            'watermark': ZencoderConfig.getValue(tenantAlias, 'zencoder', 'watermark') || null,
            'output': ZencoderConfig.getValue(tenantAlias, 'zencoder', 'output') || config.output
        });
    } else {
        return config;
    }
};

/**
 * Create an encoding job in Zencoder.
 *
 * @param  {Object}     config              The default configuration
 * @param  {String}     tenantAlias         The alias of the tenant whose configuration to use
 * @param  {String}     path                The path to video/audio file
 * @param  {Function}   callback            Standard callback function
 * @param  {Error}      callback.err        An error that occurred, if any
 * @param  {Response}   callback.response   The HTTP Response object
 * @param  {String}     callback.body       The body of the response
 */
var job = module.exports.job = function(config, tenantAlias, path, contentId, revisionId, isVideo, callback) {
    var config = getConfig(config, tenantAlias);
    var zencoder = new Zencoder(config.apiKey);
    var tenant = TenantsAPI.getTenant(tenantAlias);
    var baseUrl = util.format('%s/%s', config.output, contentId);

    var options = {
        input: path,
        pass_through: contentId,
        private: true,
        output: {
            notifications: util.format('https://%s/%s', tenant.host, '/api/zencoder/encodingNotification'),
            base_url: baseUrl,
            public: true,
        }
    };

    if (isVideo) {
        _.extend(options.output, {
                filename: revisionId + '.mp4',
                height: PreviewConstants.SIZES.IMAGE.MEDIUM,
                format: 'mp4',
                thumbnails: [
                    {
                        label: 'large',
                        number: 1,
                        base_url: baseUrl,
                        width: PreviewConstants.SIZES.IMAGE.LARGE,
                        height: PreviewConstants.SIZES.IMAGE.LARGE,
                        filename: revisionId + '_l.png'
                    },
                    {
                        label: 'small',
                        number: 1,
                        base_url: baseUrl,
                        width: PreviewConstants.SIZES.IMAGE.SMALL,
                        height: PreviewConstants.SIZES.IMAGE.SMALL,
                        filename: revisionId + '_s.png'
                    }
                ]
            });
        }
        if (config.watermark) {
            _.extend(options.output, {
                    watermarks: [
                        {
                            url: config.watermark,
                            opacity: 0.3
                        }
                    ]
            });
    } else {
        _.extend(options, {
                filename: revisionId + '.mp3',
                skip_video: true,
                format: 'mp3'
        });
    }

    log().debug('Creating Zencoder-job with options %s', options);
    zencoder.Job.create(options, function(err, data, res) {
        if (err) {
            return callback(err);
        }

        return callback(null, res, data);
    });
};
