// oae-doc/lib/api.js
/*!
 * Copyright 2012 Sakai Foundation (SF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://www.osedu.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var fs = require('fs');
var _ = require('underscore');
var dox = require('dox');
var path = require('path');
var IO = require('oae-util/lib/io');
var Modules = require('oae-util/lib/modules');
var Validator = require('oae-util/lib/validator').Validator;
var log = require('oae-logger').logger('oae-doc');

var cachedDocs = {
  'backend': {},
  'frontend': {}
};

/**
 * Retrieves the docs for the requested module
 * 
 * @param  {String}     name            The name of the module to get docs for
 * @param  {String}     type            The type of the module to get docs for
 * @param  {Function}   callback        The function to call when complete
 * @param  {Object}     callback.doc    The documentation for the module requested
 */
var getDoc = module.exports.getDoc = function(name, type, callback) {
    var validator = new Validator();
    validator.check(name, {'code': 400, 'msg': 'Missing name'}).notEmpty();
    validator.check(type, {'code': 400, 'msg': 'Missing type'}).notEmpty();
    validator.check(type, {'code': 400, 'msg': 'The type is not valid'}).isIn(["backend","frontend"]);
    if (type === "backend") {  
      validator.check(name.substring(0, 4), {'code': 400, 'msg': 'No such OAE module'}).equals('oae-');
      validator.check(name, {'code': 400, 'msg': 'Module name can\'t contain \'.\''}).notContains('.');
    }
    if (validator.hasErrors()) {
      return callback(validator.getFirstError());
    }
    if (cachedDocs[type] && cachedDocs[type][name]) {
      return callback(null, cachedDocs[type][name]);
    } else {
      return callback({'code': 404, 'msg': 'No documentation for this module was found'},cachedDocs[type]);
    }
};

/**
 * Parse module doc into json
 *
 * @param   {string}      module          The module to get docs for
 * @param   {Function}    callback        The function to call when complete
 * @param   {Object}      callback.doc    The documentation for the module requested
 * @api private
 */
var readDoc = function(module, callback) {

    var dir = 'node_modules/' + module + '/lib';

    /*!
     * Filters out all non-javascript files and folders, as we don't want to generate documentation for these,
     * and returns the filtered array.
     *
     * @param  {String[]}   libs       The unfiltered Array of file names in the module lib directory
     * @return {String[]}              The returned, filtered, Array of file names that can be used to generate documentation for
     */
    var filterLibs = function(libs) {
        var allowedLibs = [];
        for (var i = 0; i < libs.length; i++) {
            if (libs[i].indexOf('.js') !== -1) {
                allowedLibs.push(libs[i]);
            }
        }
        return allowedLibs;
    };

    IO.getFileListForFolder(dir, function(err, libs) {
        if (err) {
            log().warn({'err': err, 'dir': dir}, 'Failed getting file list to parse dox documentation.');
            return callback({'code': 404, 'msg': 'No documentation for this module was found'});
        }
        
        // Filter out all non-javascript files
        libs = filterLibs(libs);

        var toDo = libs.length;
        var done = 0;
        var doc = {};

        for (var i=0; i < libs.length; i++) {
            var lib = libs[i];
            (function(lib) {
                IO.readFile(dir + '/' + lib, function(err, data) {
                    done++;
                    if (!err) {
                        try {
                            doc[lib] = dox.parseComments(data);
                        } catch (ex) {
                            log().warn({
                                err: ex,
                                data: data
                            }, 'Failed parsing comment data with dox for file %s. Ignoring.', dir + '/' + lib);
                        }
                    } else {
                        log().error('Failed reading ' + dir + '/' + lib);
                    }

                    if (done === toDo) {
                        return callback(err, doc);
                    }
                });
            })(lib);
        }

        if (!libs.length) {
            callback({'code': 404, 'msg': 'No documentation for this module was found'});
        }
    });
};
/**
 * Parse UI API Element doc into json
 *
 * @param   {string}      module            The frontend module to get docs for
 * @param   {Function}    callback          The function to call when complete
 * @param   {Object}      callback.doc      The documentation for the module requested
 * @api private
 */
var readUIDoc = function(module, callback) {
  var dir = '../3akai-ux/shared/oae/api';
  var doc = {};
  IO.readFile(dir + '/oae.api.' + module + '.js', function(err, data) {
    if (!err) {
      try {
        doc[module] = dox.parseComments(data);
      } catch (err) {
        log().warn({
          err: err,
          data: data
        }, 'Failed parsing comment data with dox for file');
      }
    } else {
      log().error('Failed reading' + module);
    }
    return callback(err, doc);
  });
};
/**
 * Initializes the documentation by fetching all modules + configuration for those modules and caching it
 * 
 * @param   {Function}   callback              The function to call when complete
 */
var initializeDocumentation = module.exports.initializeDocumentation = function(callback) {
  var modules = Modules.getAvailableModules();
  var toDo = modules.length;
  var done = 0;

  for (var i = 0; i < toDo; i++) {
    var module = modules[i];
    (function(module) {
      readDoc(module, function(err, doc) {
        done++;
        cachedDocs.backend[module] = doc;
        if( done === toDo)
          loadAvailableFrontendModules(callback);
      });
    })(module);
  }
};
/**
 * Function that takes all the frontend modules and puts it in an array
 * 
 * @param   {Function}   callback     The function to call when complete
 */
var loadAvailableFrontendModules = function (callback) {
  var _modules = [];
  IO.getFileListForFolder('../3akai-ux/shared/oae/api', function(modules) {
    _.forEach(modules, function(module) {
        var extension = path.extname(module);
        var basename = path.basename(module, extension);
        if (basename.substring(0, 7) === 'oae.api' && extension.toLowerCase() === '.js' && module !== 'oae.api.js') {
          _modules.push(basename.substr(8));
        }
    });
    initializeFrontendModules(_modules, callback);
  });
}
/**
 * Function that loads the frontend documentation into the cacheDocs
 * 
 * @param  {Array}      modules         Array with all the frontend module names
 * @param  {Function}   callback        The function to call when complete
 */
var initializeFrontendModules = function (modules, callback) { 
  var toDo = modules.length;
  var done = 0;

  for (var i = 0; i < toDo; i++) {
    var module = modules[i];
    (function(module) {
      readUIDoc(module, function(err, doc) {
        done++;
        cachedDocs.frontend[module] = doc;
        if( done === toDo)
          callback();
      });
    })(module);
  }
}; 

/**
 * Returns the available frontend module names from cache
 *
 * @return   {String[]}   Returns an Array of strings representing the names of the available frontend modules
 */
var getFrontendDocumentationNames = module.exports.getFrontendDocumentationNames = function() {
  return _.keys(cachedDocs.frontend);
};

/**
 * Returns the available backend module names from cache
 *
 * @return   {String[]}   Returns an Array of strings representing the names of the available backend modules
 */
var getBackendDocumentationNames = module.exports.getBackendDocumentationNames = function() {
  return _.keys(cachedDocs.backend);
}
