// oae-tests/rest/util.js
/*
 * Copyright 2012 Sakai Foundation (SF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://www.osedu.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var assert = require('assert');
var request = require('request');
var util = require('util');

/**
 * When we encounter an unexpected status code.
 *
 * @param {String}      message     A message explaining the error,
 * @param {Response}    response    The response as received from the server.
 */
var RestError = module.exports.RestError = function RestError(message, response) {
  this.name = 'RestError';
  this.message = message;
  this.code = response.statusCode;
  this.response = response;
};
util.inherits(RestError, Error);

/**
 * When we encounter a local problem (ie: not something that the server did.)
 *
 * @param {String}      message     A message explaining what went wrong.
 * @param {Response}    response    The response as received from the server.
 */
var OaeError = module.exports.OaeError = function OaeError(message, response) {
  this.name = 'OaeError';
  this.message = message;
  this.code = 500;
  this.request = response.request;
  this.response = response;
};
util.inherits(OaeError, Error);


/**
 * Perform a local user login using the REST API.
 *
 * @param {Context}         context         A context object that can be used to log on.
 * @param {Function(err)}   callback        Standard callback function
 * @param {Object}          callback.err    Error object containing error message (if any)
 */
var login = module.exports.login = function(context, callback) {
    request.post({
        'url': 'http://' + context.baseUrl + '/api/auth/login',
        'method': 'POST',
        'form': {
            'username': context.user.username,
            'password': context.user.password
        }
    },function(err, response, body) {
        if (err) {
            return callback(new OaeError('Something went wrong trying to contact the server.', response));
        } else if (response.statusCode !== 200) {
            return callback(new OaeError('Authentication was not successful: ' + body, response));
        }

        callback();
    });
};

/**
 * Perform a local user logout using the REST API.
 *
 * @param {Context}         context         A context object that can be used to log out.
 * @param {Function(err)}   callback        Standard callback function
 * @param {Object}          callback.err    Error object containing error message (if any)
 */
var logout = module.exports.logout = function(context, callback) {
    request.post({
        'uri': 'http://' + context.baseUrl + '/api/auth/logout',
        'followRedirect': false
    }, function(err, response, body) {
        if (err) {
            return callback(new OaeError('Something went wrong trying to contact the server.', response));
        }
        callback();
    });
};

/**
 * Performs a local user login using the REST API if the passed in context is different
 * than the currently logged in user.
 *
 * @param {Context}         context         A context object that can be used to log on.
 * @param {Function(err)}   callback        Standard callback function
 * @param {Object}          callback.err    Error object containing error message (if any)
 */
var switchUser = module.exports.switchUser = function (context, callback) {
    if (request.userId !== context.user.id) {
        logout(context, function(err) {
            if (err) {
                return callback(err);
            }
            login(context, function(err) {
                if (!err) {
                    request.userId = context.user.id;
                }
                callback(err);
            });
        });
    } else {
        callback();
    }
};

/**
 * Parses the JSON response in a safe way.
 * In case we can't parse it we pass back an OaeError with the body, request
 * and response attached to it
 *
 * @param {String}                  body                The response body
 * @param {Response}                response            The response object
 * @param {Function(err, resp)}     callback            A callback method
 * @param {Object}                  callback.err        Error object containing error message
 * @param {Object}                  callback.response   The parsed server response.
 */
var parseResponse = module.exports.parseResponse = function(body, response, callback) {
  try {
    var json = JSON.parse(body);
    callback(null, json);
  } catch (err) {
    callback(new OaeError('Couldn\'t parse the response.', response));
  }
};
