// oae-util/lib/io.js
/*!
 * Copyright 2014 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var fs = require('fs');

var log = require('oae-logger').logger('IO');

/**
 * Get a list of all of the files and folders inside of a folder. Hidden files and folder (starting with
 * '.', like .DS_STORE) will be excluded from the returned list.
 *
 * @param  {String}      foldername         Path of the folder for which we should get the containing files and folders
 * @param  {Function}    callback           Standard callback function
 * @param  {Object}      callback.err       An error that occurred, if any
 * @param  {String[]}    callback.files     Array containing all of the file and foldernames that exist inside of the given folder
 */
var getFileListForFolder = module.exports.getFileListForFolder = function(foldername, callback) {
    fs.exists(foldername, function(exists) {
        if (!exists) {
            return callback(null, []);
        }

        fs.stat(foldername, function(err, stat) {
            if (err) {
                return callback(err);
            } else if (!stat.isDirectory()) {
                return callback(null, []);
            }

            fs.readdir(foldername, function(err, files) {
                if (err) {
                    return callback(err);
                }

                var finalFiles = [];
                for (var f = 0; f < files.length; f++) {
                    if (files[f].substring(0, 1) !== '.') {
                        finalFiles.push(files[f]);
                    }
                }

                callback(null, finalFiles);
            });
        });

    });
};

/**
 * Copy a file
 *
 * @param  {String}   source        Path to the source file that needs to be copied
 * @param  {String}   dest          Path to the destination to which the file needs to be coppied
 * @param  {Function} callback      Standard callback function
 * @param  {Object}   callback.err  An error that occurred, if any
 */
var copyFile = module.exports.copyFile = function(source, dest, callback){
    var ins = fs.createReadStream(source);
    var outs = fs.createWriteStream(dest);
    // Clean up if there's an error reading the source file
    ins.once('error', function(err){
        destroyStream(outs);
        log().error({'err': err}, 'Wasn\'t able to copy the file %s to %s.', source, dest);
        callback({'code': 500, 'msg': err});
    });
    // Clean up if there's an error writing the destination file
    outs.once('error', function(err){
        ins.removeAllListeners('error');
        outs.removeAllListeners('close');
        ins.destroy();
        log().error({'err': err}, 'Wasn\'t able to copy the file %s to %s.', source, dest);
        callback({'code': 500, 'msg': err});
    });
    outs.once('close', function(){
        callback(null);
    });
    ins.pipe(outs);
};

/**
 * Move a file. This will try to rename it first. In case this has to go across partitions, this will
 * fall back to copy and delete
 *
 * @param  {String}   source        Path to the source file that needs to be moved
 * @param  {String}   dest          Path to the destination to which the file needs to be moved
 * @param  {Function} callback      Standard callback function
 * @param  {Object}   callback.err  An error that occurred, if any
 */
var moveFile = module.exports.moveFile = function(source, dest, callback){
    fs.rename(source, dest, function(err) {
        if (err) {
            // The `EXDEV` error will be thrown when a file is being moved across partitions
            // In that case, we copy and delete the file instead of moving it
            if (err.code !== 'EXDEV') {
                log().error({'err': err}, 'Wasn\'t able to rename the file  %s to %s.', source, dest);
                return callback({'code': 500, 'msg': err});
            } else {
                copyFile(source, dest, function(err) {
                    if (err) {
                        return callback({'code': 500, 'msg': err});
                    } else {
                        fs.unlink(source, callback);
                    }
                });
            }
        } else {
            callback(null);
        }
    });
};

/**
 * Destroy a stream and remove all the listeners.
 *
 * @param  {Stream}  stream  The stream to destroy
 */
var destroyStream = module.exports.destroyStream = function(stream) {
    stream.removeAllListeners('error');
    stream.removeAllListeners('close');
    stream.destroy();
};

/**
 * Test whether or not the given path exists by checking with the file system.
 *
 * This mimics the now deprecated `fs.exists()` function without suffering from
 * the same race condition.
 *
 * @param  {String}     path                The path to test
 * @param  {Function}   callback            Standard callback function
 * @param  {Object}     callback.err        An error that occurred, if any
 * @param  {Boolean}    callback.exists     Whether the path points to an existing file or directory
 */
var exists = module.exports.exists = function(path, callback) {
    fs.open(path, 'r', function(err) {
        if (err) {
            if (err.code === 'ENOENT') {
                return callback(null, false);
            } else {
                log().error({
                    'err': err,
                    'path': path
                }, 'Unable to check whether a file or folder exists');
                return callback({'code': 500, 'msg': 'Could not check whether a file or folder exists'});
            }
        } else {
            return callback(null, true);
        }
    });
};
