// oae-util/lib/server.js
/*!
 * Copyright 2012 Sakai Foundation (SF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://www.osedu.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var express = require('express');
var RedisStore = require('connect-redis')(express);

var TelemetryAPI = require('oae-telemetry');
var log = require('oae-logger').logger('oae-server');

// Maintains a list of paths that are safe from CSRF attacks
var safePathPrefixes = [];

/**
 * Starts an express server on the specified port. This will be done for the global admin server, as well
 * as for the tenant server.
 *
 * @param  {Number}     port        The port on which the express server should be started
 * @param  {Object}     config      JSON object containing configuration values for Cassandra, Redis, logging and telemetry
 * @return {Express}                The created express server
 */
module.exports.setupServer = function(port, config) {
    // Create the express server
    var server = express();
    // Make it listen for requests on the specified port
    server.listen(port);

    // Don't output pretty JSON,
    server.set('json spaces', 0);

    // We'll be using cookies to hash the principalId.
    // Signing of the cookie will be done by the session middleware.
    server.use(express.cookieParser());

    // The bodyParser will parse the following type of requests:
    // * application/json
    // * urlencoded (regular POST requests)
    // * multipart (file uploads)
    // We pass in the config.files hash so we can configure where
    // the uploaded files will be temporarily stored.
    server.use(express.bodyParser(config.files));

    // This needs to come BEFORE passport and AFTER cookieParser.
    // The secret will be used to sign the cookie
    // We'll use Redis as a store for our sessions.
    var redisStore = new RedisStore({
        'host': config.redis.host,
        'port': config.redis.port,
        'pass': config.redis.pass,
        'db': config.redis.dbIndex
    });
    server.use(express.session({'secret': config.cookie.secret, 'store': redisStore}));

    // Add telemetry before we do anything else.
    server.use(function(req, res, next) {
        TelemetryAPI.request(req, res);
        next();
    });

    /*!
     * Referer-based CSRF protection. If the request is not safe (e.g., POST, DELETE) and the origin of the request (as
     * specified by the HTTP Referer header) does not match the target host of the request (as specified by the HTTP
     * Host header), then the request will result in a 500 error.
     *
     * While referer-based protection is not highly recommended due to spoofing possibilities in insecure environments,
     * it currently offers the best trade-off between ease of use (e.g., for cURL interoperability), effort and security
     * against CSRF attacks.
     *
     * If using a utility such as `curl` to POST requests to the API, you can bypass this by just setting the referer
     * header to "/":
     *
     * curl -X POST -e / http://my.oae.com/api/auth/login
     *
     * More information about CSRF attacks: http://en.wikipedia.org/wiki/Cross-site_request_forgery
     */
    server.use(function(req, res, next) {
        if (!_isSafeMethod(req.method) && !_isSafePath(req) && !_isSameOrigin(req)) {
            log().warn({
                'method': req.method,
                'host': req.headers.host,
                'referer': req.headers.referer,
                'targetPath': req.path
            }, 'CSRF validation failed: attempted to execute unsafe operation from untrusted origin.');
            return _abort(res, 500, 'CSRF validation failed: attempted to execute unsafe method from untrusted origin.');
        }

        next();
    });

    return server;
};

/**
 * Add a path to the list of safe paths. Paths added here will not be protected against CSRF
 * attacks. This is common for endpoints that have other verification mechanisms such as Shibboleth.
 *
 * @param  {String}     pathPrefix  A path prefix that will not be validated against CSRF attacks
 */
var addSafePathPrefix = module.exports.addSafePathPrefix = function(pathPrefix) {
    log().info('Adding %s to list of paths that are not CSRF-protected.');
    safePathPrefixes.push(pathPrefix);
};

/**
 * This method is used to bind server functionality after all modules have had an opportunity to do so. This can be useful for things such
 * as:
 *
 *  * Response code logging / telemetry
 *  * Default "catch-all" error handling
 *
 * @param  {Express}    server      The express server for which the initialized should be finalized
 */
module.exports.postInitializeServer = function(server) {

    // Explicitly apply the routes to the end
    server.use(server.router);

    // Catch-all error handler
    var serverTelemetry = TelemetryAPI.telemetry('server');
    server.use(function(err, req, res, next) {
        serverTelemetry.incr('error');
        log(req.ctx).error({
            err: err,
            req: req,
            res: res
        }, 'Unhandled error in the request chain, caught at the default error handler.');
        res.send(500, 'An unexpected error occurred.');
    });
};

/**
 * Abort a request with a given code and response message.
 *
 * @param  {Response}   res     The express response object
 * @param  {Number}     code    The HTTP response code
 * @param  {String}     message The message body to provide as a reason for aborting the request
 * @api private
 */
var _abort = function(res, code, message) {
    res.setHeader('Connection', 'Close');
    return res.send(code, message);
};

/**
 * Determines if the target path for a request is considered "safe" from CSRF attacks.
 *
 * @param  {Request}    req     The express request object
 * @return {Boolean}            `true` if the path is safe from CSRF attacks, `false` otherwise
 * @api private
 */
var _isSafePath = function(req) {
    var path = req.path;
    var matchingPaths = _.filter(safePathPrefixes, function(safePathPrefix) {
        return (path.indexOf(safePathPrefix) === 0);
    });
    return (matchingPaths.length > 0);
};

/**
 * Determine whether or not the given request method is considered "safe"
 *
 * @param  {String}     method  The request method
 * @return {Boolean}            `true` if the request method is safe (e.g., GET, HEAD), `false` otherwise
 * @api private
 */
var _isSafeMethod = function(method) {
    return (method === 'GET' || method === 'HEAD');
};

/**
 * Determine whether or not the origin host of the given request is the same as the target host.
 *
 * @param  {Request}    req     The express request object to test
 * @return {Boolean}            `true` if the request is of the same origin as the target host, `false` otherwise
 * @api private
 */
var _isSameOrigin = function(req) {
    var host = req.headers.host;
    var referer = req.headers.referer;

    if (!referer) {
        return false;
    }

    if (referer.indexOf('/') !== 0) {
        // Verify the host portion against the host header
        referer = referer.split('://')[1];
        if (!referer || referer.split('/')[0] !== host) {
            // If there is nothing after the protocol (e.g., "http://") or the host before the first slash does not match
            // we deem it not to be the same origin.
            return false;
        } else {
            return true;
        }
    } else {
        // If the referer is a relative uri, it must be from same origin.
        return true;
    }
};
