// oae-util/lib/taskqueue.js
/*!
 * Copyright 2013 Apereo Foundation (AF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');

var MQ = require('./mq');

/**
 * ## The Task Queue API.
 *
 * This is a thin wrapper around the MQ utility abstracting most of the
 * pain of Exchanges, Queues, bindings, etc..
 *
 * It can be used as a simple task queue where tasks can be submitted to
 * and consumed from.
 */

var Constants = module.exports.Constants = {
    'DEFAULT_TASK_EXCHANGE_NAME': 'oae-taskexchange',
    'DEFAULT_TASK_EXCHANGE_OPTS': {
        'type': 'direct',
        'durable': true,
        'autoDelete': false
    },
    'DEFAULT_TASK_QUEUE_OPTS': {
        'durable': true,
        'autoDelete': false,
        'arguments': {
            // Additional information on highly available RabbitMQ queues can be found at http://www.rabbitmq.com/ha.html.
            // We use `all` as the policy: Queue is mirrored across all nodes in the cluster.
            // When a new node is added to the cluster, the queue will be mirrored to that node.
            'x-ha-policy': 'all'
        }
    },
    'DEFAULT_TASK_QUEUE_PUBLISH_OPTS': {
        'deliveryMode': 2                   // 2 indicates 'persistent'
    },
    'DEFAULT_TASK_QUEUE_SUBSCRIBE_OPTS': {
        'ack': true,
        'prefetchCount': 15
    }
};

/**
 * Initializes the task queue logic so that it can start sending and receiving tasks
 *
 * @param  {Function}   callback    Standard callback function
 */
var init = module.exports.init = function(callback) {
    MQ.declareExchange(Constants.DEFAULT_TASK_EXCHANGE_NAME, Constants.DEFAULT_TASK_EXCHANGE_OPTS, callback);
};


/**
 * A task queue is a simple queue where messages are considered tasks.
 * A queue will be created for each unique `taskQueueId`. The queue is bound to
 * the default exchange and new tasks will be published with the `taskQueue`
 * as routing key. This ensures that a submitted task ends up in the correct queue.
 *
 * @param  {String}     taskQueueId             The task queue to which the consumer should be bound
 * @param  {Function}   listener                A function that will be executed each time a task is received
 * @param  {Object}     listener.data           The data that is present in the task
 * @param  {Function}   listener.callback       A function that should be executed when the task has been completed
 * @param  {Object}     [options]               A set of options that can help with either declaring the queue or subscribing to it
 * @param  {Object}     [options.queue]         A set of options that can override the `Constants.DEFAULT_TASK_QUEUE_OPTS`.
 * @param  {Object}     [options.subscribe]     A set of options that can override the `Constants.DEFAULT_TASK_QUEUE_SUBSCRIBE_OPTS`.
 * @param  {Function}   callback                Standard callback function
 */
var bind = module.exports.bind = function(taskQueueId, listener, options, callback) {
    options = options || {};
    options.queue = options.queue || {};
    options.subscribe = options.subscribe || {};

    // 1. Declare the queue.
    var queueOptions = _.defaults(options.queue, Constants.DEFAULT_TASK_QUEUE_OPTS);
    _declareQueue(taskQueueId, queueOptions, function(err) {
        if (err) {
            return callback(err);
        }

        /*
         * 2. Bind queue to the default exchange
         *
         * We use the `taskQueueId` for both the name as the queue and the routing key.
         */
        MQ.bindQueueToExchange(taskQueueId, Constants.DEFAULT_TASK_EXCHANGE_NAME, taskQueueId, function(err) {
            if (err) {
                return callback(err);
            }


            // 3. Subscribe to the queue
            var subscribeOptions = _.defaults(options.subscribe, Constants.DEFAULT_TASK_QUEUE_SUBSCRIBE_OPTS);
            MQ.subscribeQueue(taskQueueId, subscribeOptions, listener, callback);
        });
    });
};

/**
 * Declares a queue as long as it hasn't been declared before
 *
 * @see bind
 */
var _declareQueue = function(taskQueueId, queueOptions, callback) {
    if (MQ.isQueueDeclared(taskQueueId)) {
        return callback();
    } else {
        MQ.declareQueue(taskQueueId, queueOptions, callback);
    }
};

/**
 * Stop consuming tasks from the task queue.
 *
 * @param  {String}     taskQueueId     The name of the task queue to stop consuming messages from.
 * @param  {Function}   callback        Standard callback function
 * @param  {Object}     callback.err    An error that occurred, if any
 */
var unbind = module.exports.unbind = function(taskQueueId, callback) {
    MQ.unsubscribeQueue(taskQueueId, callback);
};

/**
 * Submits a task to the queue.
 *
 * @param  {String}     taskQueueId     The name of the task queue where the data should be submitted to
 * @param  {Object}     taskData        The data that should be made available to the consumer of this task
 * @param  {Function}   callback        Standard callback function
 */
var submit = module.exports.submit = function(taskQueueId, taskData, callback) {
    MQ.submit(Constants.DEFAULT_TASK_EXCHANGE_NAME, taskQueueId, taskData, null, callback);
};
