// oae-util/lib/validator.js
/*
 * Copyright 2012 Sakai Foundation (SF) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 * 
 *     http://www.osedu.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('underscore');
var Validator = module.exports.Validator = require('validator').Validator;

/**
 * Wrapper function around node-validator that makes sure that a validation
 * Failure doesn't throw an error and just collects all of the errors
 * @param {Object}  msg     Error that should be recorded if the validation fails
 */
Validator.prototype.error = function (msg) {
    this._errors.push(msg);
};

/**
 * Wrapper function around node-validator that retrieves all of the validation
 * errors that have been encountered for a given Validator instance
 * @returns {Array<Object>}     Array containing all of the validation errors
 */
Validator.prototype.getErrors = function () {
    if (this._errors.length === 1) {
        return this._errors[0];
    } else if (this._errors.length === 0) {
        return null;
    }
    return this._errors;
};

/**
 * Check whether or not a given Validator instance has seen any validation errors
 * @returns {Boolean}     Returns true when validation errors have occured and false otherwise
 */
Validator.prototype.hasErrors = function () {
    return this._errors.length ? true : false;
};

///////////////////////
// Custom validators //
///////////////////////

/**
 * Check whether or not a user object represents an actual user
 * @param {User} user   User object that needs to be checked for validity
 */
Validator.prototype.isUser = function(user) {
    if (!_.isObject(user) || !user.id) {
        this.error(this.msg || 'Not a valid user');
    }
    return this;
};
